<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.6.3" />
<title>cnvpytor.root API documentation</title>
<meta name="description" content="cnvpytor.root â€¦" />
<link href='https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.0/normalize.min.css' rel='stylesheet'>
<link href='https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/8.0.0/sanitize.min.css' rel='stylesheet'>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet">
<style>.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{font-weight:bold}#index h4 + ul{margin-bottom:.6em}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase;cursor:pointer}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>cnvpytor.root</code></h1>
</header>
<section id="section-intro">
<p>cnvpytor.root</p>
<p>class Root: main CNVpytor class</p>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">&#34;&#34;&#34; cnvpytor.root

class Root: main CNVpytor class
&#34;&#34;&#34;
from __future__ import absolute_import, print_function, division

from .io import *
from .bam import Bam
from .vcf import Vcf
from .fasta import Fasta
from .genome import Genome
import numpy as np
import logging

_logger = logging.getLogger(&#34;cnvpytor.root&#34;)


class Root:

    def __init__(self, filename, max_cores=8):
        &#34;&#34;&#34;
        Class constructor opens CNVpytor file. The class implements all core CNVpytor calculations.

        Parameters
        ----------
        filename : str
            CNVpytor filename
        max_cores : int
            Maximal number of cores used in parallel calculations

        &#34;&#34;&#34;
        _logger.debug(&#34;App class init: filename &#39;%s&#39;; max cores %d.&#34; % (filename, max_cores))
        self.io = IO(filename)
        self.max_cores = max_cores
        self.io_gc = self.io
        self.io_mask = self.io
        if self.io.signal_exists(None, None, &#34;reference genome&#34;) and self.io.signal_exists(None, None, &#34;use reference&#34;):
            rg_name = np.array(self.io.get_signal(None, None, &#34;reference genome&#34;)).astype(&#34;str&#34;)[0]
            if rg_name in Genome.reference_genomes:
                rg_use = self.io.get_signal(None, None, &#34;use reference&#34;)
                if &#34;gc_file&#34; in Genome.reference_genomes[rg_name] and rg_use[0] == 1:
                    _logger.debug(&#34;Using GC content from database for reference genome &#39;%s&#39;.&#34; % rg_name)
                    self.io_gc = IO(Genome.reference_genomes[rg_name][&#34;gc_file&#34;], ro=True, buffer=True)
                if &#34;mask_file&#34; in Genome.reference_genomes[rg_name] and rg_use[1] == 1:
                    _logger.debug(&#34;Using strict mask from database for reference genome &#39;%s&#39;.&#34; % rg_name)
                    self.io_mask = IO(Genome.reference_genomes[rg_name][&#34;mask_file&#34;], ro=True, buffer=True)

    def read_bam(self, bf, chroms):
        bamf = Bam(bf)
        if bamf.reference_genome:
            self.io.create_signal(None, None, &#34;reference genome&#34;, np.array([np.string_(bamf.reference_genome)]))
            self.io.create_signal(None, None, &#34;use reference&#34;, np.array([1, 1]).astype(&#34;uint8&#34;))

        def read_chromosome(x):
            chrs, lens = x
            _logger.info(&#34;Reading data for chromosome %s with length %d&#34; % (chrs, lens))
            l_rd_p, l_rd_u, l_his_read_frg = bamf.read_chromosome(chrs)
            return l_rd_p, l_rd_u, l_his_read_frg

        chrname, chrlen = bamf.get_chr_len()
        chr_len = [(c, l) for (c, l) in zip(chrname, chrlen) if len(chroms) == 0 or c in chroms]
        if self.max_cores == 1:
            count = 0
            cum_his_read_frg = None
            for cl in chr_len:
                rd_p, rd_u, his_read_frg = read_chromosome(cl)
                if not rd_p is None:
                    if cum_his_read_frg is None:
                        cum_his_read_frg = his_read_frg
                    else:
                        cum_his_read_frg += his_read_frg
                    self.io.save_rd(cl[0], rd_p, rd_u)
                    count += 1
            if not cum_his_read_frg is None:
                self.io.create_signal(None, None, &#34;read frg dist&#34;, cum_his_read_frg)
            return count
        else:
            from .pool import parmap
            res = parmap(read_chromosome, chr_len, cores=self.max_cores)
            count = 0
            cum_his_read_frg = None
            for c, r in zip(chr_len, res):
                if not r[0] is None:
                    if cum_his_read_frg is None:
                        cum_his_read_frg = r[2]
                    else:
                        cum_his_read_frg += r[2]
                    self.io.save_rd(c[0], r[0], r[1])
                    count += 1
            if not cum_his_read_frg is None:
                self.io.create_signal(None, None, &#34;read frg dist&#34;, cum_his_read_frg)
            return count

    def rd_stat(self, chroms=[], plot=True):
        rd_stat = []
        auto, sex, mt = False, False, False

        rd_gc_chromosomes = {}
        for c in self.io_gc.gc_chromosomes():
            rd_name = self.io.rd_chromosome_name(c)
            if not rd_name is None:
                rd_gc_chromosomes[rd_name] = c

        for c in rd_gc_chromosomes:
            if (len(chroms) == 0 or c in chroms):
                rd_p, rd_u = self.io.read_rd(c)
                max_bin = max(int(10 * np.mean(rd_u) + 1), int(10 * np.mean(rd_p) + 1))
                dist_p, bins = np.histogram(rd_p, bins=range(max_bin + 1))
                dist_u, bins = np.histogram(rd_u, bins=range(max_bin + 1))
                n_p, m_p, s_p = fit_normal(bins[1:-1], dist_p[1:])[0]
                n_u, m_u, s_u = fit_normal(bins[1:-1], dist_u[1:])[0]
                _logger.info(
                    &#34;Chromosome &#39;%s&#39; stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (c, m_p, s_p))
                _logger.info(
                    &#34;Chromosome &#39;%s&#39; stat - RD unique distribution gaussian fit:  %.2f +- %.2f&#34; % (c, m_u, s_u))
                rd_stat.append((c, m_p, s_p, m_u, s_u))
                auto = auto or Genome.is_autosome(c)
                sex = sex or Genome.is_sex_chrom(c)
                mt = mt or Genome.is_mt_chrom(c)

        if auto:
            max_bin_auto = int(
                max(map(lambda x: 5 * x[1] + 5 * x[2], filter(lambda x: Genome.is_autosome(x[0]), rd_stat))))
            _logger.debug(&#34;Max RD for autosomes calculated: %d&#34; % max_bin_auto)
            dist_p_auto = np.zeros(max_bin_auto)
            dist_u_auto = np.zeros(max_bin_auto)
            dist_p_gc_auto = np.zeros((max_bin_auto, 101))
            bins_auto = range(max_bin_auto + 1)
        if sex:
            max_bin_sex = int(
                max(map(lambda x: 5 * x[1] + 5 * x[2], filter(lambda x: Genome.is_sex_chrom(x[0]), rd_stat))))
            _logger.debug(&#34;Max RD for sex chromosomes calculated: %d&#34; % max_bin_sex)
            dist_p_sex = np.zeros(max_bin_sex)
            dist_u_sex = np.zeros(max_bin_sex)
            dist_p_gc_sex = np.zeros((max_bin_sex, 101))
            bins_sex = range(max_bin_sex + 1)

        if mt:
            max_bin_mt = int(
                max(map(lambda x: 5 * x[1] + 5 * x[2], filter(lambda x: Genome.is_mt_chrom(x[0]), rd_stat))))
            _logger.debug(&#34;Max RD for mitochondria calculated: %d&#34; % max_bin_mt)
            bin_size_mt = max_bin_mt // 100
            bins_mt = range(0, max_bin_mt // bin_size_mt * bin_size_mt + bin_size_mt, bin_size_mt)
            n_bins_mt = len(bins_mt) - 1
            _logger.debug(&#34;Using %d bin size, %d bins for mitochondria chromosome.&#34; % (bin_size_mt, n_bins_mt))
            dist_p_mt = np.zeros(n_bins_mt)
            dist_u_mt = np.zeros(n_bins_mt)
            dist_p_gc_mt = np.zeros((n_bins_mt, 101))

        for c in rd_gc_chromosomes:
            if len(chroms) == 0 or c in chroms:
                _logger.info(&#34;Chromosome &#39;%s&#39; stat &#34; % c)
                rd_p, rd_u = self.io.read_rd(c)
                if Genome.is_autosome(c):
                    dist_p, bins = np.histogram(rd_p, bins=bins_auto)
                    dist_u, bins = np.histogram(rd_u, bins=bins_auto)
                    gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                    gcp = gcp_decompress(gcat)
                    dist_p_gc, xbins, ybins = np.histogram2d(rd_p, gcp, bins=(bins_auto, range(102)))
                    dist_p_auto += dist_p
                    dist_u_auto += dist_u
                    dist_p_gc_auto += dist_p_gc
                elif Genome.is_sex_chrom(c):
                    dist_p, bins = np.histogram(rd_p, bins=bins_sex)
                    dist_u, bins = np.histogram(rd_u, bins=bins_sex)
                    gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                    gcp = gcp_decompress(gcat)
                    dist_p_gc, xbins, ybins = np.histogram2d(rd_p, gcp, bins=(bins_sex, range(102)))
                    dist_p_sex += dist_p
                    dist_u_sex += dist_u
                    dist_p_gc_sex += dist_p_gc
                elif Genome.is_mt_chrom(c):
                    dist_p, bins = np.histogram(rd_p, bins=bins_mt)
                    dist_u, bins = np.histogram(rd_u, bins=bins_mt)
                    gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                    gcp = gcp_decompress(gcat)
                    dist_p_gc, xbins, ybins = np.histogram2d(rd_p, gcp, bins=(bins_mt, range(102)))
                    dist_p_mt += dist_p
                    dist_u_mt += dist_u
                    dist_p_gc_mt += dist_p_gc

        if auto:
            n_auto, m_auto, s_auto = fit_normal(np.array(bins_auto[1:-1]), dist_p_auto[1:])[0]
            _logger.info(&#34;Autosomes stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (m_auto, s_auto))
            stat_auto = np.array([max_bin_auto, 1, max_bin_auto, n_auto, m_auto, s_auto])
            self.io.create_signal(None, 100, &#34;RD p dist&#34;, dist_p_auto, flags=FLAG_AUTO)
            self.io.create_signal(None, 100, &#34;RD u dist&#34;, dist_u_auto, flags=FLAG_AUTO)
            self.io.create_signal(None, 100, &#34;RD GC dist&#34;, dist_p_gc_auto, flags=FLAG_AUTO)
            self.io.create_signal(None, 100, &#34;RD stat&#34;, stat_auto, flags=FLAG_AUTO)
            gc_corr_auto = calculate_gc_correction(dist_p_gc_auto, m_auto, s_auto)
            self.io.create_signal(None, 100, &#34;GC corr&#34;, gc_corr_auto, flags=FLAG_AUTO)

        if sex:
            n_sex, m_sex, s_sex = fit_normal(np.array(bins_sex[1:-1]), dist_p_sex[1:])[0]
            _logger.info(&#34;Sex chromosomes stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (m_sex, s_sex))
            stat_sex = np.array([max_bin_sex, 1, max_bin_sex, n_sex, m_sex, s_sex])
            self.io.create_signal(None, 100, &#34;RD p dist&#34;, dist_p_sex, flags=FLAG_SEX)
            self.io.create_signal(None, 100, &#34;RD u dist&#34;, dist_u_sex, flags=FLAG_SEX)
            self.io.create_signal(None, 100, &#34;RD GC dist&#34;, dist_p_gc_sex, flags=FLAG_SEX)
            self.io.create_signal(None, 100, &#34;RD stat&#34;, stat_sex, flags=FLAG_SEX)
            gc_corr_sex = calculate_gc_correction(dist_p_gc_sex, m_sex, s_sex)
            self.io.create_signal(None, 100, &#34;GC corr&#34;, gc_corr_sex, flags=FLAG_SEX)

        if mt:
            n_mt, m_mt, s_mt = fit_normal(np.array(bins_mt[1:-1]), dist_p_mt[1:])[0]
            _logger.info(&#34;Mitochondria stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (m_mt, s_mt))
            if auto:
                _logger.info(&#34;Mitochondria stat - number of mitochondria per cell:  %.2f +- %.2f&#34; % (
                    2. * m_mt / m_auto, 2. * s_mt / m_auto + s_auto * m_mt / (m_auto * m_auto)))
            stat_mt = np.array([max_bin_mt, bin_size_mt, n_bins_mt, n_mt, m_mt, s_mt])
            self.io.create_signal(None, 100, &#34;RD p dist&#34;, dist_p_mt, flags=FLAG_MT)
            self.io.create_signal(None, 100, &#34;RD u dist&#34;, dist_u_mt, flags=FLAG_MT)
            self.io.create_signal(None, 100, &#34;RD GC dist&#34;, dist_p_gc_mt, flags=FLAG_MT)
            self.io.create_signal(None, 100, &#34;RD stat&#34;, stat_mt, flags=FLAG_MT)
            gc_corr_mt = calculate_gc_correction(dist_p_gc_mt, m_mt, s_mt, bin_size_mt)
            self.io.create_signal(None, 100, &#34;GC corr&#34;, gc_corr_mt, flags=FLAG_MT)

    def read_vcf(self, vcf_file, chroms, sample=&#39;&#39;, use_index=False):
        &#34;&#34;&#34;

        Parameters
        ----------
        vcf_file
        chroms
        sample
        use_index

        Returns
        -------

        &#34;&#34;&#34;
        vcff = Vcf(vcf_file)
        chrs = [c for c in vcff.get_chromosomes() if len(chroms) == 0 or c in chroms]

        def save_data(chr, pos, ref, alt, nref, nalt, gt, flag, qual):
            if (len(chroms) == 0 or chr in chroms) and (not pos is None) and (len(pos) &gt; 0):
                self.io.save_snp(chr, pos, ref, alt, nref, nalt, gt, flag, qual)
            # TODO: Stop reading if all form chrom list are read.

        if use_index:
            count = 0
            for c in chrs:
                _logger.info(&#34;Reading variant data for chromosome %s&#34; % c)
                pos, ref, alt, nref, nalt, gt, flag, qual = vcff.read_chromosome_snp(c, sample)

                if not pos is None and len(pos) &gt; 0:
                    self.io.save_snp(c, pos, ref, alt, nref, nalt, gt, flag, qual)
                    count += 1
            return count
        else:
            return vcff.read_all_snp(save_data, sample)

    def rd(self, bamfiles, chroms=[]):
        &#34;&#34;&#34; Read chromosomes from bam/sam/cram file(s) and store in .cnvnator file
                Arguments:
                    * list of bam file names
                    * list of chromosome names
        &#34;&#34;&#34;
        hm = 0
        for bf in bamfiles:
            hm += self.read_bam(bf, chroms)

        if self.io.signal_exists(None, None, &#34;reference genome&#34;):
            rg_name = np.array(self.io.get_signal(None, None, &#34;reference genome&#34;)).astype(&#34;str&#34;)[0]
            if &#34;mask_file&#34; in Genome.reference_genomes[rg_name]:
                _logger.info(&#34;Strict mask for reference genome &#39;%s&#39; found in database.&#34; % rg_name)
                self.io_mask = IO(Genome.reference_genomes[rg_name][&#34;mask_file&#34;])
            if &#34;gc_file&#34; in Genome.reference_genomes[rg_name]:
                _logger.info(&#34;GC content for reference genome &#39;%s&#39; found in database.&#34; % rg_name)
                self.io_gc = IO(Genome.reference_genomes[rg_name][&#34;gc_file&#34;])
                self.rd_stat()

        return hm

    def vcf(self, vcf_files, chroms=[], sample=&#39;&#39;):
        &#34;&#34;&#34; Read SNP data from variant file(s) and store in .cnvnator file
                Arguments:
                    * list of variant file names
                    * list of chromosome names
        &#34;&#34;&#34;
        hm = 0
        for vcf_file in vcf_files:
            hm += self.read_vcf(vcf_file, chroms, sample)
        return hm

    def pileup_bam(self, bamfile, chroms, pos, ref, alt, nref, nalt):
        &#34;&#34;&#34;
        TODO:

        Parameters
        ----------
        bamfile
        chroms

        &#34;&#34;&#34;
        _logger.info(&#34;Calculating pileup from bam file &#39;%s&#39;.&#34; % bamfile)
        bamf = Bam(bamfile)

        def pileup_chromosome(x):
            c, l, snpc = x
            _logger.info(&#34;Pileup chromosome %s with length %d&#34; % (c, l))
            return bamf.pileup(c, pos[snpc], ref[snpc], alt[snpc])

        chrname, chrlen = bamf.get_chr_len()
        chr_len = [(c, l, self.io.snp_chromosome_name(c)) for (c, l) in zip(chrname, chrlen) if
                   self.io.snp_chromosome_name(c) in chroms]

        if self.max_cores == 1:
            for cl in chr_len:
                r = pileup_chromosome(cl)
                nref[cl[2]] = [x + y for x, y in zip(nref[cl[2]], r[0])]
                nalt[cl[2]] = [x + y for x, y in zip(nalt[cl[2]], r[1])]

        else:
            from .pool import parmap
            res = parmap(pileup_chromosome, chr_len, cores=self.max_cores)
            for cl, r in zip(chr_len, res):
                nref[cl[2]] = [x + y for x, y in zip(nref[cl[2]], r[0])]
                nalt[cl[2]] = [x + y for x, y in zip(nalt[cl[2]], r[1])]

    def pileup(self, bamfiles, chroms=[]):
        &#34;&#34;&#34;
        Read bam/sam/cram file(s) and pile up SNP counts at positions imported with vcf(vcf_files, ...) method

        Parameters
        ----------
        bamfiles : list of str
            Bam files.
        &#34;&#34;&#34;
        chrs = [c for c in self.io.snp_chromosomes() if (len(chroms) == 0 or c in chroms)]

        pos, ref, alt, nref, nalt, gt, flag, qual = {}, {}, {}, {}, {}, {}, {}, {}
        for c in chrs:
            _logger.info(&#34;Decompressing and setting all SNP counts to zero for chromosome &#39;%s&#39;.&#34; % c)
            pos[c], ref[c], alt[c], nref[c], nalt[c], gt[c], flag[c], qual[c] = self.io.read_snp(c)
            nref[c] = [0] * len(nref[c])
            nalt[c] = [0] * len(nalt[c])
        for bf in bamfiles:
            self.pileup_bam(bf, chrs, pos, ref, alt, nref, nalt)

        for c in chrs:
            _logger.info(&#34;Saving SNP data for chromosome &#39;%s&#39; in file &#39;%s&#39;.&#34; % (c, self.io.filename))
            self.io.save_snp(c, pos[c], ref[c], alt[c], nref[c], nalt[c], gt[c], flag[c], qual[c])

    def gc(self, filename, chroms=[], make_gc_genome_file=False):
        &#34;&#34;&#34; Read GC content from reference fasta file and store in .cnvnator file
                Arguments:
                    * FASTA filename
                    * list of chromosome names
        &#34;&#34;&#34;
        fasta = Fasta(filename)
        chrname, chrlen = fasta.get_chr_len()
        chr_len_rdn = []
        if make_gc_genome_file:
            _logger.debug(&#34;GC data for all reference genome contigs will be read and stored in cnvnator file.&#34;)
            for (c, l) in zip(chrname, chrlen):
                chr_len_rdn.append((c, l, c))
        else:
            for (c, l) in zip(chrname, chrlen):
                rd_name = self.io.rd_chromosome_name(c)
                if len(chroms) == 0 or (rd_name in chroms) or (c in chroms):
                    if rd_name is None:
                        _logger.debug(&#34;Not found RD signal for chromosome &#39;%s&#39;. Skipping...&#34; % c)
                    else:
                        _logger.debug(&#34;Found RD signal for chromosome &#39;%s&#39; with name &#39;%s&#39;.&#34; % (c, rd_name))
                        chr_len_rdn.append((c, l, rd_name))
        count = 0
        for c, l, rdn in chr_len_rdn:
            _logger.info(&#34;Reading data for chromosome &#39;%s&#39; with length %d&#34; % (c, l))
            gc, at = fasta.read_chromosome_gc(c)
            if not gc is None:
                self.io.create_signal(rdn, None, &#34;GC/AT&#34;, gc_at_compress(gc, at))
                count += 1
                _logger.info(&#34;GC data for chromosome &#39;%s&#39; saved in file &#39;%s&#39;.&#34; % (rdn, self.io.filename))
        if count &gt; 0:
            _logger.info(&#34;Running RD statistics on chromosomes with imported GC data.&#34;)
            self.rd_stat()
        return count

    def copy_gc(self, filename, chroms=[]):
        &#34;&#34;&#34; Copy GC content from another cnvnator file
                Arguments:
                    * cnvnator filename
                    * list of chromosome names
        &#34;&#34;&#34;
        io_src = IO(filename)
        chr_rdn = []
        for c in io_src.chromosomes_with_signal(None, &#34;GC/AT&#34;):
            rd_name = self.io.rd_chromosome_name(c)
            if len(chroms) == 0 or (rd_name in chroms) or (c in chroms):
                if rd_name is None:
                    _logger.debug(&#34;Not found RD signal for chromosome &#39;%s&#39;. Skipping...&#34; % c)
                else:
                    _logger.debug(&#34;Found RD signal for chromosome &#39;%s&#39; with name &#39;%s&#39;.&#34; % (c, rd_name))
                    chr_rdn.append((c, rd_name))
        count = 0
        for c, rdn in chr_rdn:
            _logger.info(
                &#34;Copying GC data for chromosome &#39;%s&#39; from file &#39;%s&#39; in file &#39;%s&#39;.&#34; % (c, filename, self.io.filename))
            self.io.create_signal(rdn, None, &#34;GC/AT&#34;, io_src.get_signal(c, None, &#34;GC/AT&#34;).astype(dtype=&#34;uint8&#34;))
        return count

    def mask(self, filename, chroms=[], make_mask_genome_file=False):
        &#34;&#34;&#34; Read strict mask fasta file and store in .cnvnator file
                Arguments:
                    * FASTA filename
                    * list of chromosome names
        &#34;&#34;&#34;
        fasta = Fasta(filename)
        chrname, chrlen = fasta.get_chr_len()
        chr_len_rdn = []
        if make_mask_genome_file:
            _logger.debug(&#34;Strict mask data for all reference genome contigs will be read and stored in cnvnator file.&#34;)
            for (c, l) in zip(chrname, chrlen):
                chr_len_rdn.append((c, l, c))
        else:
            for (c, l) in zip(chrname, chrlen):
                rd_name = self.io.rd_chromosome_name(c)
                snp_name = self.io.snp_chromosome_name(c)
                if len(chroms) == 0 or (rd_name in chroms) or (snp_name in chroms) or (c in chroms):
                    if (rd_name is None) and (snp_name is None):
                        _logger.debug(&#34;Not found RD or SNP signal for chromosome &#39;%s&#39;. Skipping...&#34; % c)
                    elif (rd_name is None):
                        _logger.debug(&#34;Found SNP signal for chromosome &#39;%s&#39; with name &#39;%s&#39;.&#34; % (c, snp_name))
                        chr_len_rdn.append((c, l, snp_name))
                    else:
                        _logger.debug(&#34;Found RD signal for chromosome &#39;%s&#39; with name &#39;%s&#39;.&#34; % (c, rd_name))
                        chr_len_rdn.append((c, l, rd_name))
        count = 0
        for c, l, rdn in chr_len_rdn:
            _logger.info(&#34;Reading data for chromosome &#39;%s&#39; with length %d&#34; % (c, l))
            mask = fasta.read_chromosome_mask_p_regions(c)
            if not mask is None:
                self.io.create_signal(rdn, None, &#34;mask&#34;, mask_compress(mask))
                count += 1
                _logger.info(&#34;Strict mask data for chromosome &#39;%s&#39; saved in file &#39;%s&#39;.&#34; % (rdn, self.io.filename))
        return count

    def copy_mask(self, filename, chroms=[]):
        &#34;&#34;&#34; Copy strict mask data from another cnvnator file
                Arguments:
                    * cnvnator filename
                    * list of chromosome names
        &#34;&#34;&#34;
        io_src = IO(filename)
        chr_rdn = []
        for c in io_src.chromosomes_with_signal(None, &#34;mask&#34;):
            rd_name = self.io.rd_chromosome_name(c)
            snp_name = self.snp_chromosome_name(c)
            if len(chroms) == 0 or (rd_name in chroms) or (snp_name in chroms) or (c in chroms):
                if (rd_name is None) and (snp_name is None):
                    _logger.debug(&#34;Not found RD or SNP signal for chromosome &#39;%s&#39;. Skipping...&#34; % c)
                elif (rd_name is None):
                    _logger.debug(&#34;Found SNP signal for chromosome &#39;%s&#39; with name &#39;%s&#39;.&#34; % (c, snp_name))
                    chr_rdn.append((c, snp_name))
                else:
                    _logger.debug(&#34;Found RD signal for chromosome &#39;%s&#39; with name &#39;%s&#39;.&#34; % (c, rd_name))
                    chr_rdn.append((c, rd_name))
        count = 0
        for c, rdn in chr_rdn:
            _logger.info(
                &#34;Copying strict mask data for chromosome &#39;%s&#39; from file &#39;%s&#39; in file &#39;%s&#39;.&#34; % (
                    c, filename, self.io.filename))
            self.io.create_signal(rdn, None, &#34;mask&#34;, io_src.get_signal(c, None, &#34;mask&#34;).astype(dtype=&#34;uint32&#34;))
        return count

    def set_reference_genome(self, rg):
        &#34;&#34;&#34; Manually sets reference genome.

        Parameters
        ----------
        rg: str
            Name of reference genome

        Returns
        -------
            True if genome exists in database

        &#34;&#34;&#34;
        if rg in Genome.reference_genomes:
            self.io.create_signal(None, None, &#34;reference genome&#34;, np.array([np.string_(rg)]))
            self.io.create_signal(None, None, &#34;use reference&#34;, np.array([1, 1]).astype(&#34;uint8&#34;))

    def calculate_histograms(self, bin_sizes, chroms=[]):
        &#34;&#34;&#34;
        Calculates RD histograms and store data into cnvpytor file.

        Parameters
        ----------
        bin_sizes : list of int
            List of histogram bin sizes
        chroms : list of str
            List of chromosomes. Calculates for all available if empty.

        &#34;&#34;&#34;
        rd_gc_chromosomes = {}
        for c in self.io_gc.gc_chromosomes():
            rd_name = self.io.rd_chromosome_name(c)
            if not rd_name is None and (len(chroms) == 0 or (rd_name in chroms) or (c in chroms)):
                rd_gc_chromosomes[rd_name] = c

        rd_mask_chromosomes = {}
        for c in self.io_mask.mask_chromosomes():
            rd_name = self.io.rd_chromosome_name(c)
            if not rd_name is None and (len(chroms) == 0 or (rd_name in chroms) or (c in chroms)):
                rd_mask_chromosomes[rd_name] = c

        for bin_size in bin_sizes:
            his_stat = []
            auto, sex, mt = False, False, False
            bin_ratio = bin_size // 100

            for c in self.io.rd_chromosomes():
                if c in rd_gc_chromosomes:
                    _logger.info(&#34;Calculating histograms using bin size %d for chromosome &#39;%s&#39;.&#34; % (bin_size, c))
                    flag = FLAG_MT if Genome.is_mt_chrom(c) else FLAG_SEX if Genome.is_sex_chrom(c) else FLAG_AUTO
                    rd_p, rd_u = self.io.read_rd(c)
                    his_p = np.concatenate(
                        (rd_p, np.zeros(bin_ratio - len(rd_p) + (len(rd_p) // bin_ratio * bin_ratio))))
                    his_p.resize((len(his_p) // bin_ratio, bin_ratio))
                    his_p = his_p.sum(axis=1)
                    his_u = np.concatenate(
                        (rd_u, np.zeros(bin_ratio - len(rd_u) + (len(rd_u) // bin_ratio * bin_ratio))))
                    his_u.resize((len(his_u) // bin_ratio, bin_ratio))
                    his_u = his_u.sum(axis=1)
                    self.io.create_signal(c, bin_size, &#34;RD&#34;, his_p)
                    self.io.create_signal(c, bin_size, &#34;RD unique&#34;, his_u)

                    max_bin = max(int(10 * np.mean(his_u) + 1), int(10 * np.mean(his_p) + 1))
                    rd_bin_size = max_bin // 10000
                    if rd_bin_size == 0:
                        rd_bin_size = 1
                    rd_bins = range(0, max_bin // rd_bin_size * rd_bin_size + rd_bin_size, rd_bin_size)
                    dist_p, bins = np.histogram(his_p, bins=rd_bins)
                    dist_u, bins = np.histogram(his_u, bins=rd_bins)
                    n_p, m_p, s_p = fit_normal(bins[1:-1], dist_p[1:])[0]
                    n_u, m_u, s_u = fit_normal(bins[1:-1], dist_u[1:])[0]
                    _logger.info(
                        &#34;Chromosome &#39;%s&#39; bin size %d stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (
                            c, bin_size, m_p, s_p))
                    _logger.info(
                        &#34;Chromosome &#39;%s&#39; bin size %d stat - RD unique distribution gaussian fit:  %.2f +- %.2f&#34; % (
                            c, bin_size, m_u, s_u))
                    his_stat.append((c, m_p, s_p, m_u, s_u))
                    auto = auto or Genome.is_autosome(c)
                    sex = sex or Genome.is_sex_chrom(c)
                    mt = mt or Genome.is_mt_chrom(c)

            if auto:
                max_bin_auto = int(
                    max(map(lambda x: 5 * x[1] + 5 * x[2], filter(lambda x: Genome.is_autosome(x[0]), his_stat))))
                _logger.debug(&#34;Max RD for autosome histograms calculated: %d.&#34; % max_bin_auto)
                bin_size_auto = max_bin_auto // 1000
                if bin_size_auto == 0:
                    bin_size_auto = 1
                bins_auto = range(0, max_bin_auto // bin_size_auto * bin_size_auto + bin_size_auto, bin_size_auto)
                n_bins_auto = len(bins_auto) - 1
                _logger.debug(
                    &#34;Using %d RD bin size, %d bins for autosome RD - GC distributions.&#34; % (bin_size_auto, n_bins_auto))
                dist_p_auto = np.zeros(n_bins_auto)
                dist_u_auto = np.zeros(n_bins_auto)
                dist_p_gc_auto = np.zeros((n_bins_auto, 101))

            if sex:
                max_bin_sex = int(
                    max(map(lambda x: 5 * x[1] + 5 * x[2], filter(lambda x: Genome.is_sex_chrom(x[0]), his_stat))))
                _logger.debug(&#34;Max RD for sex chromosome histograms calculated: %d.&#34; % max_bin_sex)
                bin_size_sex = max_bin_sex // 1000
                if bin_size_sex == 0:
                    bin_size_sex = 1
                bins_sex = range(0, max_bin_sex // bin_size_sex * bin_size_sex + bin_size_sex, bin_size_sex)
                n_bins_sex = len(bins_sex) - 1
                _logger.debug(
                    &#34;Using %d RD bin size, %d bins for sex chromosome RD - GC distributions.&#34; % (
                        bin_size_sex, n_bins_sex))
                dist_p_sex = np.zeros(n_bins_sex)
                dist_u_sex = np.zeros(n_bins_sex)
                dist_p_gc_sex = np.zeros((n_bins_sex, 101))

            if mt:
                max_bin_mt = int(
                    max(map(lambda x: 5 * x[1] + 5 * x[2], filter(lambda x: Genome.is_mt_chrom(x[0]), his_stat))))
                _logger.debug(&#34;Max RD for mitochondria histogram calculated: %d.&#34; % max_bin_mt)
                bin_size_mt = max_bin_mt // 1000
                if bin_size_mt == 0:
                    bin_size_mt = 1
                bins_mt = range(0, max_bin_mt // bin_size_mt * bin_size_mt + bin_size_mt, bin_size_mt)
                n_bins_mt = len(bins_mt) - 1
                _logger.debug(&#34;Using %d bin size, %d bins for mitochondria chromosome.&#34; % (bin_size_mt, n_bins_mt))
                dist_p_mt = np.zeros(n_bins_mt)
                dist_u_mt = np.zeros(n_bins_mt)
                dist_p_gc_mt = np.zeros((n_bins_mt, 101))

            _logger.info(&#34;Calculating global statistics.&#34;)
            for c in self.io.rd_chromosomes():
                if c in rd_gc_chromosomes:
                    _logger.debug(&#34;Chromosome &#39;%s&#39;.&#34; % c)
                    his_p = self.io.get_signal(c, bin_size, &#34;RD&#34;)
                    his_u = self.io.get_signal(c, bin_size, &#34;RD unique&#34;)
                    if Genome.is_autosome(c):
                        dist_p, bins = np.histogram(his_p, bins=bins_auto)
                        dist_u, bins = np.histogram(his_u, bins=bins_auto)
                        gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                        dist_p_gc, xbins, ybins = np.histogram2d(his_p, gcp_decompress(gcat, bin_ratio),
                                                                 bins=(bins_auto, range(102)))
                        dist_p_auto += dist_p
                        dist_u_auto += dist_u
                        dist_p_gc_auto += dist_p_gc
                    elif Genome.is_sex_chrom(c):
                        dist_p, bins = np.histogram(his_p, bins=bins_sex)
                        dist_u, bins = np.histogram(his_u, bins=bins_sex)
                        gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                        dist_p_gc, xbins, ybins = np.histogram2d(his_p, gcp_decompress(gcat, bin_ratio),
                                                                 bins=(bins_sex, range(102)))
                        dist_p_sex += dist_p
                        dist_u_sex += dist_u
                        dist_p_gc_sex += dist_p_gc
                    elif Genome.is_mt_chrom(c) and (bin_size &lt;= 1000):
                        dist_p, bins = np.histogram(his_p, bins=bins_mt)
                        dist_u, bins = np.histogram(his_u, bins=bins_mt)
                        gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                        dist_p_gc, xbins, ybins = np.histogram2d(his_p, gcp_decompress(gcat, bin_ratio),
                                                                 bins=(bins_mt, range(102)))
                        dist_p_mt += dist_p
                        dist_u_mt += dist_u
                        dist_p_gc_mt += dist_p_gc

            if auto:
                n_auto, m_auto, s_auto = fit_normal(np.array(bins_auto[1:-1]), dist_p_auto[1:])[0]
                _logger.info(&#34;Autosomes stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (m_auto, s_auto))
                stat_auto = np.array([max_bin_auto, bin_size_auto, n_bins_auto, n_auto, m_auto, s_auto])
                self.io.create_signal(None, bin_size, &#34;RD p dist&#34;, dist_p_auto, flags=FLAG_AUTO)
                self.io.create_signal(None, bin_size, &#34;RD u dist&#34;, dist_u_auto, flags=FLAG_AUTO)
                self.io.create_signal(None, bin_size, &#34;RD GC dist&#34;, dist_p_gc_auto, flags=FLAG_AUTO)
                self.io.create_signal(None, bin_size, &#34;RD stat&#34;, stat_auto, flags=FLAG_AUTO)
                gc_corr_auto = calculate_gc_correction(dist_p_gc_auto, m_auto, s_auto, bin_size_auto)
                self.io.create_signal(None, bin_size, &#34;GC corr&#34;, gc_corr_auto, flags=FLAG_AUTO)

            if sex:
                n_sex, m_sex, s_sex = fit_normal(np.array(bins_sex[1:-1]), dist_p_sex[1:])[0]
                _logger.info(
                    &#34;Sex chromosomes stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (m_sex, s_sex))
                stat_sex = np.array([max_bin_sex, bin_size_sex, n_bins_sex, n_sex, m_sex, s_sex])
                self.io.create_signal(None, bin_size, &#34;RD p dist&#34;, dist_p_sex, flags=FLAG_SEX)
                self.io.create_signal(None, bin_size, &#34;RD u dist&#34;, dist_u_sex, flags=FLAG_SEX)
                self.io.create_signal(None, bin_size, &#34;RD GC dist&#34;, dist_p_gc_sex, flags=FLAG_SEX)
                self.io.create_signal(None, bin_size, &#34;RD stat&#34;, stat_sex, flags=FLAG_SEX)
                gc_corr_sex = calculate_gc_correction(dist_p_gc_sex, m_sex, s_sex, bin_size_sex)
                self.io.create_signal(None, bin_size, &#34;GC corr&#34;, gc_corr_sex, flags=FLAG_SEX)

            if mt:
                n_mt, m_mt, s_mt = fit_normal(np.array(bins_mt[1:-1]), dist_p_mt[1:])[0]
                _logger.info(&#34;Mitochondria stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (m_mt, s_mt))
                if auto:
                    _logger.info(&#34;Mitochondria stat - number of mitochondria per cell:  %.2f +- %.2f&#34; % (
                        2. * m_mt / m_auto, 2. * s_mt / m_auto + s_auto * m_mt / (m_auto * m_auto)))
                stat_mt = np.array([max_bin_mt, bin_size_mt, n_bins_mt, n_mt, m_mt, s_mt])
                self.io.create_signal(None, bin_size, &#34;RD p dist&#34;, dist_p_mt, flags=FLAG_MT)
                self.io.create_signal(None, bin_size, &#34;RD u dist&#34;, dist_u_mt, flags=FLAG_MT)
                self.io.create_signal(None, bin_size, &#34;RD GC dist&#34;, dist_p_gc_mt, flags=FLAG_MT)
                self.io.create_signal(None, bin_size, &#34;RD stat&#34;, stat_mt, flags=FLAG_MT)
                gc_corr_mt = calculate_gc_correction(dist_p_gc_mt, m_mt, s_mt, bin_size_mt)
                self.io.create_signal(None, bin_size, &#34;GC corr&#34;, gc_corr_mt, flags=FLAG_MT)

            for c in self.io.rd_chromosomes():
                if c in rd_gc_chromosomes:
                    _logger.info(
                        &#34;Calculating GC corrected RD histogram using bin size %d for chromosome &#39;%s&#39;.&#34; % (bin_size, c))
                    flag = FLAG_MT if Genome.is_mt_chrom(c) else FLAG_SEX if Genome.is_sex_chrom(c) else FLAG_AUTO
                    his_p = self.io.get_signal(c, bin_size, &#34;RD&#34;)
                    _logger.debug(&#34;Calculating GC corrected RD&#34;)
                    gc_corr = self.io.get_signal(None, bin_size, &#34;GC corr&#34;, flag)
                    gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                    his_p_corr = his_p / np.array(list(map(lambda x: gc_corr[int(x)], gcp_decompress(gcat, bin_ratio))))

                    self.io.create_signal(c, bin_size, &#34;RD&#34;, his_p_corr, flags=FLAG_GC_CORR)

            for c in self.io.rd_chromosomes():
                if (c in rd_gc_chromosomes) and (c in rd_mask_chromosomes):
                    _logger.info(&#34;Calculating P-mask histograms using bin size %d for chromosome &#39;%s&#39;.&#34; % (bin_size, c))
                    flag = FLAG_MT if Genome.is_mt_chrom(c) else FLAG_SEX if Genome.is_sex_chrom(c) else FLAG_AUTO
                    stat = self.io.get_signal(None, 100, &#34;RD stat&#34;, flag)
                    rd_p, rd_u = self.io.read_rd(c)
                    mask = mask_decompress(self.io_mask.get_signal(rd_mask_chromosomes[c], None, &#34;mask&#34;))
                    bin_ratio = bin_size // 100
                    n_bins = len(rd_p) // bin_ratio + 1
                    his_p = np.zeros(n_bins)
                    his_p_corr = np.zeros(n_bins)
                    his_u = np.zeros(n_bins)
                    his_count = np.zeros(n_bins)
                    for (b, e) in mask:
                        for p in range((b - 1) // 100 + 2, (e - 1) // 100):
                            his_p[p // bin_ratio] += rd_p[p]
                            his_u[p // bin_ratio] += rd_u[p]
                            his_count[p // bin_ratio] += 1
                    np.seterr(divide=&#39;ignore&#39;, invalid=&#39;ignore&#39;)
                    his_p = his_p / his_count * bin_ratio
                    his_u = his_u / his_count * bin_ratio
                    gc_corr = self.io.get_signal(None, bin_size, &#34;GC corr&#34;, flag)
                    gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                    his_p_corr = his_p / np.array(list(map(lambda x: gc_corr[int(x)], gcp_decompress(gcat, bin_ratio))))
                    self.io.create_signal(c, bin_size, &#34;RD&#34;, his_p, flags=FLAG_USEMASK)
                    self.io.create_signal(c, bin_size, &#34;RD unique&#34;, his_u, flags=FLAG_USEMASK)
                    self.io.create_signal(c, bin_size, &#34;RD&#34;, his_p_corr, flags=FLAG_GC_CORR | FLAG_USEMASK)

    def mask_snps(self):
        &#34;&#34;&#34;
        Flags SNPs in P-region (sets second bit of the flag to 1 for SNP inside P region, or to 0 otherwise).
        Requires imported mask data or recognized reference genome with mask data.

        &#34;&#34;&#34;
        snp_mask_chromosomes = {}
        for c in self.io_mask.mask_chromosomes():
            snp_name = self.io.snp_chromosome_name(c)
            if snp_name is not None:
                snp_mask_chromosomes[snp_name] = c

        for c in self.io.snp_chromosomes():
            if c in snp_mask_chromosomes:
                _logger.info(&#34;Masking SNP data for chromosome &#39;%s&#39;.&#34; % c)
                pos, ref, alt, nref, nalt, gt, flag, qual = self.io.read_snp(c)
                mask = mask_decompress(self.io_mask.get_signal(snp_mask_chromosomes[c], None, &#34;mask&#34;))
                mask_ix = 0
                for snp_ix in range(len(pos)):
                    while mask_ix &lt; len(mask) and mask[mask_ix][1] &lt; pos[snp_ix]:
                        mask_ix += 1
                    if mask_ix &lt; len(mask) and mask[mask_ix][0] &lt; pos[snp_ix] and pos[snp_ix] &lt; (mask[mask_ix][1] + 1):
                        flag[snp_ix] = flag[snp_ix] | 2
                    else:
                        flag[snp_ix] = flag[snp_ix] &amp; 1
                self.io.save_snp(c, pos, ref, alt, nref, nalt, gt, flag, qual, update=True)

    def calculate_baf(self, bin_sizes, chroms=[], use_mask=True, use_id=False, res=200, reduce_noise=False):
        &#34;&#34;&#34;
        Calculates BAF histograms and store data into cnvpytor file.

        Parameters
        ----------
        bin_sizes : list of int
            List of histogram bin sizes
        chroms : list of str
            List of chromosomes. Calculates for all available if empty.
        use_mask : bool
            Use P-mask filter if True. Default: True.
        use_id : bool
            Use id flag filter if True. Default: False.
        res: int
            Likelihood function resolution. Default: 200.


        &#34;&#34;&#34;
        snp_flag = (FLAG_USEMASK if use_mask else 0) | (FLAG_USEID if use_id else 0)
        for c in self.io.snp_chromosomes():
            if len(chroms) == 0 or c in chroms:
                _logger.info(&#34;Calculating BAF histograms for chromosome &#39;%s&#39;.&#34; % c)
                pos, ref, alt, nref, nalt, gt, flag, qual = self.io.read_snp(c)
                max_bin = {}
                count = {}
                hets = {}
                count_h1 = {}
                count_h2 = {}
                baf = {}
                maf = {}
                likelihood = {}
                i1 = {}
                i2 = {}
                lh_x = np.arange(1.0 / res, 1., 1.0 / res)
                for bs in bin_sizes:
                    max_bin[bs] = (pos[-1] - 1) // bs + 1
                    count[bs] = np.zeros(max_bin[bs])
                    hets[bs] = np.zeros(max_bin[bs])
                    count_h1[bs] = np.zeros(max_bin[bs])
                    count_h2[bs] = np.zeros(max_bin[bs])
                    baf[bs] = np.zeros(max_bin[bs])
                    maf[bs] = np.zeros(max_bin[bs])
                    likelihood[bs] = np.ones((max_bin[bs], res - 1)).astype(&#34;float&#34;) / (res - 1)
                    i1[bs] = np.zeros(max_bin[bs])
                    i2[bs] = np.zeros(max_bin[bs])

                for i in range(len(pos)):
                    if (gt[i] == 1 or gt[i] == 5 or gt[i] == 6) and (nalt[i] + nref[i]) &gt; 0 and (
                            not use_id or (flag[i] &amp; 1)) and (not use_mask or (flag[i] &amp; 2)):
                        for bs in bin_sizes:
                            b = (pos[i] - 1) // bs
                            count[bs][b] += 1
                            if gt[i] == 5:
                                count_h1[bs][b] += nref[i]
                                count_h2[bs][b] += nalt[i]
                                hets[bs][b] += 1
                                snp_baf = 1.0 * nalt[i] / (nalt[i] + nref[i])
                                likelihood[bs][b] *= beta(nalt[i], nref[i], lh_x, phased=True)
                                s = np.sum(likelihood[bs][b])
                                if s != 0.0:
                                    likelihood[bs][b] /= s
                            elif gt[i] == 6:
                                count_h1[bs][b] += nalt[i]
                                count_h2[bs][b] += nref[i]
                                hets[bs][b] += 1
                                snp_baf = 1.0 * nref[i] / (nalt[i] + nref[i])
                                likelihood[bs][b] *= beta(nref[i], nalt[i], lh_x, phased=True)
                                s = np.sum(likelihood[bs][b])
                                if s != 0.0:
                                    likelihood[bs][b] /= s
                            else:
                                snp_baf = 1.0 * nalt[i] / (nalt[i] + nref[i])
                                hets[bs][b] += 1
                                if reduce_noise:
                                    likelihood[bs][b] *= beta(nalt[i] + (1 if nalt[i] &lt; nref[i] else 0),
                                                              nref[i] + (1 if nref[i] &lt; nalt[i] else 0), lh_x)
                                else:
                                    likelihood[bs][b] *= beta(nalt[i] + 1, nref[i] + 1, lh_x)
                                s = np.sum(likelihood[bs][b])
                                if s != 0.0:
                                    likelihood[bs][b] /= s

                            baf[bs][b] += snp_baf
                            maf[bs][b] += 1.0 - snp_baf if snp_baf &gt; 0.5 else snp_baf

                for bs in bin_sizes:
                    for i in range(max_bin[bs]):
                        if count[bs][i] &gt; 0:
                            baf[bs][i] /= count[bs][i]
                            maf[bs][i] /= count[bs][i]
                            max_lh = np.amax(likelihood[bs][i])
                            ix = np.where(likelihood[bs][i] == max_lh)[0][0]
                            i1[bs][i] = 1.0 * (res // 2 - 1 - ix) / res if ix &lt;= (res // 2 - 1) else 1.0 * (
                                    ix - res // 2 + 1) / res
                            i2[bs][i] = likelihood[bs][i][res // 2 - 1] / max_lh
                    _logger.info(&#34;Saving BAF histograms with bin size %d for chromosome &#39;%s&#39;.&#34; % (bs, c))
                    self.io.create_signal(c, bs, &#34;SNP bin count&#34;, count[bs].astype(&#34;uint16&#34;), snp_flag)
                    self.io.create_signal(c, bs, &#34;SNP bin hets&#34;, hets[bs].astype(&#34;uint16&#34;), snp_flag)
                    self.io.create_signal(c, bs, &#34;SNP bin count h1&#34;, count_h1[bs].astype(&#34;uint16&#34;), snp_flag)
                    self.io.create_signal(c, bs, &#34;SNP bin count h2&#34;, count_h2[bs].astype(&#34;uint16&#34;), snp_flag)
                    self.io.create_signal(c, bs, &#34;SNP baf&#34;, baf[bs].astype(&#34;float32&#34;), snp_flag)
                    self.io.create_signal(c, bs, &#34;SNP maf&#34;, maf[bs].astype(&#34;float32&#34;), snp_flag)
                    self.io.create_signal(c, bs, &#34;SNP likelihood&#34;, likelihood[bs].astype(&#34;float32&#34;), snp_flag)
                    self.io.create_signal(c, bs, &#34;SNP i1&#34;, i1[bs].astype(&#34;float32&#34;), snp_flag)
                    self.io.create_signal(c, bs, &#34;SNP i2&#34;, i2[bs].astype(&#34;float32&#34;), snp_flag)

    def call_baf(self, bin_sizes, chroms=[], use_mask=True, use_id=False, odec=0.9, omin=None, mcount=None,
                 max_distance=0.1):
        &#34;&#34;&#34;

        Returns
        -------

        &#34;&#34;&#34;
        for bin_size in bin_sizes:
            if omin is None:
                overlap_min = 1e-7 * bin_size
            else:
                overlap_min = omin
            if mcount is None:
                min_count = bin_size // 10000
            else:
                min_count = mcount

            snp_flag = (FLAG_USEMASK if use_mask else 0) | (FLAG_USEID if use_id else 0)
            for c in self.io.snp_chromosomes():
                if len(chroms) == 0 or c in chroms and self.io.signal_exists(c, bin_size, &#34;SNP likelihood&#34;, snp_flag):
                    # _logger.info(
                    #    &#34;Caling CNV-s by merging BAF likelihood with bin size %d for chromosome &#39;%s&#39;.&#34; % (bin_size, c))
                    likelihood = list(self.io.get_signal(c, bin_size, &#34;SNP likelihood&#34;, snp_flag).astype(&#34;float64&#34;))
                    snp_hets = self.io.get_signal(c, bin_size, &#34;SNP bin hets&#34;, snp_flag)

                    bins = len(likelihood)
                    res = likelihood[0].size
                    segments = [[i] for i in range(bins) if
                                snp_hets[i] &gt;= min_count and np.sum(likelihood[i]) &gt; 0.0]
                    likelihood = [likelihood[i] for i in range(bins) if
                                  snp_hets[i] &gt;= min_count and np.sum(likelihood[i]) &gt; 0.0]
                    overlaps = [likelihood_overlap(likelihood[i], likelihood[i + 1]) for i in range(len(segments) - 1)]

                    iter = 0
                    while len(overlaps) &gt; 0:
                        maxo = max(overlaps)
                        mino = max(maxo * odec, overlap_min)
                        if maxo &lt; overlap_min:
                            break
                        i = 0
                        while i &lt; len(overlaps):
                            if overlaps[i] &gt; mino:
                                nlh = likelihood[i] * likelihood[i + 1]
                                likelihood[i] = nlh / np.sum(nlh)
                                segments[i] += segments[i + 1]
                                del likelihood[i + 1]
                                del segments[i + 1]
                                del overlaps[i]
                                if i &lt; len(overlaps):
                                    overlaps[i] = likelihood_overlap(likelihood[i], likelihood[i + 1])
                                if i &gt; 0:
                                    overlaps[i - 1] = likelihood_overlap(likelihood[i - 1], likelihood[i])
                            else:
                                i = i + 1
                        iter = iter + 1

                    overlaps = [[
                        likelihood_overlap(likelihood[i], likelihood[j])
                        if (segments[j][0] - segments[i][-1]) &lt; max_distance * (
                                len(segments[i]) + len(segments[j])) and i &lt; j
                        else 0 for j in range(len(segments))]
                        for i in range(len(segments))]

                    iter = 0
                    ons = -1

                    while True:
                        overlaps = [likelihood_overlap(likelihood[i], likelihood[j]) for i in range(len(likelihood))
                                    for j in range(i + 1, len(likelihood)) if
                                    (segments[j][0] - segments[i][-1]) &lt; max_distance * (
                                                len(segments[i]) + len(segments[j]))]
                        if len(overlaps) == 0:
                            break
                        maxo = max(overlaps)
                        mino = max(maxo * odec, overlap_min)
                        if maxo &lt; omin:
                            break
                        i, j = 0, 1
                        while i &lt; len(segments) - 1:
                            #                            if overlaps[i][j] &gt; mino:
                            if likelihood_overlap(likelihood[i], likelihood[j]) &gt; mino and (
                                    segments[j][0] - segments[i][-1]) &lt; max_distance * (
                                    len(segments[i]) + len(segments[j])):
                                nlh = likelihood[i] * likelihood[j]
                                likelihood[i] = nlh / np.sum(nlh)
                                segments[i] += segments[j]
                                segments[i] = sorted(segments[i])
                                del likelihood[j]
                                del segments[j]

                                if j &gt;= len(segments):
                                    i += 1
                                    j = i + 1
                            else:
                                j += 1
                                if j &gt;= len(segments):
                                    i += 1
                                    j = i + 1

                        iter = iter + 1
                        if ons == len(segments):
                            break
                        ons = len(segments)

                    for i in range(len(segments)):
                        max_lh = np.amax(likelihood[i])
                        ix = np.where(likelihood[i] == max_lh)[0][0]
                        i1 = 1.0 * (res // 2 - ix) / (res + 1) if ix &lt;= (res // 2) else 1.0 * (
                                ix - res // 2) / (res + 1)
                        i2 = likelihood[i][res // 2] / max_lh

                        print(c + &#34;:&#34; + str(segments[i][0] * bin_size + 1) + &#34;-&#34; + str(
                            segments[i][-1] * bin_size + bin_size),
                              bin_size, i1,
                              i2, len(segments[i]))

    def ls(self):
        self.io.ls()</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="cnvpytor.root.Root"><code class="flex name class">
<span>class <span class="ident">Root</span></span>
<span>(</span><span>filename, max_cores=8)</span>
</code></dt>
<dd>
<section class="desc"><p>Class constructor opens CNVpytor file. The class implements all core CNVpytor calculations.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>filename</code></strong> :&ensp;<code>str</code></dt>
<dd>CNVpytor filename</dd>
<dt><strong><code>max_cores</code></strong> :&ensp;<code>int</code></dt>
<dd>Maximal number of cores used in parallel calculations</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">class Root: main CNVpytor class</code></pre>
</details>
<h3>Methods</h3>
<dl>
<dt id="cnvpytor.root.Root.calculate_baf"><code class="name flex">
<span>def <span class="ident">calculate_baf</span></span>(<span>self, bin_sizes, chroms=[], use_mask=True, use_id=False, res=200, reduce_noise=False)</span>
</code></dt>
<dd>
<section class="desc"><p>Calculates BAF histograms and store data into cnvpytor file.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>bin_sizes</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>List of histogram bin sizes</dd>
<dt><strong><code>chroms</code></strong> :&ensp;<code>list</code> of <code>str</code></dt>
<dd>List of chromosomes. Calculates for all available if empty.</dd>
<dt><strong><code>use_mask</code></strong> :&ensp;<code>bool</code></dt>
<dd>Use P-mask filter if True. Default: True.</dd>
<dt><strong><code>use_id</code></strong> :&ensp;<code>bool</code></dt>
<dd>Use id flag filter if True. Default: False.</dd>
<dt><strong><code>res</code></strong> :&ensp;<code>int</code></dt>
<dd>Likelihood function resolution. Default: 200.</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def calculate_baf(self, bin_sizes, chroms=[], use_mask=True, use_id=False, res=200, reduce_noise=False):
    &#34;&#34;&#34;
    Calculates BAF histograms and store data into cnvpytor file.

    Parameters
    ----------
    bin_sizes : list of int
        List of histogram bin sizes
    chroms : list of str
        List of chromosomes. Calculates for all available if empty.
    use_mask : bool
        Use P-mask filter if True. Default: True.
    use_id : bool
        Use id flag filter if True. Default: False.
    res: int
        Likelihood function resolution. Default: 200.


    &#34;&#34;&#34;
    snp_flag = (FLAG_USEMASK if use_mask else 0) | (FLAG_USEID if use_id else 0)
    for c in self.io.snp_chromosomes():
        if len(chroms) == 0 or c in chroms:
            _logger.info(&#34;Calculating BAF histograms for chromosome &#39;%s&#39;.&#34; % c)
            pos, ref, alt, nref, nalt, gt, flag, qual = self.io.read_snp(c)
            max_bin = {}
            count = {}
            hets = {}
            count_h1 = {}
            count_h2 = {}
            baf = {}
            maf = {}
            likelihood = {}
            i1 = {}
            i2 = {}
            lh_x = np.arange(1.0 / res, 1., 1.0 / res)
            for bs in bin_sizes:
                max_bin[bs] = (pos[-1] - 1) // bs + 1
                count[bs] = np.zeros(max_bin[bs])
                hets[bs] = np.zeros(max_bin[bs])
                count_h1[bs] = np.zeros(max_bin[bs])
                count_h2[bs] = np.zeros(max_bin[bs])
                baf[bs] = np.zeros(max_bin[bs])
                maf[bs] = np.zeros(max_bin[bs])
                likelihood[bs] = np.ones((max_bin[bs], res - 1)).astype(&#34;float&#34;) / (res - 1)
                i1[bs] = np.zeros(max_bin[bs])
                i2[bs] = np.zeros(max_bin[bs])

            for i in range(len(pos)):
                if (gt[i] == 1 or gt[i] == 5 or gt[i] == 6) and (nalt[i] + nref[i]) &gt; 0 and (
                        not use_id or (flag[i] &amp; 1)) and (not use_mask or (flag[i] &amp; 2)):
                    for bs in bin_sizes:
                        b = (pos[i] - 1) // bs
                        count[bs][b] += 1
                        if gt[i] == 5:
                            count_h1[bs][b] += nref[i]
                            count_h2[bs][b] += nalt[i]
                            hets[bs][b] += 1
                            snp_baf = 1.0 * nalt[i] / (nalt[i] + nref[i])
                            likelihood[bs][b] *= beta(nalt[i], nref[i], lh_x, phased=True)
                            s = np.sum(likelihood[bs][b])
                            if s != 0.0:
                                likelihood[bs][b] /= s
                        elif gt[i] == 6:
                            count_h1[bs][b] += nalt[i]
                            count_h2[bs][b] += nref[i]
                            hets[bs][b] += 1
                            snp_baf = 1.0 * nref[i] / (nalt[i] + nref[i])
                            likelihood[bs][b] *= beta(nref[i], nalt[i], lh_x, phased=True)
                            s = np.sum(likelihood[bs][b])
                            if s != 0.0:
                                likelihood[bs][b] /= s
                        else:
                            snp_baf = 1.0 * nalt[i] / (nalt[i] + nref[i])
                            hets[bs][b] += 1
                            if reduce_noise:
                                likelihood[bs][b] *= beta(nalt[i] + (1 if nalt[i] &lt; nref[i] else 0),
                                                          nref[i] + (1 if nref[i] &lt; nalt[i] else 0), lh_x)
                            else:
                                likelihood[bs][b] *= beta(nalt[i] + 1, nref[i] + 1, lh_x)
                            s = np.sum(likelihood[bs][b])
                            if s != 0.0:
                                likelihood[bs][b] /= s

                        baf[bs][b] += snp_baf
                        maf[bs][b] += 1.0 - snp_baf if snp_baf &gt; 0.5 else snp_baf

            for bs in bin_sizes:
                for i in range(max_bin[bs]):
                    if count[bs][i] &gt; 0:
                        baf[bs][i] /= count[bs][i]
                        maf[bs][i] /= count[bs][i]
                        max_lh = np.amax(likelihood[bs][i])
                        ix = np.where(likelihood[bs][i] == max_lh)[0][0]
                        i1[bs][i] = 1.0 * (res // 2 - 1 - ix) / res if ix &lt;= (res // 2 - 1) else 1.0 * (
                                ix - res // 2 + 1) / res
                        i2[bs][i] = likelihood[bs][i][res // 2 - 1] / max_lh
                _logger.info(&#34;Saving BAF histograms with bin size %d for chromosome &#39;%s&#39;.&#34; % (bs, c))
                self.io.create_signal(c, bs, &#34;SNP bin count&#34;, count[bs].astype(&#34;uint16&#34;), snp_flag)
                self.io.create_signal(c, bs, &#34;SNP bin hets&#34;, hets[bs].astype(&#34;uint16&#34;), snp_flag)
                self.io.create_signal(c, bs, &#34;SNP bin count h1&#34;, count_h1[bs].astype(&#34;uint16&#34;), snp_flag)
                self.io.create_signal(c, bs, &#34;SNP bin count h2&#34;, count_h2[bs].astype(&#34;uint16&#34;), snp_flag)
                self.io.create_signal(c, bs, &#34;SNP baf&#34;, baf[bs].astype(&#34;float32&#34;), snp_flag)
                self.io.create_signal(c, bs, &#34;SNP maf&#34;, maf[bs].astype(&#34;float32&#34;), snp_flag)
                self.io.create_signal(c, bs, &#34;SNP likelihood&#34;, likelihood[bs].astype(&#34;float32&#34;), snp_flag)
                self.io.create_signal(c, bs, &#34;SNP i1&#34;, i1[bs].astype(&#34;float32&#34;), snp_flag)
                self.io.create_signal(c, bs, &#34;SNP i2&#34;, i2[bs].astype(&#34;float32&#34;), snp_flag)</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.calculate_histograms"><code class="name flex">
<span>def <span class="ident">calculate_histograms</span></span>(<span>self, bin_sizes, chroms=[])</span>
</code></dt>
<dd>
<section class="desc"><p>Calculates RD histograms and store data into cnvpytor file.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>bin_sizes</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>List of histogram bin sizes</dd>
<dt><strong><code>chroms</code></strong> :&ensp;<code>list</code> of <code>str</code></dt>
<dd>List of chromosomes. Calculates for all available if empty.</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def calculate_histograms(self, bin_sizes, chroms=[]):
    &#34;&#34;&#34;
    Calculates RD histograms and store data into cnvpytor file.

    Parameters
    ----------
    bin_sizes : list of int
        List of histogram bin sizes
    chroms : list of str
        List of chromosomes. Calculates for all available if empty.

    &#34;&#34;&#34;
    rd_gc_chromosomes = {}
    for c in self.io_gc.gc_chromosomes():
        rd_name = self.io.rd_chromosome_name(c)
        if not rd_name is None and (len(chroms) == 0 or (rd_name in chroms) or (c in chroms)):
            rd_gc_chromosomes[rd_name] = c

    rd_mask_chromosomes = {}
    for c in self.io_mask.mask_chromosomes():
        rd_name = self.io.rd_chromosome_name(c)
        if not rd_name is None and (len(chroms) == 0 or (rd_name in chroms) or (c in chroms)):
            rd_mask_chromosomes[rd_name] = c

    for bin_size in bin_sizes:
        his_stat = []
        auto, sex, mt = False, False, False
        bin_ratio = bin_size // 100

        for c in self.io.rd_chromosomes():
            if c in rd_gc_chromosomes:
                _logger.info(&#34;Calculating histograms using bin size %d for chromosome &#39;%s&#39;.&#34; % (bin_size, c))
                flag = FLAG_MT if Genome.is_mt_chrom(c) else FLAG_SEX if Genome.is_sex_chrom(c) else FLAG_AUTO
                rd_p, rd_u = self.io.read_rd(c)
                his_p = np.concatenate(
                    (rd_p, np.zeros(bin_ratio - len(rd_p) + (len(rd_p) // bin_ratio * bin_ratio))))
                his_p.resize((len(his_p) // bin_ratio, bin_ratio))
                his_p = his_p.sum(axis=1)
                his_u = np.concatenate(
                    (rd_u, np.zeros(bin_ratio - len(rd_u) + (len(rd_u) // bin_ratio * bin_ratio))))
                his_u.resize((len(his_u) // bin_ratio, bin_ratio))
                his_u = his_u.sum(axis=1)
                self.io.create_signal(c, bin_size, &#34;RD&#34;, his_p)
                self.io.create_signal(c, bin_size, &#34;RD unique&#34;, his_u)

                max_bin = max(int(10 * np.mean(his_u) + 1), int(10 * np.mean(his_p) + 1))
                rd_bin_size = max_bin // 10000
                if rd_bin_size == 0:
                    rd_bin_size = 1
                rd_bins = range(0, max_bin // rd_bin_size * rd_bin_size + rd_bin_size, rd_bin_size)
                dist_p, bins = np.histogram(his_p, bins=rd_bins)
                dist_u, bins = np.histogram(his_u, bins=rd_bins)
                n_p, m_p, s_p = fit_normal(bins[1:-1], dist_p[1:])[0]
                n_u, m_u, s_u = fit_normal(bins[1:-1], dist_u[1:])[0]
                _logger.info(
                    &#34;Chromosome &#39;%s&#39; bin size %d stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (
                        c, bin_size, m_p, s_p))
                _logger.info(
                    &#34;Chromosome &#39;%s&#39; bin size %d stat - RD unique distribution gaussian fit:  %.2f +- %.2f&#34; % (
                        c, bin_size, m_u, s_u))
                his_stat.append((c, m_p, s_p, m_u, s_u))
                auto = auto or Genome.is_autosome(c)
                sex = sex or Genome.is_sex_chrom(c)
                mt = mt or Genome.is_mt_chrom(c)

        if auto:
            max_bin_auto = int(
                max(map(lambda x: 5 * x[1] + 5 * x[2], filter(lambda x: Genome.is_autosome(x[0]), his_stat))))
            _logger.debug(&#34;Max RD for autosome histograms calculated: %d.&#34; % max_bin_auto)
            bin_size_auto = max_bin_auto // 1000
            if bin_size_auto == 0:
                bin_size_auto = 1
            bins_auto = range(0, max_bin_auto // bin_size_auto * bin_size_auto + bin_size_auto, bin_size_auto)
            n_bins_auto = len(bins_auto) - 1
            _logger.debug(
                &#34;Using %d RD bin size, %d bins for autosome RD - GC distributions.&#34; % (bin_size_auto, n_bins_auto))
            dist_p_auto = np.zeros(n_bins_auto)
            dist_u_auto = np.zeros(n_bins_auto)
            dist_p_gc_auto = np.zeros((n_bins_auto, 101))

        if sex:
            max_bin_sex = int(
                max(map(lambda x: 5 * x[1] + 5 * x[2], filter(lambda x: Genome.is_sex_chrom(x[0]), his_stat))))
            _logger.debug(&#34;Max RD for sex chromosome histograms calculated: %d.&#34; % max_bin_sex)
            bin_size_sex = max_bin_sex // 1000
            if bin_size_sex == 0:
                bin_size_sex = 1
            bins_sex = range(0, max_bin_sex // bin_size_sex * bin_size_sex + bin_size_sex, bin_size_sex)
            n_bins_sex = len(bins_sex) - 1
            _logger.debug(
                &#34;Using %d RD bin size, %d bins for sex chromosome RD - GC distributions.&#34; % (
                    bin_size_sex, n_bins_sex))
            dist_p_sex = np.zeros(n_bins_sex)
            dist_u_sex = np.zeros(n_bins_sex)
            dist_p_gc_sex = np.zeros((n_bins_sex, 101))

        if mt:
            max_bin_mt = int(
                max(map(lambda x: 5 * x[1] + 5 * x[2], filter(lambda x: Genome.is_mt_chrom(x[0]), his_stat))))
            _logger.debug(&#34;Max RD for mitochondria histogram calculated: %d.&#34; % max_bin_mt)
            bin_size_mt = max_bin_mt // 1000
            if bin_size_mt == 0:
                bin_size_mt = 1
            bins_mt = range(0, max_bin_mt // bin_size_mt * bin_size_mt + bin_size_mt, bin_size_mt)
            n_bins_mt = len(bins_mt) - 1
            _logger.debug(&#34;Using %d bin size, %d bins for mitochondria chromosome.&#34; % (bin_size_mt, n_bins_mt))
            dist_p_mt = np.zeros(n_bins_mt)
            dist_u_mt = np.zeros(n_bins_mt)
            dist_p_gc_mt = np.zeros((n_bins_mt, 101))

        _logger.info(&#34;Calculating global statistics.&#34;)
        for c in self.io.rd_chromosomes():
            if c in rd_gc_chromosomes:
                _logger.debug(&#34;Chromosome &#39;%s&#39;.&#34; % c)
                his_p = self.io.get_signal(c, bin_size, &#34;RD&#34;)
                his_u = self.io.get_signal(c, bin_size, &#34;RD unique&#34;)
                if Genome.is_autosome(c):
                    dist_p, bins = np.histogram(his_p, bins=bins_auto)
                    dist_u, bins = np.histogram(his_u, bins=bins_auto)
                    gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                    dist_p_gc, xbins, ybins = np.histogram2d(his_p, gcp_decompress(gcat, bin_ratio),
                                                             bins=(bins_auto, range(102)))
                    dist_p_auto += dist_p
                    dist_u_auto += dist_u
                    dist_p_gc_auto += dist_p_gc
                elif Genome.is_sex_chrom(c):
                    dist_p, bins = np.histogram(his_p, bins=bins_sex)
                    dist_u, bins = np.histogram(his_u, bins=bins_sex)
                    gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                    dist_p_gc, xbins, ybins = np.histogram2d(his_p, gcp_decompress(gcat, bin_ratio),
                                                             bins=(bins_sex, range(102)))
                    dist_p_sex += dist_p
                    dist_u_sex += dist_u
                    dist_p_gc_sex += dist_p_gc
                elif Genome.is_mt_chrom(c) and (bin_size &lt;= 1000):
                    dist_p, bins = np.histogram(his_p, bins=bins_mt)
                    dist_u, bins = np.histogram(his_u, bins=bins_mt)
                    gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                    dist_p_gc, xbins, ybins = np.histogram2d(his_p, gcp_decompress(gcat, bin_ratio),
                                                             bins=(bins_mt, range(102)))
                    dist_p_mt += dist_p
                    dist_u_mt += dist_u
                    dist_p_gc_mt += dist_p_gc

        if auto:
            n_auto, m_auto, s_auto = fit_normal(np.array(bins_auto[1:-1]), dist_p_auto[1:])[0]
            _logger.info(&#34;Autosomes stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (m_auto, s_auto))
            stat_auto = np.array([max_bin_auto, bin_size_auto, n_bins_auto, n_auto, m_auto, s_auto])
            self.io.create_signal(None, bin_size, &#34;RD p dist&#34;, dist_p_auto, flags=FLAG_AUTO)
            self.io.create_signal(None, bin_size, &#34;RD u dist&#34;, dist_u_auto, flags=FLAG_AUTO)
            self.io.create_signal(None, bin_size, &#34;RD GC dist&#34;, dist_p_gc_auto, flags=FLAG_AUTO)
            self.io.create_signal(None, bin_size, &#34;RD stat&#34;, stat_auto, flags=FLAG_AUTO)
            gc_corr_auto = calculate_gc_correction(dist_p_gc_auto, m_auto, s_auto, bin_size_auto)
            self.io.create_signal(None, bin_size, &#34;GC corr&#34;, gc_corr_auto, flags=FLAG_AUTO)

        if sex:
            n_sex, m_sex, s_sex = fit_normal(np.array(bins_sex[1:-1]), dist_p_sex[1:])[0]
            _logger.info(
                &#34;Sex chromosomes stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (m_sex, s_sex))
            stat_sex = np.array([max_bin_sex, bin_size_sex, n_bins_sex, n_sex, m_sex, s_sex])
            self.io.create_signal(None, bin_size, &#34;RD p dist&#34;, dist_p_sex, flags=FLAG_SEX)
            self.io.create_signal(None, bin_size, &#34;RD u dist&#34;, dist_u_sex, flags=FLAG_SEX)
            self.io.create_signal(None, bin_size, &#34;RD GC dist&#34;, dist_p_gc_sex, flags=FLAG_SEX)
            self.io.create_signal(None, bin_size, &#34;RD stat&#34;, stat_sex, flags=FLAG_SEX)
            gc_corr_sex = calculate_gc_correction(dist_p_gc_sex, m_sex, s_sex, bin_size_sex)
            self.io.create_signal(None, bin_size, &#34;GC corr&#34;, gc_corr_sex, flags=FLAG_SEX)

        if mt:
            n_mt, m_mt, s_mt = fit_normal(np.array(bins_mt[1:-1]), dist_p_mt[1:])[0]
            _logger.info(&#34;Mitochondria stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (m_mt, s_mt))
            if auto:
                _logger.info(&#34;Mitochondria stat - number of mitochondria per cell:  %.2f +- %.2f&#34; % (
                    2. * m_mt / m_auto, 2. * s_mt / m_auto + s_auto * m_mt / (m_auto * m_auto)))
            stat_mt = np.array([max_bin_mt, bin_size_mt, n_bins_mt, n_mt, m_mt, s_mt])
            self.io.create_signal(None, bin_size, &#34;RD p dist&#34;, dist_p_mt, flags=FLAG_MT)
            self.io.create_signal(None, bin_size, &#34;RD u dist&#34;, dist_u_mt, flags=FLAG_MT)
            self.io.create_signal(None, bin_size, &#34;RD GC dist&#34;, dist_p_gc_mt, flags=FLAG_MT)
            self.io.create_signal(None, bin_size, &#34;RD stat&#34;, stat_mt, flags=FLAG_MT)
            gc_corr_mt = calculate_gc_correction(dist_p_gc_mt, m_mt, s_mt, bin_size_mt)
            self.io.create_signal(None, bin_size, &#34;GC corr&#34;, gc_corr_mt, flags=FLAG_MT)

        for c in self.io.rd_chromosomes():
            if c in rd_gc_chromosomes:
                _logger.info(
                    &#34;Calculating GC corrected RD histogram using bin size %d for chromosome &#39;%s&#39;.&#34; % (bin_size, c))
                flag = FLAG_MT if Genome.is_mt_chrom(c) else FLAG_SEX if Genome.is_sex_chrom(c) else FLAG_AUTO
                his_p = self.io.get_signal(c, bin_size, &#34;RD&#34;)
                _logger.debug(&#34;Calculating GC corrected RD&#34;)
                gc_corr = self.io.get_signal(None, bin_size, &#34;GC corr&#34;, flag)
                gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                his_p_corr = his_p / np.array(list(map(lambda x: gc_corr[int(x)], gcp_decompress(gcat, bin_ratio))))

                self.io.create_signal(c, bin_size, &#34;RD&#34;, his_p_corr, flags=FLAG_GC_CORR)

        for c in self.io.rd_chromosomes():
            if (c in rd_gc_chromosomes) and (c in rd_mask_chromosomes):
                _logger.info(&#34;Calculating P-mask histograms using bin size %d for chromosome &#39;%s&#39;.&#34; % (bin_size, c))
                flag = FLAG_MT if Genome.is_mt_chrom(c) else FLAG_SEX if Genome.is_sex_chrom(c) else FLAG_AUTO
                stat = self.io.get_signal(None, 100, &#34;RD stat&#34;, flag)
                rd_p, rd_u = self.io.read_rd(c)
                mask = mask_decompress(self.io_mask.get_signal(rd_mask_chromosomes[c], None, &#34;mask&#34;))
                bin_ratio = bin_size // 100
                n_bins = len(rd_p) // bin_ratio + 1
                his_p = np.zeros(n_bins)
                his_p_corr = np.zeros(n_bins)
                his_u = np.zeros(n_bins)
                his_count = np.zeros(n_bins)
                for (b, e) in mask:
                    for p in range((b - 1) // 100 + 2, (e - 1) // 100):
                        his_p[p // bin_ratio] += rd_p[p]
                        his_u[p // bin_ratio] += rd_u[p]
                        his_count[p // bin_ratio] += 1
                np.seterr(divide=&#39;ignore&#39;, invalid=&#39;ignore&#39;)
                his_p = his_p / his_count * bin_ratio
                his_u = his_u / his_count * bin_ratio
                gc_corr = self.io.get_signal(None, bin_size, &#34;GC corr&#34;, flag)
                gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                his_p_corr = his_p / np.array(list(map(lambda x: gc_corr[int(x)], gcp_decompress(gcat, bin_ratio))))
                self.io.create_signal(c, bin_size, &#34;RD&#34;, his_p, flags=FLAG_USEMASK)
                self.io.create_signal(c, bin_size, &#34;RD unique&#34;, his_u, flags=FLAG_USEMASK)
                self.io.create_signal(c, bin_size, &#34;RD&#34;, his_p_corr, flags=FLAG_GC_CORR | FLAG_USEMASK)</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.call_baf"><code class="name flex">
<span>def <span class="ident">call_baf</span></span>(<span>self, bin_sizes, chroms=[], use_mask=True, use_id=False, odec=0.9, omin=None, mcount=None, max_distance=0.1)</span>
</code></dt>
<dd>
<section class="desc"><h2 id="returns">Returns</h2></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def call_baf(self, bin_sizes, chroms=[], use_mask=True, use_id=False, odec=0.9, omin=None, mcount=None,
             max_distance=0.1):
    &#34;&#34;&#34;

    Returns
    -------

    &#34;&#34;&#34;
    for bin_size in bin_sizes:
        if omin is None:
            overlap_min = 1e-7 * bin_size
        else:
            overlap_min = omin
        if mcount is None:
            min_count = bin_size // 10000
        else:
            min_count = mcount

        snp_flag = (FLAG_USEMASK if use_mask else 0) | (FLAG_USEID if use_id else 0)
        for c in self.io.snp_chromosomes():
            if len(chroms) == 0 or c in chroms and self.io.signal_exists(c, bin_size, &#34;SNP likelihood&#34;, snp_flag):
                # _logger.info(
                #    &#34;Caling CNV-s by merging BAF likelihood with bin size %d for chromosome &#39;%s&#39;.&#34; % (bin_size, c))
                likelihood = list(self.io.get_signal(c, bin_size, &#34;SNP likelihood&#34;, snp_flag).astype(&#34;float64&#34;))
                snp_hets = self.io.get_signal(c, bin_size, &#34;SNP bin hets&#34;, snp_flag)

                bins = len(likelihood)
                res = likelihood[0].size
                segments = [[i] for i in range(bins) if
                            snp_hets[i] &gt;= min_count and np.sum(likelihood[i]) &gt; 0.0]
                likelihood = [likelihood[i] for i in range(bins) if
                              snp_hets[i] &gt;= min_count and np.sum(likelihood[i]) &gt; 0.0]
                overlaps = [likelihood_overlap(likelihood[i], likelihood[i + 1]) for i in range(len(segments) - 1)]

                iter = 0
                while len(overlaps) &gt; 0:
                    maxo = max(overlaps)
                    mino = max(maxo * odec, overlap_min)
                    if maxo &lt; overlap_min:
                        break
                    i = 0
                    while i &lt; len(overlaps):
                        if overlaps[i] &gt; mino:
                            nlh = likelihood[i] * likelihood[i + 1]
                            likelihood[i] = nlh / np.sum(nlh)
                            segments[i] += segments[i + 1]
                            del likelihood[i + 1]
                            del segments[i + 1]
                            del overlaps[i]
                            if i &lt; len(overlaps):
                                overlaps[i] = likelihood_overlap(likelihood[i], likelihood[i + 1])
                            if i &gt; 0:
                                overlaps[i - 1] = likelihood_overlap(likelihood[i - 1], likelihood[i])
                        else:
                            i = i + 1
                    iter = iter + 1

                overlaps = [[
                    likelihood_overlap(likelihood[i], likelihood[j])
                    if (segments[j][0] - segments[i][-1]) &lt; max_distance * (
                            len(segments[i]) + len(segments[j])) and i &lt; j
                    else 0 for j in range(len(segments))]
                    for i in range(len(segments))]

                iter = 0
                ons = -1

                while True:
                    overlaps = [likelihood_overlap(likelihood[i], likelihood[j]) for i in range(len(likelihood))
                                for j in range(i + 1, len(likelihood)) if
                                (segments[j][0] - segments[i][-1]) &lt; max_distance * (
                                            len(segments[i]) + len(segments[j]))]
                    if len(overlaps) == 0:
                        break
                    maxo = max(overlaps)
                    mino = max(maxo * odec, overlap_min)
                    if maxo &lt; omin:
                        break
                    i, j = 0, 1
                    while i &lt; len(segments) - 1:
                        #                            if overlaps[i][j] &gt; mino:
                        if likelihood_overlap(likelihood[i], likelihood[j]) &gt; mino and (
                                segments[j][0] - segments[i][-1]) &lt; max_distance * (
                                len(segments[i]) + len(segments[j])):
                            nlh = likelihood[i] * likelihood[j]
                            likelihood[i] = nlh / np.sum(nlh)
                            segments[i] += segments[j]
                            segments[i] = sorted(segments[i])
                            del likelihood[j]
                            del segments[j]

                            if j &gt;= len(segments):
                                i += 1
                                j = i + 1
                        else:
                            j += 1
                            if j &gt;= len(segments):
                                i += 1
                                j = i + 1

                    iter = iter + 1
                    if ons == len(segments):
                        break
                    ons = len(segments)

                for i in range(len(segments)):
                    max_lh = np.amax(likelihood[i])
                    ix = np.where(likelihood[i] == max_lh)[0][0]
                    i1 = 1.0 * (res // 2 - ix) / (res + 1) if ix &lt;= (res // 2) else 1.0 * (
                            ix - res // 2) / (res + 1)
                    i2 = likelihood[i][res // 2] / max_lh

                    print(c + &#34;:&#34; + str(segments[i][0] * bin_size + 1) + &#34;-&#34; + str(
                        segments[i][-1] * bin_size + bin_size),
                          bin_size, i1,
                          i2, len(segments[i]))</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.copy_gc"><code class="name flex">
<span>def <span class="ident">copy_gc</span></span>(<span>self, filename, chroms=[])</span>
</code></dt>
<dd>
<section class="desc"><p>Copy GC content from another cnvnator file</p>
<h2 id="arguments">Arguments</h2>
<ul>
<li>cnvnator filename</li>
<li>list of chromosome names</li>
</ul></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def copy_gc(self, filename, chroms=[]):
    &#34;&#34;&#34; Copy GC content from another cnvnator file
            Arguments:
                * cnvnator filename
                * list of chromosome names
    &#34;&#34;&#34;
    io_src = IO(filename)
    chr_rdn = []
    for c in io_src.chromosomes_with_signal(None, &#34;GC/AT&#34;):
        rd_name = self.io.rd_chromosome_name(c)
        if len(chroms) == 0 or (rd_name in chroms) or (c in chroms):
            if rd_name is None:
                _logger.debug(&#34;Not found RD signal for chromosome &#39;%s&#39;. Skipping...&#34; % c)
            else:
                _logger.debug(&#34;Found RD signal for chromosome &#39;%s&#39; with name &#39;%s&#39;.&#34; % (c, rd_name))
                chr_rdn.append((c, rd_name))
    count = 0
    for c, rdn in chr_rdn:
        _logger.info(
            &#34;Copying GC data for chromosome &#39;%s&#39; from file &#39;%s&#39; in file &#39;%s&#39;.&#34; % (c, filename, self.io.filename))
        self.io.create_signal(rdn, None, &#34;GC/AT&#34;, io_src.get_signal(c, None, &#34;GC/AT&#34;).astype(dtype=&#34;uint8&#34;))
    return count</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.copy_mask"><code class="name flex">
<span>def <span class="ident">copy_mask</span></span>(<span>self, filename, chroms=[])</span>
</code></dt>
<dd>
<section class="desc"><p>Copy strict mask data from another cnvnator file</p>
<h2 id="arguments">Arguments</h2>
<ul>
<li>cnvnator filename</li>
<li>list of chromosome names</li>
</ul></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def copy_mask(self, filename, chroms=[]):
    &#34;&#34;&#34; Copy strict mask data from another cnvnator file
            Arguments:
                * cnvnator filename
                * list of chromosome names
    &#34;&#34;&#34;
    io_src = IO(filename)
    chr_rdn = []
    for c in io_src.chromosomes_with_signal(None, &#34;mask&#34;):
        rd_name = self.io.rd_chromosome_name(c)
        snp_name = self.snp_chromosome_name(c)
        if len(chroms) == 0 or (rd_name in chroms) or (snp_name in chroms) or (c in chroms):
            if (rd_name is None) and (snp_name is None):
                _logger.debug(&#34;Not found RD or SNP signal for chromosome &#39;%s&#39;. Skipping...&#34; % c)
            elif (rd_name is None):
                _logger.debug(&#34;Found SNP signal for chromosome &#39;%s&#39; with name &#39;%s&#39;.&#34; % (c, snp_name))
                chr_rdn.append((c, snp_name))
            else:
                _logger.debug(&#34;Found RD signal for chromosome &#39;%s&#39; with name &#39;%s&#39;.&#34; % (c, rd_name))
                chr_rdn.append((c, rd_name))
    count = 0
    for c, rdn in chr_rdn:
        _logger.info(
            &#34;Copying strict mask data for chromosome &#39;%s&#39; from file &#39;%s&#39; in file &#39;%s&#39;.&#34; % (
                c, filename, self.io.filename))
        self.io.create_signal(rdn, None, &#34;mask&#34;, io_src.get_signal(c, None, &#34;mask&#34;).astype(dtype=&#34;uint32&#34;))
    return count</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.gc"><code class="name flex">
<span>def <span class="ident">gc</span></span>(<span>self, filename, chroms=[], make_gc_genome_file=False)</span>
</code></dt>
<dd>
<section class="desc"><p>Read GC content from reference fasta file and store in .cnvnator file</p>
<h2 id="arguments">Arguments</h2>
<ul>
<li>FASTA filename</li>
<li>list of chromosome names</li>
</ul></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def gc(self, filename, chroms=[], make_gc_genome_file=False):
    &#34;&#34;&#34; Read GC content from reference fasta file and store in .cnvnator file
            Arguments:
                * FASTA filename
                * list of chromosome names
    &#34;&#34;&#34;
    fasta = Fasta(filename)
    chrname, chrlen = fasta.get_chr_len()
    chr_len_rdn = []
    if make_gc_genome_file:
        _logger.debug(&#34;GC data for all reference genome contigs will be read and stored in cnvnator file.&#34;)
        for (c, l) in zip(chrname, chrlen):
            chr_len_rdn.append((c, l, c))
    else:
        for (c, l) in zip(chrname, chrlen):
            rd_name = self.io.rd_chromosome_name(c)
            if len(chroms) == 0 or (rd_name in chroms) or (c in chroms):
                if rd_name is None:
                    _logger.debug(&#34;Not found RD signal for chromosome &#39;%s&#39;. Skipping...&#34; % c)
                else:
                    _logger.debug(&#34;Found RD signal for chromosome &#39;%s&#39; with name &#39;%s&#39;.&#34; % (c, rd_name))
                    chr_len_rdn.append((c, l, rd_name))
    count = 0
    for c, l, rdn in chr_len_rdn:
        _logger.info(&#34;Reading data for chromosome &#39;%s&#39; with length %d&#34; % (c, l))
        gc, at = fasta.read_chromosome_gc(c)
        if not gc is None:
            self.io.create_signal(rdn, None, &#34;GC/AT&#34;, gc_at_compress(gc, at))
            count += 1
            _logger.info(&#34;GC data for chromosome &#39;%s&#39; saved in file &#39;%s&#39;.&#34; % (rdn, self.io.filename))
    if count &gt; 0:
        _logger.info(&#34;Running RD statistics on chromosomes with imported GC data.&#34;)
        self.rd_stat()
    return count</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.ls"><code class="name flex">
<span>def <span class="ident">ls</span></span>(<span>self)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def ls(self):
    self.io.ls()</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.mask"><code class="name flex">
<span>def <span class="ident">mask</span></span>(<span>self, filename, chroms=[], make_mask_genome_file=False)</span>
</code></dt>
<dd>
<section class="desc"><p>Read strict mask fasta file and store in .cnvnator file</p>
<h2 id="arguments">Arguments</h2>
<ul>
<li>FASTA filename</li>
<li>list of chromosome names</li>
</ul></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def mask(self, filename, chroms=[], make_mask_genome_file=False):
    &#34;&#34;&#34; Read strict mask fasta file and store in .cnvnator file
            Arguments:
                * FASTA filename
                * list of chromosome names
    &#34;&#34;&#34;
    fasta = Fasta(filename)
    chrname, chrlen = fasta.get_chr_len()
    chr_len_rdn = []
    if make_mask_genome_file:
        _logger.debug(&#34;Strict mask data for all reference genome contigs will be read and stored in cnvnator file.&#34;)
        for (c, l) in zip(chrname, chrlen):
            chr_len_rdn.append((c, l, c))
    else:
        for (c, l) in zip(chrname, chrlen):
            rd_name = self.io.rd_chromosome_name(c)
            snp_name = self.io.snp_chromosome_name(c)
            if len(chroms) == 0 or (rd_name in chroms) or (snp_name in chroms) or (c in chroms):
                if (rd_name is None) and (snp_name is None):
                    _logger.debug(&#34;Not found RD or SNP signal for chromosome &#39;%s&#39;. Skipping...&#34; % c)
                elif (rd_name is None):
                    _logger.debug(&#34;Found SNP signal for chromosome &#39;%s&#39; with name &#39;%s&#39;.&#34; % (c, snp_name))
                    chr_len_rdn.append((c, l, snp_name))
                else:
                    _logger.debug(&#34;Found RD signal for chromosome &#39;%s&#39; with name &#39;%s&#39;.&#34; % (c, rd_name))
                    chr_len_rdn.append((c, l, rd_name))
    count = 0
    for c, l, rdn in chr_len_rdn:
        _logger.info(&#34;Reading data for chromosome &#39;%s&#39; with length %d&#34; % (c, l))
        mask = fasta.read_chromosome_mask_p_regions(c)
        if not mask is None:
            self.io.create_signal(rdn, None, &#34;mask&#34;, mask_compress(mask))
            count += 1
            _logger.info(&#34;Strict mask data for chromosome &#39;%s&#39; saved in file &#39;%s&#39;.&#34; % (rdn, self.io.filename))
    return count</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.mask_snps"><code class="name flex">
<span>def <span class="ident">mask_snps</span></span>(<span>self)</span>
</code></dt>
<dd>
<section class="desc"><p>Flags SNPs in P-region (sets second bit of the flag to 1 for SNP inside P region, or to 0 otherwise).
Requires imported mask data or recognized reference genome with mask data.</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def mask_snps(self):
    &#34;&#34;&#34;
    Flags SNPs in P-region (sets second bit of the flag to 1 for SNP inside P region, or to 0 otherwise).
    Requires imported mask data or recognized reference genome with mask data.

    &#34;&#34;&#34;
    snp_mask_chromosomes = {}
    for c in self.io_mask.mask_chromosomes():
        snp_name = self.io.snp_chromosome_name(c)
        if snp_name is not None:
            snp_mask_chromosomes[snp_name] = c

    for c in self.io.snp_chromosomes():
        if c in snp_mask_chromosomes:
            _logger.info(&#34;Masking SNP data for chromosome &#39;%s&#39;.&#34; % c)
            pos, ref, alt, nref, nalt, gt, flag, qual = self.io.read_snp(c)
            mask = mask_decompress(self.io_mask.get_signal(snp_mask_chromosomes[c], None, &#34;mask&#34;))
            mask_ix = 0
            for snp_ix in range(len(pos)):
                while mask_ix &lt; len(mask) and mask[mask_ix][1] &lt; pos[snp_ix]:
                    mask_ix += 1
                if mask_ix &lt; len(mask) and mask[mask_ix][0] &lt; pos[snp_ix] and pos[snp_ix] &lt; (mask[mask_ix][1] + 1):
                    flag[snp_ix] = flag[snp_ix] | 2
                else:
                    flag[snp_ix] = flag[snp_ix] &amp; 1
            self.io.save_snp(c, pos, ref, alt, nref, nalt, gt, flag, qual, update=True)</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.pileup"><code class="name flex">
<span>def <span class="ident">pileup</span></span>(<span>self, bamfiles, chroms=[])</span>
</code></dt>
<dd>
<section class="desc"><p>Read bam/sam/cram file(s) and pile up SNP counts at positions imported with vcf(vcf_files, &hellip;) method</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>bamfiles</code></strong> :&ensp;<code>list</code> of <code>str</code></dt>
<dd>Bam files.</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def pileup(self, bamfiles, chroms=[]):
    &#34;&#34;&#34;
    Read bam/sam/cram file(s) and pile up SNP counts at positions imported with vcf(vcf_files, ...) method

    Parameters
    ----------
    bamfiles : list of str
        Bam files.
    &#34;&#34;&#34;
    chrs = [c for c in self.io.snp_chromosomes() if (len(chroms) == 0 or c in chroms)]

    pos, ref, alt, nref, nalt, gt, flag, qual = {}, {}, {}, {}, {}, {}, {}, {}
    for c in chrs:
        _logger.info(&#34;Decompressing and setting all SNP counts to zero for chromosome &#39;%s&#39;.&#34; % c)
        pos[c], ref[c], alt[c], nref[c], nalt[c], gt[c], flag[c], qual[c] = self.io.read_snp(c)
        nref[c] = [0] * len(nref[c])
        nalt[c] = [0] * len(nalt[c])
    for bf in bamfiles:
        self.pileup_bam(bf, chrs, pos, ref, alt, nref, nalt)

    for c in chrs:
        _logger.info(&#34;Saving SNP data for chromosome &#39;%s&#39; in file &#39;%s&#39;.&#34; % (c, self.io.filename))
        self.io.save_snp(c, pos[c], ref[c], alt[c], nref[c], nalt[c], gt[c], flag[c], qual[c])</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.pileup_bam"><code class="name flex">
<span>def <span class="ident">pileup_bam</span></span>(<span>self, bamfile, chroms, pos, ref, alt, nref, nalt)</span>
</code></dt>
<dd>
<section class="desc"><p>TODO:</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>bamfile</code></strong></dt>
<dd>&nbsp;</dd>
<dt><strong><code>chroms</code></strong></dt>
<dd>&nbsp;</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def pileup_bam(self, bamfile, chroms, pos, ref, alt, nref, nalt):
    &#34;&#34;&#34;
    TODO:

    Parameters
    ----------
    bamfile
    chroms

    &#34;&#34;&#34;
    _logger.info(&#34;Calculating pileup from bam file &#39;%s&#39;.&#34; % bamfile)
    bamf = Bam(bamfile)

    def pileup_chromosome(x):
        c, l, snpc = x
        _logger.info(&#34;Pileup chromosome %s with length %d&#34; % (c, l))
        return bamf.pileup(c, pos[snpc], ref[snpc], alt[snpc])

    chrname, chrlen = bamf.get_chr_len()
    chr_len = [(c, l, self.io.snp_chromosome_name(c)) for (c, l) in zip(chrname, chrlen) if
               self.io.snp_chromosome_name(c) in chroms]

    if self.max_cores == 1:
        for cl in chr_len:
            r = pileup_chromosome(cl)
            nref[cl[2]] = [x + y for x, y in zip(nref[cl[2]], r[0])]
            nalt[cl[2]] = [x + y for x, y in zip(nalt[cl[2]], r[1])]

    else:
        from .pool import parmap
        res = parmap(pileup_chromosome, chr_len, cores=self.max_cores)
        for cl, r in zip(chr_len, res):
            nref[cl[2]] = [x + y for x, y in zip(nref[cl[2]], r[0])]
            nalt[cl[2]] = [x + y for x, y in zip(nalt[cl[2]], r[1])]</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.rd"><code class="name flex">
<span>def <span class="ident">rd</span></span>(<span>self, bamfiles, chroms=[])</span>
</code></dt>
<dd>
<section class="desc"><p>Read chromosomes from bam/sam/cram file(s) and store in .cnvnator file</p>
<h2 id="arguments">Arguments</h2>
<ul>
<li>list of bam file names</li>
<li>list of chromosome names</li>
</ul></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def rd(self, bamfiles, chroms=[]):
    &#34;&#34;&#34; Read chromosomes from bam/sam/cram file(s) and store in .cnvnator file
            Arguments:
                * list of bam file names
                * list of chromosome names
    &#34;&#34;&#34;
    hm = 0
    for bf in bamfiles:
        hm += self.read_bam(bf, chroms)

    if self.io.signal_exists(None, None, &#34;reference genome&#34;):
        rg_name = np.array(self.io.get_signal(None, None, &#34;reference genome&#34;)).astype(&#34;str&#34;)[0]
        if &#34;mask_file&#34; in Genome.reference_genomes[rg_name]:
            _logger.info(&#34;Strict mask for reference genome &#39;%s&#39; found in database.&#34; % rg_name)
            self.io_mask = IO(Genome.reference_genomes[rg_name][&#34;mask_file&#34;])
        if &#34;gc_file&#34; in Genome.reference_genomes[rg_name]:
            _logger.info(&#34;GC content for reference genome &#39;%s&#39; found in database.&#34; % rg_name)
            self.io_gc = IO(Genome.reference_genomes[rg_name][&#34;gc_file&#34;])
            self.rd_stat()

    return hm</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.rd_stat"><code class="name flex">
<span>def <span class="ident">rd_stat</span></span>(<span>self, chroms=[], plot=True)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def rd_stat(self, chroms=[], plot=True):
    rd_stat = []
    auto, sex, mt = False, False, False

    rd_gc_chromosomes = {}
    for c in self.io_gc.gc_chromosomes():
        rd_name = self.io.rd_chromosome_name(c)
        if not rd_name is None:
            rd_gc_chromosomes[rd_name] = c

    for c in rd_gc_chromosomes:
        if (len(chroms) == 0 or c in chroms):
            rd_p, rd_u = self.io.read_rd(c)
            max_bin = max(int(10 * np.mean(rd_u) + 1), int(10 * np.mean(rd_p) + 1))
            dist_p, bins = np.histogram(rd_p, bins=range(max_bin + 1))
            dist_u, bins = np.histogram(rd_u, bins=range(max_bin + 1))
            n_p, m_p, s_p = fit_normal(bins[1:-1], dist_p[1:])[0]
            n_u, m_u, s_u = fit_normal(bins[1:-1], dist_u[1:])[0]
            _logger.info(
                &#34;Chromosome &#39;%s&#39; stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (c, m_p, s_p))
            _logger.info(
                &#34;Chromosome &#39;%s&#39; stat - RD unique distribution gaussian fit:  %.2f +- %.2f&#34; % (c, m_u, s_u))
            rd_stat.append((c, m_p, s_p, m_u, s_u))
            auto = auto or Genome.is_autosome(c)
            sex = sex or Genome.is_sex_chrom(c)
            mt = mt or Genome.is_mt_chrom(c)

    if auto:
        max_bin_auto = int(
            max(map(lambda x: 5 * x[1] + 5 * x[2], filter(lambda x: Genome.is_autosome(x[0]), rd_stat))))
        _logger.debug(&#34;Max RD for autosomes calculated: %d&#34; % max_bin_auto)
        dist_p_auto = np.zeros(max_bin_auto)
        dist_u_auto = np.zeros(max_bin_auto)
        dist_p_gc_auto = np.zeros((max_bin_auto, 101))
        bins_auto = range(max_bin_auto + 1)
    if sex:
        max_bin_sex = int(
            max(map(lambda x: 5 * x[1] + 5 * x[2], filter(lambda x: Genome.is_sex_chrom(x[0]), rd_stat))))
        _logger.debug(&#34;Max RD for sex chromosomes calculated: %d&#34; % max_bin_sex)
        dist_p_sex = np.zeros(max_bin_sex)
        dist_u_sex = np.zeros(max_bin_sex)
        dist_p_gc_sex = np.zeros((max_bin_sex, 101))
        bins_sex = range(max_bin_sex + 1)

    if mt:
        max_bin_mt = int(
            max(map(lambda x: 5 * x[1] + 5 * x[2], filter(lambda x: Genome.is_mt_chrom(x[0]), rd_stat))))
        _logger.debug(&#34;Max RD for mitochondria calculated: %d&#34; % max_bin_mt)
        bin_size_mt = max_bin_mt // 100
        bins_mt = range(0, max_bin_mt // bin_size_mt * bin_size_mt + bin_size_mt, bin_size_mt)
        n_bins_mt = len(bins_mt) - 1
        _logger.debug(&#34;Using %d bin size, %d bins for mitochondria chromosome.&#34; % (bin_size_mt, n_bins_mt))
        dist_p_mt = np.zeros(n_bins_mt)
        dist_u_mt = np.zeros(n_bins_mt)
        dist_p_gc_mt = np.zeros((n_bins_mt, 101))

    for c in rd_gc_chromosomes:
        if len(chroms) == 0 or c in chroms:
            _logger.info(&#34;Chromosome &#39;%s&#39; stat &#34; % c)
            rd_p, rd_u = self.io.read_rd(c)
            if Genome.is_autosome(c):
                dist_p, bins = np.histogram(rd_p, bins=bins_auto)
                dist_u, bins = np.histogram(rd_u, bins=bins_auto)
                gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                gcp = gcp_decompress(gcat)
                dist_p_gc, xbins, ybins = np.histogram2d(rd_p, gcp, bins=(bins_auto, range(102)))
                dist_p_auto += dist_p
                dist_u_auto += dist_u
                dist_p_gc_auto += dist_p_gc
            elif Genome.is_sex_chrom(c):
                dist_p, bins = np.histogram(rd_p, bins=bins_sex)
                dist_u, bins = np.histogram(rd_u, bins=bins_sex)
                gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                gcp = gcp_decompress(gcat)
                dist_p_gc, xbins, ybins = np.histogram2d(rd_p, gcp, bins=(bins_sex, range(102)))
                dist_p_sex += dist_p
                dist_u_sex += dist_u
                dist_p_gc_sex += dist_p_gc
            elif Genome.is_mt_chrom(c):
                dist_p, bins = np.histogram(rd_p, bins=bins_mt)
                dist_u, bins = np.histogram(rd_u, bins=bins_mt)
                gcat = self.io_gc.get_signal(rd_gc_chromosomes[c], None, &#34;GC/AT&#34;)
                gcp = gcp_decompress(gcat)
                dist_p_gc, xbins, ybins = np.histogram2d(rd_p, gcp, bins=(bins_mt, range(102)))
                dist_p_mt += dist_p
                dist_u_mt += dist_u
                dist_p_gc_mt += dist_p_gc

    if auto:
        n_auto, m_auto, s_auto = fit_normal(np.array(bins_auto[1:-1]), dist_p_auto[1:])[0]
        _logger.info(&#34;Autosomes stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (m_auto, s_auto))
        stat_auto = np.array([max_bin_auto, 1, max_bin_auto, n_auto, m_auto, s_auto])
        self.io.create_signal(None, 100, &#34;RD p dist&#34;, dist_p_auto, flags=FLAG_AUTO)
        self.io.create_signal(None, 100, &#34;RD u dist&#34;, dist_u_auto, flags=FLAG_AUTO)
        self.io.create_signal(None, 100, &#34;RD GC dist&#34;, dist_p_gc_auto, flags=FLAG_AUTO)
        self.io.create_signal(None, 100, &#34;RD stat&#34;, stat_auto, flags=FLAG_AUTO)
        gc_corr_auto = calculate_gc_correction(dist_p_gc_auto, m_auto, s_auto)
        self.io.create_signal(None, 100, &#34;GC corr&#34;, gc_corr_auto, flags=FLAG_AUTO)

    if sex:
        n_sex, m_sex, s_sex = fit_normal(np.array(bins_sex[1:-1]), dist_p_sex[1:])[0]
        _logger.info(&#34;Sex chromosomes stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (m_sex, s_sex))
        stat_sex = np.array([max_bin_sex, 1, max_bin_sex, n_sex, m_sex, s_sex])
        self.io.create_signal(None, 100, &#34;RD p dist&#34;, dist_p_sex, flags=FLAG_SEX)
        self.io.create_signal(None, 100, &#34;RD u dist&#34;, dist_u_sex, flags=FLAG_SEX)
        self.io.create_signal(None, 100, &#34;RD GC dist&#34;, dist_p_gc_sex, flags=FLAG_SEX)
        self.io.create_signal(None, 100, &#34;RD stat&#34;, stat_sex, flags=FLAG_SEX)
        gc_corr_sex = calculate_gc_correction(dist_p_gc_sex, m_sex, s_sex)
        self.io.create_signal(None, 100, &#34;GC corr&#34;, gc_corr_sex, flags=FLAG_SEX)

    if mt:
        n_mt, m_mt, s_mt = fit_normal(np.array(bins_mt[1:-1]), dist_p_mt[1:])[0]
        _logger.info(&#34;Mitochondria stat - RD parity distribution gaussian fit:  %.2f +- %.2f&#34; % (m_mt, s_mt))
        if auto:
            _logger.info(&#34;Mitochondria stat - number of mitochondria per cell:  %.2f +- %.2f&#34; % (
                2. * m_mt / m_auto, 2. * s_mt / m_auto + s_auto * m_mt / (m_auto * m_auto)))
        stat_mt = np.array([max_bin_mt, bin_size_mt, n_bins_mt, n_mt, m_mt, s_mt])
        self.io.create_signal(None, 100, &#34;RD p dist&#34;, dist_p_mt, flags=FLAG_MT)
        self.io.create_signal(None, 100, &#34;RD u dist&#34;, dist_u_mt, flags=FLAG_MT)
        self.io.create_signal(None, 100, &#34;RD GC dist&#34;, dist_p_gc_mt, flags=FLAG_MT)
        self.io.create_signal(None, 100, &#34;RD stat&#34;, stat_mt, flags=FLAG_MT)
        gc_corr_mt = calculate_gc_correction(dist_p_gc_mt, m_mt, s_mt, bin_size_mt)
        self.io.create_signal(None, 100, &#34;GC corr&#34;, gc_corr_mt, flags=FLAG_MT)</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.read_bam"><code class="name flex">
<span>def <span class="ident">read_bam</span></span>(<span>self, bf, chroms)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def read_bam(self, bf, chroms):
    bamf = Bam(bf)
    if bamf.reference_genome:
        self.io.create_signal(None, None, &#34;reference genome&#34;, np.array([np.string_(bamf.reference_genome)]))
        self.io.create_signal(None, None, &#34;use reference&#34;, np.array([1, 1]).astype(&#34;uint8&#34;))

    def read_chromosome(x):
        chrs, lens = x
        _logger.info(&#34;Reading data for chromosome %s with length %d&#34; % (chrs, lens))
        l_rd_p, l_rd_u, l_his_read_frg = bamf.read_chromosome(chrs)
        return l_rd_p, l_rd_u, l_his_read_frg

    chrname, chrlen = bamf.get_chr_len()
    chr_len = [(c, l) for (c, l) in zip(chrname, chrlen) if len(chroms) == 0 or c in chroms]
    if self.max_cores == 1:
        count = 0
        cum_his_read_frg = None
        for cl in chr_len:
            rd_p, rd_u, his_read_frg = read_chromosome(cl)
            if not rd_p is None:
                if cum_his_read_frg is None:
                    cum_his_read_frg = his_read_frg
                else:
                    cum_his_read_frg += his_read_frg
                self.io.save_rd(cl[0], rd_p, rd_u)
                count += 1
        if not cum_his_read_frg is None:
            self.io.create_signal(None, None, &#34;read frg dist&#34;, cum_his_read_frg)
        return count
    else:
        from .pool import parmap
        res = parmap(read_chromosome, chr_len, cores=self.max_cores)
        count = 0
        cum_his_read_frg = None
        for c, r in zip(chr_len, res):
            if not r[0] is None:
                if cum_his_read_frg is None:
                    cum_his_read_frg = r[2]
                else:
                    cum_his_read_frg += r[2]
                self.io.save_rd(c[0], r[0], r[1])
                count += 1
        if not cum_his_read_frg is None:
            self.io.create_signal(None, None, &#34;read frg dist&#34;, cum_his_read_frg)
        return count</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.read_vcf"><code class="name flex">
<span>def <span class="ident">read_vcf</span></span>(<span>self, vcf_file, chroms, sample='', use_index=False)</span>
</code></dt>
<dd>
<section class="desc"><h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>vcf_file</code></strong></dt>
<dd>&nbsp;</dd>
<dt><strong><code>chroms</code></strong></dt>
<dd>&nbsp;</dd>
<dt><strong><code>sample</code></strong></dt>
<dd>&nbsp;</dd>
<dt><strong><code>use_index</code></strong></dt>
<dd>&nbsp;</dd>
</dl>
<h2 id="returns">Returns</h2></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def read_vcf(self, vcf_file, chroms, sample=&#39;&#39;, use_index=False):
    &#34;&#34;&#34;

    Parameters
    ----------
    vcf_file
    chroms
    sample
    use_index

    Returns
    -------

    &#34;&#34;&#34;
    vcff = Vcf(vcf_file)
    chrs = [c for c in vcff.get_chromosomes() if len(chroms) == 0 or c in chroms]

    def save_data(chr, pos, ref, alt, nref, nalt, gt, flag, qual):
        if (len(chroms) == 0 or chr in chroms) and (not pos is None) and (len(pos) &gt; 0):
            self.io.save_snp(chr, pos, ref, alt, nref, nalt, gt, flag, qual)
        # TODO: Stop reading if all form chrom list are read.

    if use_index:
        count = 0
        for c in chrs:
            _logger.info(&#34;Reading variant data for chromosome %s&#34; % c)
            pos, ref, alt, nref, nalt, gt, flag, qual = vcff.read_chromosome_snp(c, sample)

            if not pos is None and len(pos) &gt; 0:
                self.io.save_snp(c, pos, ref, alt, nref, nalt, gt, flag, qual)
                count += 1
        return count
    else:
        return vcff.read_all_snp(save_data, sample)</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.set_reference_genome"><code class="name flex">
<span>def <span class="ident">set_reference_genome</span></span>(<span>self, rg)</span>
</code></dt>
<dd>
<section class="desc"><p>Manually sets reference genome.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>rg</code></strong> :&ensp;<code>str</code></dt>
<dd>Name of reference genome</dd>
</dl>
<h2 id="returns">Returns</h2>
<pre><code>True if genome exists in database
</code></pre></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def set_reference_genome(self, rg):
    &#34;&#34;&#34; Manually sets reference genome.

    Parameters
    ----------
    rg: str
        Name of reference genome

    Returns
    -------
        True if genome exists in database

    &#34;&#34;&#34;
    if rg in Genome.reference_genomes:
        self.io.create_signal(None, None, &#34;reference genome&#34;, np.array([np.string_(rg)]))
        self.io.create_signal(None, None, &#34;use reference&#34;, np.array([1, 1]).astype(&#34;uint8&#34;))</code></pre>
</details>
</dd>
<dt id="cnvpytor.root.Root.vcf"><code class="name flex">
<span>def <span class="ident">vcf</span></span>(<span>self, vcf_files, chroms=[], sample='')</span>
</code></dt>
<dd>
<section class="desc"><p>Read SNP data from variant file(s) and store in .cnvnator file</p>
<h2 id="arguments">Arguments</h2>
<ul>
<li>list of variant file names</li>
<li>list of chromosome names</li>
</ul></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def vcf(self, vcf_files, chroms=[], sample=&#39;&#39;):
    &#34;&#34;&#34; Read SNP data from variant file(s) and store in .cnvnator file
            Arguments:
                * list of variant file names
                * list of chromosome names
    &#34;&#34;&#34;
    hm = 0
    for vcf_file in vcf_files:
        hm += self.read_vcf(vcf_file, chroms, sample)
    return hm</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="cnvpytor" href="index.html">cnvpytor</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="cnvpytor.root.Root" href="#cnvpytor.root.Root">Root</a></code></h4>
<ul class="">
<li><code><a title="cnvpytor.root.Root.calculate_baf" href="#cnvpytor.root.Root.calculate_baf">calculate_baf</a></code></li>
<li><code><a title="cnvpytor.root.Root.calculate_histograms" href="#cnvpytor.root.Root.calculate_histograms">calculate_histograms</a></code></li>
<li><code><a title="cnvpytor.root.Root.call_baf" href="#cnvpytor.root.Root.call_baf">call_baf</a></code></li>
<li><code><a title="cnvpytor.root.Root.copy_gc" href="#cnvpytor.root.Root.copy_gc">copy_gc</a></code></li>
<li><code><a title="cnvpytor.root.Root.copy_mask" href="#cnvpytor.root.Root.copy_mask">copy_mask</a></code></li>
<li><code><a title="cnvpytor.root.Root.gc" href="#cnvpytor.root.Root.gc">gc</a></code></li>
<li><code><a title="cnvpytor.root.Root.ls" href="#cnvpytor.root.Root.ls">ls</a></code></li>
<li><code><a title="cnvpytor.root.Root.mask" href="#cnvpytor.root.Root.mask">mask</a></code></li>
<li><code><a title="cnvpytor.root.Root.mask_snps" href="#cnvpytor.root.Root.mask_snps">mask_snps</a></code></li>
<li><code><a title="cnvpytor.root.Root.pileup" href="#cnvpytor.root.Root.pileup">pileup</a></code></li>
<li><code><a title="cnvpytor.root.Root.pileup_bam" href="#cnvpytor.root.Root.pileup_bam">pileup_bam</a></code></li>
<li><code><a title="cnvpytor.root.Root.rd" href="#cnvpytor.root.Root.rd">rd</a></code></li>
<li><code><a title="cnvpytor.root.Root.rd_stat" href="#cnvpytor.root.Root.rd_stat">rd_stat</a></code></li>
<li><code><a title="cnvpytor.root.Root.read_bam" href="#cnvpytor.root.Root.read_bam">read_bam</a></code></li>
<li><code><a title="cnvpytor.root.Root.read_vcf" href="#cnvpytor.root.Root.read_vcf">read_vcf</a></code></li>
<li><code><a title="cnvpytor.root.Root.set_reference_genome" href="#cnvpytor.root.Root.set_reference_genome">set_reference_genome</a></code></li>
<li><code><a title="cnvpytor.root.Root.vcf" href="#cnvpytor.root.Root.vcf">vcf</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.6.3</a>.</p>
</footer>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script>hljs.initHighlightingOnLoad()</script>
</body>
</html>