<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.6.3" />
<title>cnvpytor.utils API documentation</title>
<meta name="description" content="cnvpytor.utils â€¦" />
<link href='https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.0/normalize.min.css' rel='stylesheet'>
<link href='https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/8.0.0/sanitize.min.css' rel='stylesheet'>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet">
<style>.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{font-weight:bold}#index h4 + ul{margin-bottom:.6em}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase;cursor:pointer}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>cnvpytor.utils</code></h1>
</header>
<section id="section-intro">
<p>cnvpytor.utils</p>
<p>Misc functions</p>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">&#34;&#34;&#34; cnvpytor.utils

Misc functions

&#34;&#34;&#34;
from __future__ import absolute_import, print_function, division
import numpy as np
from argparse import ArgumentTypeError
from scipy.stats import norm
from scipy.optimize import curve_fit
import logging
import readline

_logger = logging.getLogger(&#34;cnvpytor.utils&#34;)


def gc_at_compress(gc, at):
    &#34;&#34;&#34;
    Commpress GC/AT content of 100bins using fact that #at+#gc=100 in large majority of bins.
    Transforms #at, #gc -&gt; #at, 100-#at-#gc

    Parameters
    ----------
    gc : list of int
        Binned GC content (100bp bins).
    at : list of int
        Binned AT content (100bp bins).

    Returns
    -------
    gcat : numpy.ndarray
        Array contains compressed GC/AT content.

    &#34;&#34;&#34;
    cgc = np.array(gc)
    cat = 100 - np.array(at) - cgc
    return np.concatenate((cgc, cat)).astype(&#34;uint8&#34;)


def gc_at_decompress(gcat):
    &#34;&#34;&#34;
    Decompress GT/AC content - inverse function of gc_at_compress(gc, at).

    Parameters
    ----------
    gcat : numpy.ndarray
        Array contains compressed GC/AT content.

    Returns
    -------
    gc : list of int
        Binned GC content (100bp bins).
    at : list of int
        Binned AT content (100bp bins).

    &#34;&#34;&#34;
    return list(gcat[:gcat.size // 2]), list(100 - gcat[:gcat.size // 2] - gcat[gcat.size // 2:])


def gcp_decompress(gcat, bin_ratio=1):
    &#34;&#34;&#34;
    Decompress GT/AC content and calculate GC percentage.

    Parameters
    ----------
    gcat : numpy.ndarray
        Array contains compressed GC/AT content.

    Returns
    -------
    gcp : list of int
        GC percentage.

    &#34;&#34;&#34;
    gc, at = gcat[:gcat.size // 2].astype(&#34;float&#34;), (100 - gcat[:gcat.size // 2] - gcat[gcat.size // 2:]).astype(
        &#34;float&#34;)
    if bin_ratio == 1:
        return 100. * gc / (at + gc + 1e-10)
    n = len(gc)
    his_gc = np.concatenate(
        (np.array(gc), np.zeros(bin_ratio - n + (n // bin_ratio * bin_ratio)))).astype(&#34;float&#34;)
    his_gc.resize((len(his_gc) // bin_ratio, bin_ratio))
    his_gc = his_gc.sum(axis=1)
    his_at = np.concatenate(
        (np.array(at), np.zeros(bin_ratio - n + (n // bin_ratio * bin_ratio)))).astype(&#34;float&#34;)
    his_at.resize((len(his_at) // bin_ratio, bin_ratio))
    his_at = his_at.sum(axis=1)

    return 100. * his_gc / (his_at + his_gc + 1e-10)


bp_to_binary = {&#39;A&#39;: 0, &#39;T&#39;: 3, &#39;G&#39;: 1, &#39;C&#39;: 2, &#39;.&#39;: 4}
binary_to_bp = {0: &#39;A&#39;, 3: &#39;T&#39;, 1: &#39;G&#39;, 2: &#39;C&#39;, 4: &#39;.&#39;}


def snp_compress(pos, ref, alt, nref, nalt, gt, flag, qual):
    &#34;&#34;&#34;
    Commpress SNP information binary in four arrays:
            snp_pos - diferences in positions
            snp_counts - 32 bits: ref count (16 bits), alt count (16 bits)
            snp_desc -  16 bits: ref (3 bits), alt (3 bits), gt (3 bits), flag (2 bits)
            snp_qual - 8 bits

    &#34;&#34;&#34;
    snp_pos = list(map(lambda i, j: i - j,
                       pos, [0] + pos[:-1]))
    snp_counts = list(map(lambda i, j: (i &lt;&lt; 16) + j,
                          nref, nalt))
    snp_desc = list(map(lambda r, a, g, f: ((((((bp_to_binary[r] &lt;&lt; 3) + bp_to_binary[a]) &lt;&lt; 3) + g) &lt;&lt; 2) + f),
                        ref, alt, gt, flag))
    snp_qual = qual
    return np.array(snp_pos, dtype=&#34;uint32&#34;), np.array(snp_desc, dtype=&#34;uint16&#34;), np.array(snp_counts,
                                                                                           dtype=&#34;uint32&#34;), np.array(
        snp_qual, dtype=&#34;uint8&#34;)


def snp_decompress(snp_pos, snp_desc, snp_counts, snp_qual):
    &#34;&#34;&#34;
    Decommpress SNP information

    &#34;&#34;&#34;
    pos = []
    s = 0
    for i in snp_pos:
        s += i
        pos.append(s)
    ref = []
    alt = []
    gt = []
    flag = []
    for i in snp_desc:
        ref.append(binary_to_bp[i &gt;&gt; 8])
        alt.append(binary_to_bp[(i &gt;&gt; 5) &amp; int(&#39;111&#39;, 2)])
        gt.append((i &gt;&gt; 2) &amp; int(&#39;111&#39;, 2))
        flag.append(i &amp; int(&#39;11&#39;, 2))
    nref = []
    nalt = []
    for i in snp_counts:
        nref.append(i &gt;&gt; 16)
        nalt.append(i &amp; int(&#39;ffff&#39;, 16))
    qual = list(snp_qual)

    return pos, ref, alt, nref, nalt, gt, flag, qual


def mask_compress(mask):
    &#34;&#34;&#34; Commpress strict mask P value intervals.
    Argument:
        mask - list of intervals [(start_1, end_1), (start_2, end_2),...]
    &#34;&#34;&#34;
    pos = [p for interval in mask for p in interval]
    cmask = list(map(lambda i, j: i - j, pos, [0] + pos[:-1]))
    return np.array(cmask, dtype=&#34;uint32&#34;)


def mask_decompress(cmask):
    &#34;&#34;&#34; Decommpress strict mask P value intervals.
    &#34;&#34;&#34;
    pos = []
    s = 0
    for i in cmask:
        s += i
        pos.append(s)
    return list(zip(pos[::2], pos[1::2]))


def rd_compress(rd_p, rd_u, data_type=&#34;uint16&#34;):
    &#34;&#34;&#34; Compres RD signal
    &#34;&#34;&#34;
    return rd_p.astype(data_type), (rd_p - rd_u).astype(data_type)


def rd_decompress(crd_p, crd_u):
    &#34;&#34;&#34; Decommpress SNP information
    &#34;&#34;&#34;
    return np.array(crd_p), np.array(crd_p) + np.array(crd_u)


def binsize_type(x):
    x = int(x)
    if x % 100 != 0 or x &lt;= 0:
        raise ArgumentTypeError(&#34;Bin size should be positive integer divisible by 100!&#34;)
    return x


def normal_overlap(m1, s1, m2, s2):
    if m1 &gt; m2:
        m1, m2 = m2, m1
        s1, s2 = s2, s1
    a = 1. / (2. * s1 ** 2) - 1. / (2. * s2 ** 2)
    b = m2 / (s2 ** 2) - m1 / (s1 ** 2)
    c = m1 ** 2 / (2 * s1 ** 2) - m2 ** 2 / (2 * s2 ** 2) - np.log(s2 / s1)
    print(np.roots([a, b, c]))
    roots = sorted(np.roots([a, b, c]))
    if len(roots) == 0:
        return 1
    elif len(roots) == 1:
        r = roots[0]
        return norm.cdf(r, m2, s2) + (1. - norm.cdf(r, m1, s1))
    elif s1 &lt; s2:
        return norm.cdf(roots[1], m2, s2) - norm.cdf(roots[0], m2, s2) + 1 - norm.cdf(roots[1], m1, s1) + norm.cdf(
            roots[0], m1, s1)
    return norm.cdf(roots[1], m1, s1) - norm.cdf(roots[0], m1, s1) + 1 - norm.cdf(roots[1], m2, s2) + norm.cdf(
        roots[0], m2, s2)


def normal_merge(m1, s1, m2, s2):
    if s1 == 0 and s2 == 0:
        return 0.5 * (m1 + m2), 0
    else:
        return (m1 * s2 * s2 + m2 * s1 * s1) / (s1 * s1 + s2 * s2), np.sqrt(s1 * s1 * s2 * s2 / (s1 * s1 + s2 * s2))


def normal(x, a, x0, sigma):
    return a * np.exp(-(x - x0) ** 2 / (2 * sigma ** 2))


def fit_normal(x, y):
    &#34;&#34;&#34; Fit Gaussian
    &#34;&#34;&#34;
    if sum(y) == 0:
        _logger.debug(&#34;Problem with fit: all data points have zero value. Return zeros instead fit parameters!&#34;)
        return [0, 0, 0], None
    mean = sum(x * y) / sum(y)
    sigma = np.sqrt(sum(y * (x - mean) ** 2) / sum(y))
    _logger.debug(&#34;%f %f %d&#34; % (mean, sigma, len(x)))
    if sigma == 0:
        _logger.debug(&#34;Problem with fit: sigma equals zero. Return zeros instead fit parameters!&#34;)
        return [0, 0, 0], None
    area = sum(y[:-1] * (x[1:] - x[:-1]))
    try:
        popt, pcov = curve_fit(normal, x, y, p0=[area, mean, sigma])
        return popt, pcov
    except ValueError:
        _logger.warning(&#34;Problem with fit: optimal parameters not found. Using mean and std instead!&#34;)
        return [area, mean, sigma], None
    except RuntimeError:
        _logger.warning(&#34;Problem with fit: optimal parameters not found. Using mean and std instead!&#34;)
        return [area, mean, sigma], None


def calculate_gc_correction(his_rd_gc, mean, sigma, bin_size=1):
    &#34;&#34;&#34; Calculate GC correction from RD-GC histogram
    &#34;&#34;&#34;
    max_bin = int(max(2 * mean, mean + 5 * sigma) / bin_size)
    his = his_rd_gc[1:max_bin, :]
    rd = np.repeat((np.arange(1, max_bin * bin_size, bin_size)[:max_bin - 1]).reshape((max_bin - 1, 1)), 101, axis=1)
    np.seterr(divide=&#39;ignore&#39;, invalid=&#39;ignore&#39;)
    gc_corr = np.sum(rd * his, axis=0) / np.sum(his, axis=0)
    gc_corr[np.isnan(gc_corr)] = 1.0
    gc_corr = gc_corr / (np.sum(gc_corr * np.sum(his, axis=0)) / np.sum(np.sum(his, axis=0)))
    return gc_corr


def beta(k, m, p, phased=False):
    &#34;&#34;&#34;
    Returns likelihood beta function f(p) where 0 &lt;= p &lt;= 1.
    Function is not normalized.

    Parameters
    ----------
    k : int
        Number of haplotype 1 reads.
    m : int
        Number of haplotype 2 reads.
    p : float
        Parameter.
    phased : bool
        Likelihood will be symmetrised if not phased.

    Returns
    -------
    f : float
        Value od likelihood function,.

    &#34;&#34;&#34;
    if k == m or phased:
        return 1.0 * p ** k * (1. - p) ** m
    else:
        return 1.0 * p ** k * (1. - p) ** m + 1.0 * p ** m * (1. - p) ** k


def log_beta(k, m, p, phased=False):
    &#34;&#34;&#34;
    Returns logarithm of likelihood beta function f(p) where 0 &lt;= p &lt;= 1.
    Function is not normalized.

    Parameters
    ----------
    k : int
        Number of haplotype 1 reads.
    m : int
        Number of haplotype 2 reads.
    p : float
        Parameter.
    phased : bool
        Likelihood will be symmetrised if not phased.

    Returns
    -------
    f : float
        Value od likelihood function,.

    &#34;&#34;&#34;
    if k == m or phased:
        return np.log(1.0 * p ** k * (1. - p) ** m)
    else:
        return np.log(1.0 * p ** k * (1. - p) ** m + 1.0 * p ** m * (1. - p) ** k)


def likelihood_overlap(lk1, lk2):
    &#34;&#34;&#34;
    Returns overlap area of two likelihood functions.

    Parameters
    ----------
    lk1 : numpy.ndarray
        First likelihood function.
    lk2 : numpy.ndarray
        Second likelihood function.

    Returns
    -------
    overlap : float
        Overlap area.

    &#34;&#34;&#34;
    return np.sum(np.min((lk1, lk2), axis=0))


def decode_position(s):
    &#34;&#34;&#34;

    Parameters
    ----------
    s : str

    Returns
    -------
    str
    &#34;&#34;&#34;
    return int(s.replace(&#34;K&#34;, &#34;000&#34;).replace(&#34;k&#34;, &#34;000&#34;).replace(&#34;M&#34;, &#34;000000&#34;).replace(&#34;m&#34;, &#34;000000&#34;))


def decode_region(s):
    &#34;&#34;&#34;

    Parameters
    ----------
    s : str

    Returns
    -------
    list of tuples

    &#34;&#34;&#34;
    regs = s.split(&#34;,&#34;)
    ret = []
    for r in regs:
        chr_interval = r.split(&#34;:&#34;)
        begin_end = chr_interval[1].split(&#34;-&#34;)
        ret.append((chr_interval[0], (decode_position(begin_end[0]), decode_position(begin_end[1]))))
    return ret


class PromptCompleter:
    def __init__(self, command_tree):
        &#34;&#34;&#34;
        Initialize completer class for GNU readline tab comleter.

        Parameters
        ----------
        command_tree : dictionary
            Example: command_tree = {&#34;comannd1&#34;:None,&#34;command2&#34;:{&#34;subcommand2.1&#34;:None,subcommand2.1:None}}

        &#34;&#34;&#34;
        self.command_tree = command_tree

    def _traverse(self, tokens, tree):
        if tree is None or len(tokens) == 0:
            return []
        if len(tokens) == 1:
            return [x + &#39; &#39; for x in tree if x.startswith(tokens[0])]
        else:
            if tokens[0] in tree.keys():
                return self._traverse(tokens[1:], tree[tokens[0]])
        return []

    def complete(self, text, state):
        try:
            tokens = readline.get_line_buffer().split()
            if not tokens or readline.get_line_buffer()[-1] == &#39; &#39;:
                tokens.append(&#34;&#34;)
            results = self._traverse(tokens, self.command_tree) + [None]
            return results[state]
        except Exception as e:
            print(e)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="cnvpytor.utils.beta"><code class="name flex">
<span>def <span class="ident">beta</span></span>(<span>k, m, p, phased=False)</span>
</code></dt>
<dd>
<section class="desc"><p>Returns likelihood beta function f(p) where 0 &lt;= p &lt;= 1.
Function is not normalized.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>k</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of haplotype 1 reads.</dd>
<dt><strong><code>m</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of haplotype 2 reads.</dd>
<dt><strong><code>p</code></strong> :&ensp;<code>float</code></dt>
<dd>Parameter.</dd>
<dt><strong><code>phased</code></strong> :&ensp;<code>bool</code></dt>
<dd>Likelihood will be symmetrised if not phased.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>f</code></strong> :&ensp;<code>float</code></dt>
<dd>Value od likelihood function,.</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def beta(k, m, p, phased=False):
    &#34;&#34;&#34;
    Returns likelihood beta function f(p) where 0 &lt;= p &lt;= 1.
    Function is not normalized.

    Parameters
    ----------
    k : int
        Number of haplotype 1 reads.
    m : int
        Number of haplotype 2 reads.
    p : float
        Parameter.
    phased : bool
        Likelihood will be symmetrised if not phased.

    Returns
    -------
    f : float
        Value od likelihood function,.

    &#34;&#34;&#34;
    if k == m or phased:
        return 1.0 * p ** k * (1. - p) ** m
    else:
        return 1.0 * p ** k * (1. - p) ** m + 1.0 * p ** m * (1. - p) ** k</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.binsize_type"><code class="name flex">
<span>def <span class="ident">binsize_type</span></span>(<span>x)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def binsize_type(x):
    x = int(x)
    if x % 100 != 0 or x &lt;= 0:
        raise ArgumentTypeError(&#34;Bin size should be positive integer divisible by 100!&#34;)
    return x</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.calculate_gc_correction"><code class="name flex">
<span>def <span class="ident">calculate_gc_correction</span></span>(<span>his_rd_gc, mean, sigma, bin_size=1)</span>
</code></dt>
<dd>
<section class="desc"><p>Calculate GC correction from RD-GC histogram</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def calculate_gc_correction(his_rd_gc, mean, sigma, bin_size=1):
    &#34;&#34;&#34; Calculate GC correction from RD-GC histogram
    &#34;&#34;&#34;
    max_bin = int(max(2 * mean, mean + 5 * sigma) / bin_size)
    his = his_rd_gc[1:max_bin, :]
    rd = np.repeat((np.arange(1, max_bin * bin_size, bin_size)[:max_bin - 1]).reshape((max_bin - 1, 1)), 101, axis=1)
    np.seterr(divide=&#39;ignore&#39;, invalid=&#39;ignore&#39;)
    gc_corr = np.sum(rd * his, axis=0) / np.sum(his, axis=0)
    gc_corr[np.isnan(gc_corr)] = 1.0
    gc_corr = gc_corr / (np.sum(gc_corr * np.sum(his, axis=0)) / np.sum(np.sum(his, axis=0)))
    return gc_corr</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.decode_position"><code class="name flex">
<span>def <span class="ident">decode_position</span></span>(<span>s)</span>
</code></dt>
<dd>
<section class="desc"><h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>s</code></strong> :&ensp;<code>str</code></dt>
<dd>&nbsp;</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>str</code></dt>
<dd>&nbsp;</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def decode_position(s):
    &#34;&#34;&#34;

    Parameters
    ----------
    s : str

    Returns
    -------
    str
    &#34;&#34;&#34;
    return int(s.replace(&#34;K&#34;, &#34;000&#34;).replace(&#34;k&#34;, &#34;000&#34;).replace(&#34;M&#34;, &#34;000000&#34;).replace(&#34;m&#34;, &#34;000000&#34;))</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.decode_region"><code class="name flex">
<span>def <span class="ident">decode_region</span></span>(<span>s)</span>
</code></dt>
<dd>
<section class="desc"><h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>s</code></strong> :&ensp;<code>str</code></dt>
<dd>&nbsp;</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>list</code> of <code>tuples</code></dt>
<dd>&nbsp;</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def decode_region(s):
    &#34;&#34;&#34;

    Parameters
    ----------
    s : str

    Returns
    -------
    list of tuples

    &#34;&#34;&#34;
    regs = s.split(&#34;,&#34;)
    ret = []
    for r in regs:
        chr_interval = r.split(&#34;:&#34;)
        begin_end = chr_interval[1].split(&#34;-&#34;)
        ret.append((chr_interval[0], (decode_position(begin_end[0]), decode_position(begin_end[1]))))
    return ret</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.fit_normal"><code class="name flex">
<span>def <span class="ident">fit_normal</span></span>(<span>x, y)</span>
</code></dt>
<dd>
<section class="desc"><p>Fit Gaussian</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def fit_normal(x, y):
    &#34;&#34;&#34; Fit Gaussian
    &#34;&#34;&#34;
    if sum(y) == 0:
        _logger.debug(&#34;Problem with fit: all data points have zero value. Return zeros instead fit parameters!&#34;)
        return [0, 0, 0], None
    mean = sum(x * y) / sum(y)
    sigma = np.sqrt(sum(y * (x - mean) ** 2) / sum(y))
    _logger.debug(&#34;%f %f %d&#34; % (mean, sigma, len(x)))
    if sigma == 0:
        _logger.debug(&#34;Problem with fit: sigma equals zero. Return zeros instead fit parameters!&#34;)
        return [0, 0, 0], None
    area = sum(y[:-1] * (x[1:] - x[:-1]))
    try:
        popt, pcov = curve_fit(normal, x, y, p0=[area, mean, sigma])
        return popt, pcov
    except ValueError:
        _logger.warning(&#34;Problem with fit: optimal parameters not found. Using mean and std instead!&#34;)
        return [area, mean, sigma], None
    except RuntimeError:
        _logger.warning(&#34;Problem with fit: optimal parameters not found. Using mean and std instead!&#34;)
        return [area, mean, sigma], None</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.gc_at_compress"><code class="name flex">
<span>def <span class="ident">gc_at_compress</span></span>(<span>gc, at)</span>
</code></dt>
<dd>
<section class="desc"><p>Commpress GC/AT content of 100bins using fact that #at+#gc=100 in large majority of bins.
Transforms #at, #gc -&gt; #at, 100-#at-#gc</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>gc</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>Binned GC content (100bp bins).</dd>
<dt><strong><code>at</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>Binned AT content (100bp bins).</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>gcat</code></strong> :&ensp;<code>numpy.ndarray</code></dt>
<dd>Array contains compressed GC/AT content.</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def gc_at_compress(gc, at):
    &#34;&#34;&#34;
    Commpress GC/AT content of 100bins using fact that #at+#gc=100 in large majority of bins.
    Transforms #at, #gc -&gt; #at, 100-#at-#gc

    Parameters
    ----------
    gc : list of int
        Binned GC content (100bp bins).
    at : list of int
        Binned AT content (100bp bins).

    Returns
    -------
    gcat : numpy.ndarray
        Array contains compressed GC/AT content.

    &#34;&#34;&#34;
    cgc = np.array(gc)
    cat = 100 - np.array(at) - cgc
    return np.concatenate((cgc, cat)).astype(&#34;uint8&#34;)</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.gc_at_decompress"><code class="name flex">
<span>def <span class="ident">gc_at_decompress</span></span>(<span>gcat)</span>
</code></dt>
<dd>
<section class="desc"><p>Decompress GT/AC content - inverse function of gc_at_compress(gc, at).</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>gcat</code></strong> :&ensp;<code>numpy.ndarray</code></dt>
<dd>Array contains compressed GC/AT content.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>gc</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>Binned GC content (100bp bins).</dd>
<dt><strong><code>at</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>Binned AT content (100bp bins).</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def gc_at_decompress(gcat):
    &#34;&#34;&#34;
    Decompress GT/AC content - inverse function of gc_at_compress(gc, at).

    Parameters
    ----------
    gcat : numpy.ndarray
        Array contains compressed GC/AT content.

    Returns
    -------
    gc : list of int
        Binned GC content (100bp bins).
    at : list of int
        Binned AT content (100bp bins).

    &#34;&#34;&#34;
    return list(gcat[:gcat.size // 2]), list(100 - gcat[:gcat.size // 2] - gcat[gcat.size // 2:])</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.gcp_decompress"><code class="name flex">
<span>def <span class="ident">gcp_decompress</span></span>(<span>gcat, bin_ratio=1)</span>
</code></dt>
<dd>
<section class="desc"><p>Decompress GT/AC content and calculate GC percentage.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>gcat</code></strong> :&ensp;<code>numpy.ndarray</code></dt>
<dd>Array contains compressed GC/AT content.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>gcp</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>GC percentage.</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def gcp_decompress(gcat, bin_ratio=1):
    &#34;&#34;&#34;
    Decompress GT/AC content and calculate GC percentage.

    Parameters
    ----------
    gcat : numpy.ndarray
        Array contains compressed GC/AT content.

    Returns
    -------
    gcp : list of int
        GC percentage.

    &#34;&#34;&#34;
    gc, at = gcat[:gcat.size // 2].astype(&#34;float&#34;), (100 - gcat[:gcat.size // 2] - gcat[gcat.size // 2:]).astype(
        &#34;float&#34;)
    if bin_ratio == 1:
        return 100. * gc / (at + gc + 1e-10)
    n = len(gc)
    his_gc = np.concatenate(
        (np.array(gc), np.zeros(bin_ratio - n + (n // bin_ratio * bin_ratio)))).astype(&#34;float&#34;)
    his_gc.resize((len(his_gc) // bin_ratio, bin_ratio))
    his_gc = his_gc.sum(axis=1)
    his_at = np.concatenate(
        (np.array(at), np.zeros(bin_ratio - n + (n // bin_ratio * bin_ratio)))).astype(&#34;float&#34;)
    his_at.resize((len(his_at) // bin_ratio, bin_ratio))
    his_at = his_at.sum(axis=1)

    return 100. * his_gc / (his_at + his_gc + 1e-10)</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.likelihood_overlap"><code class="name flex">
<span>def <span class="ident">likelihood_overlap</span></span>(<span>lk1, lk2)</span>
</code></dt>
<dd>
<section class="desc"><p>Returns overlap area of two likelihood functions.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>lk1</code></strong> :&ensp;<code>numpy.ndarray</code></dt>
<dd>First likelihood function.</dd>
<dt><strong><code>lk2</code></strong> :&ensp;<code>numpy.ndarray</code></dt>
<dd>Second likelihood function.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>overlap</code></strong> :&ensp;<code>float</code></dt>
<dd>Overlap area.</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def likelihood_overlap(lk1, lk2):
    &#34;&#34;&#34;
    Returns overlap area of two likelihood functions.

    Parameters
    ----------
    lk1 : numpy.ndarray
        First likelihood function.
    lk2 : numpy.ndarray
        Second likelihood function.

    Returns
    -------
    overlap : float
        Overlap area.

    &#34;&#34;&#34;
    return np.sum(np.min((lk1, lk2), axis=0))</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.log_beta"><code class="name flex">
<span>def <span class="ident">log_beta</span></span>(<span>k, m, p, phased=False)</span>
</code></dt>
<dd>
<section class="desc"><p>Returns logarithm of likelihood beta function f(p) where 0 &lt;= p &lt;= 1.
Function is not normalized.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>k</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of haplotype 1 reads.</dd>
<dt><strong><code>m</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of haplotype 2 reads.</dd>
<dt><strong><code>p</code></strong> :&ensp;<code>float</code></dt>
<dd>Parameter.</dd>
<dt><strong><code>phased</code></strong> :&ensp;<code>bool</code></dt>
<dd>Likelihood will be symmetrised if not phased.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>f</code></strong> :&ensp;<code>float</code></dt>
<dd>Value od likelihood function,.</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def log_beta(k, m, p, phased=False):
    &#34;&#34;&#34;
    Returns logarithm of likelihood beta function f(p) where 0 &lt;= p &lt;= 1.
    Function is not normalized.

    Parameters
    ----------
    k : int
        Number of haplotype 1 reads.
    m : int
        Number of haplotype 2 reads.
    p : float
        Parameter.
    phased : bool
        Likelihood will be symmetrised if not phased.

    Returns
    -------
    f : float
        Value od likelihood function,.

    &#34;&#34;&#34;
    if k == m or phased:
        return np.log(1.0 * p ** k * (1. - p) ** m)
    else:
        return np.log(1.0 * p ** k * (1. - p) ** m + 1.0 * p ** m * (1. - p) ** k)</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.mask_compress"><code class="name flex">
<span>def <span class="ident">mask_compress</span></span>(<span>mask)</span>
</code></dt>
<dd>
<section class="desc"><p>Commpress strict mask P value intervals.</p>
<h2 id="argument">Argument</h2>
<p>mask - list of intervals [(start_1, end_1), (start_2, end_2),&hellip;]</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def mask_compress(mask):
    &#34;&#34;&#34; Commpress strict mask P value intervals.
    Argument:
        mask - list of intervals [(start_1, end_1), (start_2, end_2),...]
    &#34;&#34;&#34;
    pos = [p for interval in mask for p in interval]
    cmask = list(map(lambda i, j: i - j, pos, [0] + pos[:-1]))
    return np.array(cmask, dtype=&#34;uint32&#34;)</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.mask_decompress"><code class="name flex">
<span>def <span class="ident">mask_decompress</span></span>(<span>cmask)</span>
</code></dt>
<dd>
<section class="desc"><p>Decommpress strict mask P value intervals.</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def mask_decompress(cmask):
    &#34;&#34;&#34; Decommpress strict mask P value intervals.
    &#34;&#34;&#34;
    pos = []
    s = 0
    for i in cmask:
        s += i
        pos.append(s)
    return list(zip(pos[::2], pos[1::2]))</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.normal"><code class="name flex">
<span>def <span class="ident">normal</span></span>(<span>x, a, x0, sigma)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def normal(x, a, x0, sigma):
    return a * np.exp(-(x - x0) ** 2 / (2 * sigma ** 2))</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.normal_merge"><code class="name flex">
<span>def <span class="ident">normal_merge</span></span>(<span>m1, s1, m2, s2)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def normal_merge(m1, s1, m2, s2):
    if s1 == 0 and s2 == 0:
        return 0.5 * (m1 + m2), 0
    else:
        return (m1 * s2 * s2 + m2 * s1 * s1) / (s1 * s1 + s2 * s2), np.sqrt(s1 * s1 * s2 * s2 / (s1 * s1 + s2 * s2))</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.normal_overlap"><code class="name flex">
<span>def <span class="ident">normal_overlap</span></span>(<span>m1, s1, m2, s2)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def normal_overlap(m1, s1, m2, s2):
    if m1 &gt; m2:
        m1, m2 = m2, m1
        s1, s2 = s2, s1
    a = 1. / (2. * s1 ** 2) - 1. / (2. * s2 ** 2)
    b = m2 / (s2 ** 2) - m1 / (s1 ** 2)
    c = m1 ** 2 / (2 * s1 ** 2) - m2 ** 2 / (2 * s2 ** 2) - np.log(s2 / s1)
    print(np.roots([a, b, c]))
    roots = sorted(np.roots([a, b, c]))
    if len(roots) == 0:
        return 1
    elif len(roots) == 1:
        r = roots[0]
        return norm.cdf(r, m2, s2) + (1. - norm.cdf(r, m1, s1))
    elif s1 &lt; s2:
        return norm.cdf(roots[1], m2, s2) - norm.cdf(roots[0], m2, s2) + 1 - norm.cdf(roots[1], m1, s1) + norm.cdf(
            roots[0], m1, s1)
    return norm.cdf(roots[1], m1, s1) - norm.cdf(roots[0], m1, s1) + 1 - norm.cdf(roots[1], m2, s2) + norm.cdf(
        roots[0], m2, s2)</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.rd_compress"><code class="name flex">
<span>def <span class="ident">rd_compress</span></span>(<span>rd_p, rd_u, data_type='uint16')</span>
</code></dt>
<dd>
<section class="desc"><p>Compres RD signal</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def rd_compress(rd_p, rd_u, data_type=&#34;uint16&#34;):
    &#34;&#34;&#34; Compres RD signal
    &#34;&#34;&#34;
    return rd_p.astype(data_type), (rd_p - rd_u).astype(data_type)</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.rd_decompress"><code class="name flex">
<span>def <span class="ident">rd_decompress</span></span>(<span>crd_p, crd_u)</span>
</code></dt>
<dd>
<section class="desc"><p>Decommpress SNP information</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def rd_decompress(crd_p, crd_u):
    &#34;&#34;&#34; Decommpress SNP information
    &#34;&#34;&#34;
    return np.array(crd_p), np.array(crd_p) + np.array(crd_u)</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.snp_compress"><code class="name flex">
<span>def <span class="ident">snp_compress</span></span>(<span>pos, ref, alt, nref, nalt, gt, flag, qual)</span>
</code></dt>
<dd>
<section class="desc"><p>Commpress SNP information binary in four arrays:
snp_pos - diferences in positions
snp_counts - 32 bits: ref count (16 bits), alt count (16 bits)
snp_desc -
16 bits: ref (3 bits), alt (3 bits), gt (3 bits), flag (2 bits)
snp_qual - 8 bits</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def snp_compress(pos, ref, alt, nref, nalt, gt, flag, qual):
    &#34;&#34;&#34;
    Commpress SNP information binary in four arrays:
            snp_pos - diferences in positions
            snp_counts - 32 bits: ref count (16 bits), alt count (16 bits)
            snp_desc -  16 bits: ref (3 bits), alt (3 bits), gt (3 bits), flag (2 bits)
            snp_qual - 8 bits

    &#34;&#34;&#34;
    snp_pos = list(map(lambda i, j: i - j,
                       pos, [0] + pos[:-1]))
    snp_counts = list(map(lambda i, j: (i &lt;&lt; 16) + j,
                          nref, nalt))
    snp_desc = list(map(lambda r, a, g, f: ((((((bp_to_binary[r] &lt;&lt; 3) + bp_to_binary[a]) &lt;&lt; 3) + g) &lt;&lt; 2) + f),
                        ref, alt, gt, flag))
    snp_qual = qual
    return np.array(snp_pos, dtype=&#34;uint32&#34;), np.array(snp_desc, dtype=&#34;uint16&#34;), np.array(snp_counts,
                                                                                           dtype=&#34;uint32&#34;), np.array(
        snp_qual, dtype=&#34;uint8&#34;)</code></pre>
</details>
</dd>
<dt id="cnvpytor.utils.snp_decompress"><code class="name flex">
<span>def <span class="ident">snp_decompress</span></span>(<span>snp_pos, snp_desc, snp_counts, snp_qual)</span>
</code></dt>
<dd>
<section class="desc"><p>Decommpress SNP information</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def snp_decompress(snp_pos, snp_desc, snp_counts, snp_qual):
    &#34;&#34;&#34;
    Decommpress SNP information

    &#34;&#34;&#34;
    pos = []
    s = 0
    for i in snp_pos:
        s += i
        pos.append(s)
    ref = []
    alt = []
    gt = []
    flag = []
    for i in snp_desc:
        ref.append(binary_to_bp[i &gt;&gt; 8])
        alt.append(binary_to_bp[(i &gt;&gt; 5) &amp; int(&#39;111&#39;, 2)])
        gt.append((i &gt;&gt; 2) &amp; int(&#39;111&#39;, 2))
        flag.append(i &amp; int(&#39;11&#39;, 2))
    nref = []
    nalt = []
    for i in snp_counts:
        nref.append(i &gt;&gt; 16)
        nalt.append(i &amp; int(&#39;ffff&#39;, 16))
    qual = list(snp_qual)

    return pos, ref, alt, nref, nalt, gt, flag, qual</code></pre>
</details>
</dd>
</dl>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="cnvpytor.utils.PromptCompleter"><code class="flex name class">
<span>class <span class="ident">PromptCompleter</span></span>
<span>(</span><span>command_tree)</span>
</code></dt>
<dd>
<section class="desc"><p>Initialize completer class for GNU readline tab comleter.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>command_tree</code></strong> :&ensp;<code>dictionary</code></dt>
<dd>Example: command_tree = {"comannd1":None,"command2":{"subcommand2.1":None,subcommand2.1:None}}</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">class PromptCompleter:
    def __init__(self, command_tree):
        &#34;&#34;&#34;
        Initialize completer class for GNU readline tab comleter.

        Parameters
        ----------
        command_tree : dictionary
            Example: command_tree = {&#34;comannd1&#34;:None,&#34;command2&#34;:{&#34;subcommand2.1&#34;:None,subcommand2.1:None}}

        &#34;&#34;&#34;
        self.command_tree = command_tree

    def _traverse(self, tokens, tree):
        if tree is None or len(tokens) == 0:
            return []
        if len(tokens) == 1:
            return [x + &#39; &#39; for x in tree if x.startswith(tokens[0])]
        else:
            if tokens[0] in tree.keys():
                return self._traverse(tokens[1:], tree[tokens[0]])
        return []

    def complete(self, text, state):
        try:
            tokens = readline.get_line_buffer().split()
            if not tokens or readline.get_line_buffer()[-1] == &#39; &#39;:
                tokens.append(&#34;&#34;)
            results = self._traverse(tokens, self.command_tree) + [None]
            return results[state]
        except Exception as e:
            print(e)</code></pre>
</details>
<h3>Methods</h3>
<dl>
<dt id="cnvpytor.utils.PromptCompleter.complete"><code class="name flex">
<span>def <span class="ident">complete</span></span>(<span>self, text, state)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def complete(self, text, state):
    try:
        tokens = readline.get_line_buffer().split()
        if not tokens or readline.get_line_buffer()[-1] == &#39; &#39;:
            tokens.append(&#34;&#34;)
        results = self._traverse(tokens, self.command_tree) + [None]
        return results[state]
    except Exception as e:
        print(e)</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="cnvpytor" href="index.html">cnvpytor</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="cnvpytor.utils.beta" href="#cnvpytor.utils.beta">beta</a></code></li>
<li><code><a title="cnvpytor.utils.binsize_type" href="#cnvpytor.utils.binsize_type">binsize_type</a></code></li>
<li><code><a title="cnvpytor.utils.calculate_gc_correction" href="#cnvpytor.utils.calculate_gc_correction">calculate_gc_correction</a></code></li>
<li><code><a title="cnvpytor.utils.decode_position" href="#cnvpytor.utils.decode_position">decode_position</a></code></li>
<li><code><a title="cnvpytor.utils.decode_region" href="#cnvpytor.utils.decode_region">decode_region</a></code></li>
<li><code><a title="cnvpytor.utils.fit_normal" href="#cnvpytor.utils.fit_normal">fit_normal</a></code></li>
<li><code><a title="cnvpytor.utils.gc_at_compress" href="#cnvpytor.utils.gc_at_compress">gc_at_compress</a></code></li>
<li><code><a title="cnvpytor.utils.gc_at_decompress" href="#cnvpytor.utils.gc_at_decompress">gc_at_decompress</a></code></li>
<li><code><a title="cnvpytor.utils.gcp_decompress" href="#cnvpytor.utils.gcp_decompress">gcp_decompress</a></code></li>
<li><code><a title="cnvpytor.utils.likelihood_overlap" href="#cnvpytor.utils.likelihood_overlap">likelihood_overlap</a></code></li>
<li><code><a title="cnvpytor.utils.log_beta" href="#cnvpytor.utils.log_beta">log_beta</a></code></li>
<li><code><a title="cnvpytor.utils.mask_compress" href="#cnvpytor.utils.mask_compress">mask_compress</a></code></li>
<li><code><a title="cnvpytor.utils.mask_decompress" href="#cnvpytor.utils.mask_decompress">mask_decompress</a></code></li>
<li><code><a title="cnvpytor.utils.normal" href="#cnvpytor.utils.normal">normal</a></code></li>
<li><code><a title="cnvpytor.utils.normal_merge" href="#cnvpytor.utils.normal_merge">normal_merge</a></code></li>
<li><code><a title="cnvpytor.utils.normal_overlap" href="#cnvpytor.utils.normal_overlap">normal_overlap</a></code></li>
<li><code><a title="cnvpytor.utils.rd_compress" href="#cnvpytor.utils.rd_compress">rd_compress</a></code></li>
<li><code><a title="cnvpytor.utils.rd_decompress" href="#cnvpytor.utils.rd_decompress">rd_decompress</a></code></li>
<li><code><a title="cnvpytor.utils.snp_compress" href="#cnvpytor.utils.snp_compress">snp_compress</a></code></li>
<li><code><a title="cnvpytor.utils.snp_decompress" href="#cnvpytor.utils.snp_decompress">snp_decompress</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="cnvpytor.utils.PromptCompleter" href="#cnvpytor.utils.PromptCompleter">PromptCompleter</a></code></h4>
<ul class="">
<li><code><a title="cnvpytor.utils.PromptCompleter.complete" href="#cnvpytor.utils.PromptCompleter.complete">complete</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.6.3</a>.</p>
</footer>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script>hljs.initHighlightingOnLoad()</script>
</body>
</html>