<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.6.3" />
<title>cnvpytor.vcf API documentation</title>
<meta name="description" content="cnvpytor.vcf â€¦" />
<link href='https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.0/normalize.min.css' rel='stylesheet'>
<link href='https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/8.0.0/sanitize.min.css' rel='stylesheet'>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet">
<style>.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{font-weight:bold}#index h4 + ul{margin-bottom:.6em}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase;cursor:pointer}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>cnvpytor.vcf</code></h1>
</header>
<section id="section-intro">
<p>cnvpytor.vcf</p>
<p>class Vcf: reads SNP data from VCF file</p>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">&#34;&#34;&#34; cnvpytor.vcf

class Vcf: reads SNP data from VCF file

&#34;&#34;&#34;
import pysam
import logging
import numpy as np
from .utils import int1, gt_from_list, gt_from_str

_logger = logging.getLogger(&#34;cnvpytor.vcf&#34;)


class Vcf:

    def __init__(self, filename):
        &#34;&#34;&#34;
        Opens VCF file, reads chromosome/sample names

        Parameters
        ----------
        filename : str
            Name of the VCF file.

        &#34;&#34;&#34;
        self.filename = filename
        try:
            self.file = pysam.VariantFile(filename)
        except IOError:
            _logger.error(&#34;Problem opening file &#39;%s&#39;!&#34; % filename)
            exit(0)
        except ValueError:
            _logger.error(&#34;Problem opening file &#39;%s&#39;!&#34; % filename)
            exit(0)

        self.chrs = []
        self.samples = []
        self.lengths = {}
        if self.file:
            _logger.info(&#34;File: &#34; + filename + &#34; successfully open&#34;)
            self.chrs = list(self.file.header.contigs)
            self.samples = list(self.file.header.samples)
            for c in self.chrs:
                self.lengths[c] = self.file.header.contigs[c].length
            _logger.debug(&#34;Header contigs: %s&#34; % &#34;, &#34;.join(self.chrs))
            _logger.debug(&#34;Header samples: %s&#34; % &#34;, &#34;.join(self.samples))

    def get_chromosomes(self):
        &#34;&#34;&#34;
        Get chromosome names.

        Returns
        -------
        chrs : list of str
            List of chromosome names from VCF header

        &#34;&#34;&#34;
        return self.chrs

    def get_samples(self):
        &#34;&#34;&#34;
        Get sample names.

        Returns
        -------
        samples : list of str
            List of sample names from VCF header

        &#34;&#34;&#34;
        return self.samples

    def read_chromosome_snp(self, chr_name, sample=&#39;&#39;, ad_tag=&#39;AD&#39;, gt_tag=&#39;GT&#39;, filter=True):
        &#34;&#34;&#34;
        Read SNP/indel data for given chromosome and sample.

        Parameters
        ----------
        chr_name : str
            Name of the chromosome.
        sample : str
            Name of the sample (first one if empty - default)
        ad_tag : str
            AD tag in VCF file (default: &#39;AD&#39;)
        gt_tag : str
            GT tag in VCF file (default: &#39;GT&#39;)
        filter : bool
            If True it will read only variants with PASS filter, otherwise all

        Returns
        -------
        pos : list of int
            List of SNP positions.
        ref : list of str
            List of SNP reference base (A, T, G, C or .).
        alt : list of str
            List of SNP alternative base (A, T, G, C or .).
        nref : list of int
            Count of reads contains reference SNP.
        nalt : list of int
            Count of reads contains alternative SNP.
        gt : list of int
            List of genotypes (0 - &#34;0/0&#34;, 1 - &#34;0/1&#34;, 3- &#34;1/1&#34;, 4 - &#34;0|0&#34; , 5 - &#34;0|1&#34;, 6 - &#34;1|0&#34;, 7 - &#34;1|1&#34;).
        flag : list of int
            Binary flag: first bit 1 - SNP exists in database, second bit 1 - SNP in P region of strict mask.
        qual : list of int
            SNP quality (scale 0 - 255).

        &#34;&#34;&#34;
        if sample == &#39;&#39;:
            sample = self.samples[0]
        pos = []
        ref = []
        alt = []
        nref = []
        nalt = []
        gt = []
        flag = []
        qual = []
        filter_stat = {}
        alphabet = [&#39;A&#39;, &#39;T&#39;, &#39;G&#39;, &#39;C&#39;, &#39;.&#39;]
        try:
            for rec in self.file.fetch(chr_name):
                if len(rec.filter.keys()) == 0:
                    if &#34;No filter (.)&#34; in filter_stat:
                        filter_stat[&#34;No filter (.)&#34;] += 1
                    else:
                        filter_stat[&#34;No filter (.)&#34;] = 1
                for f in rec.filter.keys():
                    if f in filter_stat:
                        filter_stat[f] += 1
                    else:
                        filter_stat[f] = 1
                if (&#34;PASS&#34; in rec.filter.keys() or not filter) and rec.alts and len(rec.alts) &gt;= 1 and (
                        gt_tag in rec.samples[sample].keys()) and (
                        ad_tag in rec.samples[sample].keys()) and len(rec.samples[sample][gt_tag]) &gt; 1 and len(
                    rec.samples[sample][ad_tag]) &gt; 1:
                    if (len(rec.samples[sample][ad_tag]) &gt; 1) and (rec.ref in alphabet) and (rec.alts[0] in alphabet):
                        pos.append(rec.pos)
                        ref.append(rec.ref)
                        alt.append(rec.alts[0])
                        flag.append(2)  # Assign P region as default (-mask_snps updates this flag)
                        if rec.qual == &#34;.&#34; or rec.qual is None:
                            qual.append(0)
                        else:
                            qual.append(int(rec.qual / 10))  # divide QUAL by factor 10 and truncate to one byte
                        if qual[-1] &gt; 255:
                            qual[-1] = 255
                        try:
                            nref.append(int(rec.samples[sample][ad_tag][0]))
                            nalt.append(int(rec.samples[sample][ad_tag][1]))
                        except ValueError:
                            nref.append(0)
                            nalt.append(0)
                        try:
                            UNICODE_EXISTS = bool(type(unicode))
                        except NameError:
                            unicode = str
                        if isinstance(rec.samples[sample][gt_tag], str) or isinstance(rec.samples[sample][gt_tag],
                                                                                      unicode):
                            gt.append(gt_from_str(rec.samples[sample][gt_tag]))
                        else:
                            gt.append(gt_from_list(rec.samples[sample][gt_tag], rec.samples[sample].phased))
        except ValueError:
            _logger.error(&#34;Variant file reading problem. Probably index file is missing or corrupted.&#34;)
            exit(0)
        _logger.debug(&#34;Chromosome &#39;%s&#39; read. Number of variants to store: %d.&#34; % (chr_name, len(pos)))
        _logger.debug(&#34;Variants filter field statistics:&#34;)
        for f in filter_stat:
            _logger.debug(&#34; * &#39;%s&#39; : %d&#34; % (f, filter_stat[f]))
        return pos, ref, alt, nref, nalt, gt, flag, qual

    def read_chromosome_snp_no_counts(self, chr_name, sample=&#39;&#39;, gt_tag=&#39;GT&#39;, filter=True):
        &#34;&#34;&#34;
        Read SNP/indel data without counts (AD tag) for given chromosome and sample.

        Parameters
        ----------
        chr_name : str
            Name of the chromosome.
        sample : str
            Name of the sample (first one if empty - default)
        gt_tag : str
            GT tag in VCF file (default: &#39;GT&#39;)
        filter : bool
            If True it will read only variants with PASS filter, otherwise all

        Returns
        -------
        pos : list of int
            List of SNP positions.
        ref : list of str
            List of SNP reference base (A, T, G, C or .).
        alt : list of str
            List of SNP alternative base (A, T, G, C or .).
        gt : list of int
            List of genotypes (0 - &#34;0/0&#34;, 1 - &#34;0/1&#34;, 3- &#34;1/1&#34;, 4 - &#34;0|0&#34; , 5 - &#34;0|1&#34;, 6 - &#34;1|0&#34;, 7 - &#34;1|1&#34;).
        flag : list of int
            Binary flag: first bit 1 - SNP exists in database, second bit 1 - SNP in P region of strict mask.
        qual : list of int
            SNP quality (scale 0 - 255).

        &#34;&#34;&#34;
        if sample == &#39;&#39;:
            sample = self.samples[0]
        pos = []
        ref = []
        alt = []
        gt = []
        flag = []
        qual = []
        alphabet = [&#39;A&#39;, &#39;T&#39;, &#39;G&#39;, &#39;C&#39;, &#39;.&#39;]
        try:
            for rec in self.file.fetch(chr_name):
                if (&#34;PASS&#34; in rec.filter.keys() or not filter) and rec.alts and len(rec.alts) &gt;= 1 and (
                        gt_tag in rec.samples[sample].keys()) and len(rec.samples[sample][gt_tag]) &gt; 1:
                    if (rec.ref in alphabet) and (rec.alts[0] in alphabet):
                        pos.append(rec.pos)
                        ref.append(rec.ref)
                        alt.append(rec.alts[0])
                        flag.append(2)  # Assign P region as default (-mask_snps updates this flag)
                        if rec.qual == &#34;.&#34; or rec.qual is None:
                            qual.append(0)
                        else:
                            qual.append(int(rec.qual / 10))  # divide QUAL by factor 10 and truncate to one byte
                        if qual[-1] &gt; 255:
                            qual[-1] = 255
                        try:
                            UNICODE_EXISTS = bool(type(unicode))
                        except NameError:
                            unicode = str
                        if isinstance(rec.samples[sample][gt_tag], str) or isinstance(rec.samples[sample][gt_tag],
                                                                                      unicode):
                            gt.append(gt_from_str(rec.samples[sample][gt_tag]))
                        else:
                            gt.append(gt_from_list(rec.samples[sample][gt_tag], rec.samples[sample].phased))

        except ValueError:
            _logger.error(&#34;Variant file reading problem. Probably index file is missing or corrupted.&#34;)
            exit(0)
        return pos, ref, alt, gt, flag, qual

    def read_chromosome_rd(self, chr_name, sample=&#39;&#39;, ad_tag=&#39;AD&#39;, dp_tag=&#39;DP&#39;, filter=False):
        &#34;&#34;&#34;
        Read RD data for given chromosome and sample.

        Parameters
        ----------
        chr_name : str
            Name of the chromosome.
        sample : str
            Name of the sample (first one if empty - default)
        ad_tag : str
            AD tag in VCF file (default: &#39;AD&#39;)
        dp_tag : str
            DP tag in VCF file (default: &#39;DP&#39;)
        filter : bool
            If True it will read only variants with PASS filter, otherwise all

        Returns
        -------
        rd_p : list of int
            binned RD signal from DP tag (100 bins)
        rd_u : list of int
            binned RD signal from AD tag (100 bins)

        &#34;&#34;&#34;
        if sample == &#39;&#39;:
            sample = self.samples[0]
        pos = []
        rd_ad = []
        rd_dp = []
        try:
            for rec in self.file.fetch(chr_name):
                if (&#34;PASS&#34; in rec.filter.keys() or not filter) and (ad_tag in rec.samples[sample].keys()) and len(
                        rec.samples[sample][ad_tag]) &gt; 1 and (dp_tag in rec.samples[sample].keys()) and (
                        rec.samples[sample][dp_tag] is not None):
                    try:
                        rd1 = sum(map(int, rec.samples[sample][ad_tag]))
                        rd2 = int(rec.samples[sample][dp_tag])
                        pos.append(rec.pos)
                        rd_ad.append(rd1)
                        rd_dp.append(rd2)
                    except ValueError:
                        pass
        except ValueError:
            _logger.error(&#34;Variant file reading problem. Probably index file is missing or corrupted.&#34;)
            exit(0)
        n = 0
        if len(pos) == 0:
            return None, None
        n = pos[-1] + 1
        if self.lengths[chr_name] is not None:
            n = self.lengths[chr_name]
        n = n // 100 + 1
        rd_p = np.zeros(n)
        rd_u = np.zeros(n)
        count = np.zeros(n)
        for p, rd1, rd2 in zip(pos, rd_dp, rd_ad):
            cgb = (p - 1) // 10000 * 100
            rd_p[cgb] += rd1
            rd_u[cgb] += rd2
            count[cgb] += 1
        for i in range(n // 100 + 1):
            if (100 * i) &lt; n:
                if count[100 * i] != 0:
                    rd_p[i * 100:(i + 1) * 100] = rd_p[i * 100] / count[i * 100]
                    rd_u[i * 100:(i + 1) * 100] = rd_u[i * 100] / count[i * 100]
                else:
                    rd_p[i * 100:(i + 1) * 100] = np.nan
                    rd_u[i * 100:(i + 1) * 100] = np.nan
        return rd_p, rd_u

    def read_all_snp(self, callback, sample=&#39;&#39;, ad_tag=&#39;AD&#39;, gt_tag=&#39;GT&#39;, filter=True):
        &#34;&#34;&#34;
        Read SNP/indel data for given sample name.

        Parameters
        ----------
        callback : callable
            Function to call after read a chromosome:
            callback(chrom, pos, ref, alt, nref, nalt, gt, flag, qual)
        sample : str
            Name of the sample (first one if empty - default)
        ad_tag : str
            AD tag in VCF file (default: &#39;AD&#39;)
        gt_tag : str
            GT tag in VCF file (default: &#39;GT&#39;)
        filter : bool
            If True it will read only variants with PASS filter, otherwise all

        Returns
        -------
        count : int
            Number of read chromosomes

        &#34;&#34;&#34;
        if sample == &#39;&#39;:
            sample = self.samples[0]
        pos = []
        ref = []
        alt = []
        nref = []
        nalt = []
        gt = []
        flag = []
        qual = []
        filter_stat = {}
        last_chrom = None
        count = 0
        alphabet = [&#39;A&#39;, &#39;T&#39;, &#39;G&#39;, &#39;C&#39;, &#39;.&#39;]

        try:
            for rec in self.file.fetch():
                if last_chrom is None:
                    last_chrom = rec.chrom
                if last_chrom != rec.chrom:
                    _logger.info(&#34;Chromosome &#39;%s&#39; read. Number of variants to store: %d.&#34; % (last_chrom, len(pos)))
                    _logger.debug(&#34;Variants filter field statistics:&#34;)
                    for f in filter_stat:
                        _logger.debug(&#34;    * &#39;%s&#39; : %d&#34; % (f, filter_stat[f]))
                    callback(last_chrom, pos, ref, alt, nref, nalt, gt, flag, qual)
                    pos = []
                    ref = []
                    alt = []
                    nref = []
                    nalt = []
                    gt = []
                    flag = []
                    qual = []
                    filter_stat = {}
                    count += 1
                if len(rec.filter.keys()) == 0:
                    if &#34;No filter (.)&#34; in filter_stat:
                        filter_stat[&#34;No filter (.)&#34;] += 1
                    else:
                        filter_stat[&#34;No filter (.)&#34;] = 1
                for f in rec.filter.keys():
                    if f in filter_stat:
                        filter_stat[f] += 1
                    else:
                        filter_stat[f] = 1

                if (&#34;PASS&#34; in rec.filter.keys() or not filter) and rec.alts and len(rec.alts) &gt;= 1 and (
                        gt_tag in rec.samples[sample].keys()) and (
                        ad_tag in rec.samples[sample].keys()) and len(rec.samples[sample][gt_tag]) &gt; 1 and len(
                    rec.samples[sample][ad_tag]) &gt; 1:
                    if (len(rec.samples[sample][ad_tag]) &gt; 1) and (rec.ref in alphabet) and (
                            rec.alts[0] in alphabet) and (rec.samples[sample][gt_tag][0] is not None) and (
                            rec.samples[sample][gt_tag][1] is not None):
                        pos.append(rec.pos)
                        ref.append(rec.ref)
                        alt.append(rec.alts[0])
                        flag.append(2)  # Assign P region as default (-mask_snps updates this flag)
                        if rec.qual == &#34;.&#34; or rec.qual is None:
                            qual.append(0)
                        else:
                            qual.append(int(rec.qual / 10))  # divide QUAL by factor 10 and truncate to one byte
                        if qual[-1] &gt; 255:
                            qual[-1] = 255
                        try:
                            nref.append(int(rec.samples[sample][ad_tag][0]))
                            nalt.append(int(rec.samples[sample][ad_tag][1]))
                        except ValueError:
                            nref.append(0)
                            nalt.append(0)
                        try:
                            UNICODE_EXISTS = bool(type(unicode))
                        except NameError:
                            unicode = str
                        if isinstance(rec.samples[sample][gt_tag], str) or isinstance(rec.samples[sample][gt_tag],
                                                                                      unicode):
                            gt.append(gt_from_str(rec.samples[sample][gt_tag]))
                        else:
                            gt.append(gt_from_list(rec.samples[sample][gt_tag], rec.samples[sample].phased))

                last_chrom = rec.chrom
            _logger.debug(&#34;Chromosome &#39;%s&#39; read. Number of variants to store: %d.&#34; % (last_chrom, len(pos)))
            _logger.debug(&#34;Variants filter field statistics:&#34;)
            for f in filter_stat:
                _logger.debug(&#34; * &#39;%s&#39; : %d&#34; % (f, filter_stat[f]))
            callback(last_chrom, pos, ref, alt, nref, nalt, gt, flag, qual)
            count += 1
            return count
        except ValueError:
            _logger.error(&#34;Variant file reading problem.&#34;)
            exit(0)

    def read_all_snp_no_counts(self, callback, sample=&#39;&#39;, gt_tag=&#39;GT&#39;, filter=True):
        &#34;&#34;&#34;
        Read SNP/indel data without counts (AD tag) for given sample name.

        Parameters
        ----------
        callback : callable
            Function to call after read a chromosome:
            callback(chrom, pos, ref, alt, gt, flag, qual)
        sample : str
            Name of the sample (first one if empty - default)
        gt_tag : str
            GT tag in VCF file (default: &#39;GT&#39;)
        filter : bool
            If True it will read only variants with PASS filter, otherwise all

        Returns
        -------
        count : int
            Number of read chromosomes

        &#34;&#34;&#34;
        if sample == &#39;&#39;:
            sample = self.samples[0]
        pos = []
        ref = []
        alt = []
        gt = []
        flag = []
        qual = []
        last_chrom = None
        count = 0
        alphabet = [&#39;A&#39;, &#39;T&#39;, &#39;G&#39;, &#39;C&#39;, &#39;.&#39;]

        try:
            for rec in self.file.fetch():
                if last_chrom is None:
                    last_chrom = rec.chrom
                if last_chrom != rec.chrom:
                    callback(last_chrom, pos, ref, alt, gt, flag, qual)
                    pos = []
                    ref = []
                    alt = []
                    gt = []
                    flag = []
                    qual = []
                    count += 1

                if (&#34;PASS&#34; in rec.filter.keys() or not filter) and rec.alts and len(rec.alts) &gt;= 1 and (
                        gt_tag in rec.samples[sample].keys()) and len(rec.samples[sample][gt_tag]) &gt; 1:
                    if (rec.ref in alphabet) and (rec.alts[0] in alphabet) and (
                            rec.samples[sample][gt_tag][0] is not None) and (
                            rec.samples[sample][gt_tag][1] is not None):
                        pos.append(rec.pos)
                        ref.append(rec.ref)
                        alt.append(rec.alts[0])
                        flag.append(2)  # Assign P region as default (-mask_snps updates this flag)
                        if rec.qual == &#34;.&#34; or rec.qual is None:
                            qual.append(0)
                        else:
                            qual.append(int(rec.qual / 10))  # divide QUAL by factor 10 and truncate to one byte
                        if qual[-1] &gt; 255:
                            qual[-1] = 255
                        try:
                            UNICODE_EXISTS = bool(type(unicode))
                        except NameError:
                            unicode = str
                        if isinstance(rec.samples[sample][gt_tag], str) or isinstance(rec.samples[sample][gt_tag],
                                                                                      unicode):
                            gt.append(gt_from_str(rec.samples[sample][gt_tag]))
                        else:
                            gt.append(gt_from_list(rec.samples[sample][gt_tag], rec.samples[sample].phased))

                last_chrom = rec.chrom
            if len(pos) &gt; 0:
                callback(last_chrom, pos, ref, alt, gt, flag, qual)
                count += 1
            return count
        except ValueError:
            _logger.error(&#34;Variant file reading problem.&#34;)
            exit(0)

    def read_all_snp_positions(self, callback, filter=True):
        &#34;&#34;&#34;
        Read SNP/indel data positions.

        Parameters
        ----------
        callback : callable
            Function to call after read a chromosome:
            callback(chrom, pos, ref, alt)
        sample : str
            Name of the sample (first one if empty - default)
        filter : bool
            If True it will read only variants with PASS filter, otherwise all

        Returns
        -------
        count : int
            Number of read chromosomes

        &#34;&#34;&#34;
        pos = []
        ref = []
        alt = []
        last_chrom = None
        count = 0
        alphabet = [&#39;A&#39;, &#39;T&#39;, &#39;G&#39;, &#39;C&#39;, &#39;.&#39;]

        try:
            for rec in self.file.fetch():
                if last_chrom is None:
                    last_chrom = rec.chrom
                if last_chrom != rec.chrom:
                    callback(last_chrom, pos, ref, alt)
                    pos = []
                    ref = []
                    alt = []
                    count += 1
                if (&#34;PASS&#34; in rec.filter.keys() or not filter) and rec.alts and len(rec.alts) &gt;= 1:
                    if (rec.ref in alphabet) and (rec.alts[0] in alphabet):
                        pos.append(rec.pos)
                        ref.append(rec.ref)
                        alt.append(rec.alts[0])
                last_chrom = rec.chrom
            if len(pos) &gt; 0:
                callback(last_chrom, pos, ref, alt)
                count += 1
            return count
        except ValueError:
            _logger.error(&#34;Variant file reading problem.&#34;)
            exit(0)

    def read_all_rd(self, callback, sample=&#39;&#39;, ad_tag=&#39;AD&#39;, dp_tag=&#39;DP&#39;, filter=False):
        &#34;&#34;&#34;
        Read RD data for given chromosome and sample.

        Parameters
        ----------
        chr_name : str
            Name of the chromosome.
        sample : str
            Name of the sample (first one if empty - default)
        ad_tag : str
            AD tag in VCF file (default: &#39;AD&#39;)
        dp_tag : str
            DP tag in VCF file (default: &#39;DP&#39;)
        filter : bool
            If True it will read only variants with PASS filter, otherwise all

        Returns
        -------
        rd_p : list of int
            binned RD signal from DP tag (100 bins)
        rd_u : list of int
            binned RD signal from AD tag (100 bins)

        &#34;&#34;&#34;
        if sample == &#39;&#39;:
            sample = self.samples[0]
        pos = []
        rd_ad = []
        rd_dp = []
        last_chrom = None
        chr_count = 0

        try:
            for rec in self.file.fetch():
                if last_chrom is None:
                    last_chrom = rec.chrom
                if last_chrom != rec.chrom and len(pos) &gt; 0:
                    n = pos[-1] + 1
                    if self.lengths[last_chrom] is not None:
                        n = self.lengths[last_chrom]
                    n = n // 100 + 1
                    rd_p = np.zeros(n)
                    rd_u = np.zeros(n)
                    count = np.zeros(n)
                    for p, rd1, rd2 in zip(pos, rd_dp, rd_ad):
                        cgb = (p - 1) // 10000 * 100
                        rd_p[cgb] += rd1
                        rd_u[cgb] += rd2
                        count[cgb] += 1
                    for i in range(n // 100 + 1):
                        if (100 * i) &lt; n:
                            if count[100 * i] != 0:
                                rd_p[i * 100:(i + 1) * 100] = rd_p[i * 100] / count[i * 100]
                                rd_u[i * 100:(i + 1) * 100] = rd_u[i * 100] / count[i * 100]
                            else:
                                rd_p[i * 100:(i + 1) * 100] = np.nan
                                rd_u[i * 100:(i + 1) * 100] = np.nan
                    callback(last_chrom, rd_p, rd_u)
                    pos = []
                    rd_ad = []
                    rd_dp = []
                    chr_count += 1

                if (&#34;PASS&#34; in rec.filter.keys() or not filter) and (ad_tag in rec.samples[sample].keys()) and len(
                        rec.samples[sample][ad_tag]) &gt; 1 and (
                        dp_tag in rec.samples[sample].keys()) and (rec.samples[sample][dp_tag] is not None):
                    try:
                        rd1 = sum(map(int, rec.samples[sample][ad_tag]))
                        rd2 = int(rec.samples[sample][dp_tag])
                        pos.append(rec.pos)
                        rd_ad.append(rd1)
                        rd_dp.append(rd2)
                    except ValueError:
                        pass
                last_chrom = rec.chrom
            if len(pos) &gt; 0:
                n = pos[-1] + 1
                if self.lengths[last_chrom] is not None:
                    n = self.lengths[last_chrom]
                n = n // 100 + 1
                rd_p = np.zeros(n)
                rd_u = np.zeros(n)
                count = np.zeros(n)

                for p, rd1, rd2 in zip(pos, rd_dp, rd_ad):
                    cgb = (p - 1) // 10000 * 100
                    rd_p[cgb] += rd1
                    rd_u[cgb] += rd2
                    count[cgb] += 1
                for i in range(n // 100 + 1):
                    if (100 * i) &lt; n:
                        if count[100 * i] != 0:
                            rd_p[i * 100:(i + 1) * 100] = rd_p[i * 100] / count[i * 100]
                            rd_u[i * 100:(i + 1) * 100] = rd_u[i * 100] / count[i * 100]
                        else:
                            rd_p[i * 100:(i + 1) * 100] = np.nan
                            rd_u[i * 100:(i + 1) * 100] = np.nan
                callback(last_chrom, rd_p, rd_u)
                chr_count += 1
            return chr_count
        except ValueError:
            _logger.error(&#34;Variant file reading problem. Probably index file is missing or corrupted.&#34;)
            exit(0)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="cnvpytor.vcf.Vcf"><code class="flex name class">
<span>class <span class="ident">Vcf</span></span>
<span>(</span><span>filename)</span>
</code></dt>
<dd>
<section class="desc"><p>Opens VCF file, reads chromosome/sample names</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>filename</code></strong> :&ensp;<code>str</code></dt>
<dd>Name of the VCF file.</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">class Vcf: reads SNP data from VCF file</code></pre>
</details>
<h3>Methods</h3>
<dl>
<dt id="cnvpytor.vcf.Vcf.get_chromosomes"><code class="name flex">
<span>def <span class="ident">get_chromosomes</span></span>(<span>self)</span>
</code></dt>
<dd>
<section class="desc"><p>Get chromosome names.</p>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>chrs</code></strong> :&ensp;<code>list</code> of <code>str</code></dt>
<dd>List of chromosome names from VCF header</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def get_chromosomes(self):
    &#34;&#34;&#34;
    Get chromosome names.

    Returns
    -------
    chrs : list of str
        List of chromosome names from VCF header

    &#34;&#34;&#34;
    return self.chrs</code></pre>
</details>
</dd>
<dt id="cnvpytor.vcf.Vcf.get_samples"><code class="name flex">
<span>def <span class="ident">get_samples</span></span>(<span>self)</span>
</code></dt>
<dd>
<section class="desc"><p>Get sample names.</p>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>samples</code></strong> :&ensp;<code>list</code> of <code>str</code></dt>
<dd>List of sample names from VCF header</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def get_samples(self):
    &#34;&#34;&#34;
    Get sample names.

    Returns
    -------
    samples : list of str
        List of sample names from VCF header

    &#34;&#34;&#34;
    return self.samples</code></pre>
</details>
</dd>
<dt id="cnvpytor.vcf.Vcf.read_all_rd"><code class="name flex">
<span>def <span class="ident">read_all_rd</span></span>(<span>self, callback, sample='', ad_tag='AD', dp_tag='DP', filter=False)</span>
</code></dt>
<dd>
<section class="desc"><p>Read RD data for given chromosome and sample.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>chr_name</code></strong> :&ensp;<code>str</code></dt>
<dd>Name of the chromosome.</dd>
<dt><strong><code>sample</code></strong> :&ensp;<code>str</code></dt>
<dd>Name of the sample (first one if empty - default)</dd>
<dt><strong><code>ad_tag</code></strong> :&ensp;<code>str</code></dt>
<dd>AD tag in VCF file (default: 'AD')</dd>
<dt><strong><code>dp_tag</code></strong> :&ensp;<code>str</code></dt>
<dd>DP tag in VCF file (default: 'DP')</dd>
<dt><strong><code>filter</code></strong> :&ensp;<code>bool</code></dt>
<dd>If True it will read only variants with PASS filter, otherwise all</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>rd_p</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>binned RD signal from DP tag (100 bins)</dd>
<dt><strong><code>rd_u</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>binned RD signal from AD tag (100 bins)</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def read_all_rd(self, callback, sample=&#39;&#39;, ad_tag=&#39;AD&#39;, dp_tag=&#39;DP&#39;, filter=False):
    &#34;&#34;&#34;
    Read RD data for given chromosome and sample.

    Parameters
    ----------
    chr_name : str
        Name of the chromosome.
    sample : str
        Name of the sample (first one if empty - default)
    ad_tag : str
        AD tag in VCF file (default: &#39;AD&#39;)
    dp_tag : str
        DP tag in VCF file (default: &#39;DP&#39;)
    filter : bool
        If True it will read only variants with PASS filter, otherwise all

    Returns
    -------
    rd_p : list of int
        binned RD signal from DP tag (100 bins)
    rd_u : list of int
        binned RD signal from AD tag (100 bins)

    &#34;&#34;&#34;
    if sample == &#39;&#39;:
        sample = self.samples[0]
    pos = []
    rd_ad = []
    rd_dp = []
    last_chrom = None
    chr_count = 0

    try:
        for rec in self.file.fetch():
            if last_chrom is None:
                last_chrom = rec.chrom
            if last_chrom != rec.chrom and len(pos) &gt; 0:
                n = pos[-1] + 1
                if self.lengths[last_chrom] is not None:
                    n = self.lengths[last_chrom]
                n = n // 100 + 1
                rd_p = np.zeros(n)
                rd_u = np.zeros(n)
                count = np.zeros(n)
                for p, rd1, rd2 in zip(pos, rd_dp, rd_ad):
                    cgb = (p - 1) // 10000 * 100
                    rd_p[cgb] += rd1
                    rd_u[cgb] += rd2
                    count[cgb] += 1
                for i in range(n // 100 + 1):
                    if (100 * i) &lt; n:
                        if count[100 * i] != 0:
                            rd_p[i * 100:(i + 1) * 100] = rd_p[i * 100] / count[i * 100]
                            rd_u[i * 100:(i + 1) * 100] = rd_u[i * 100] / count[i * 100]
                        else:
                            rd_p[i * 100:(i + 1) * 100] = np.nan
                            rd_u[i * 100:(i + 1) * 100] = np.nan
                callback(last_chrom, rd_p, rd_u)
                pos = []
                rd_ad = []
                rd_dp = []
                chr_count += 1

            if (&#34;PASS&#34; in rec.filter.keys() or not filter) and (ad_tag in rec.samples[sample].keys()) and len(
                    rec.samples[sample][ad_tag]) &gt; 1 and (
                    dp_tag in rec.samples[sample].keys()) and (rec.samples[sample][dp_tag] is not None):
                try:
                    rd1 = sum(map(int, rec.samples[sample][ad_tag]))
                    rd2 = int(rec.samples[sample][dp_tag])
                    pos.append(rec.pos)
                    rd_ad.append(rd1)
                    rd_dp.append(rd2)
                except ValueError:
                    pass
            last_chrom = rec.chrom
        if len(pos) &gt; 0:
            n = pos[-1] + 1
            if self.lengths[last_chrom] is not None:
                n = self.lengths[last_chrom]
            n = n // 100 + 1
            rd_p = np.zeros(n)
            rd_u = np.zeros(n)
            count = np.zeros(n)

            for p, rd1, rd2 in zip(pos, rd_dp, rd_ad):
                cgb = (p - 1) // 10000 * 100
                rd_p[cgb] += rd1
                rd_u[cgb] += rd2
                count[cgb] += 1
            for i in range(n // 100 + 1):
                if (100 * i) &lt; n:
                    if count[100 * i] != 0:
                        rd_p[i * 100:(i + 1) * 100] = rd_p[i * 100] / count[i * 100]
                        rd_u[i * 100:(i + 1) * 100] = rd_u[i * 100] / count[i * 100]
                    else:
                        rd_p[i * 100:(i + 1) * 100] = np.nan
                        rd_u[i * 100:(i + 1) * 100] = np.nan
            callback(last_chrom, rd_p, rd_u)
            chr_count += 1
        return chr_count
    except ValueError:
        _logger.error(&#34;Variant file reading problem. Probably index file is missing or corrupted.&#34;)
        exit(0)</code></pre>
</details>
</dd>
<dt id="cnvpytor.vcf.Vcf.read_all_snp"><code class="name flex">
<span>def <span class="ident">read_all_snp</span></span>(<span>self, callback, sample='', ad_tag='AD', gt_tag='GT', filter=True)</span>
</code></dt>
<dd>
<section class="desc"><p>Read SNP/indel data for given sample name.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>callback</code></strong> :&ensp;<code>callable</code></dt>
<dd>Function to call after read a chromosome:
callback(chrom, pos, ref, alt, nref, nalt, gt, flag, qual)</dd>
<dt><strong><code>sample</code></strong> :&ensp;<code>str</code></dt>
<dd>Name of the sample (first one if empty - default)</dd>
<dt><strong><code>ad_tag</code></strong> :&ensp;<code>str</code></dt>
<dd>AD tag in VCF file (default: 'AD')</dd>
<dt><strong><code>gt_tag</code></strong> :&ensp;<code>str</code></dt>
<dd>GT tag in VCF file (default: 'GT')</dd>
<dt><strong><code>filter</code></strong> :&ensp;<code>bool</code></dt>
<dd>If True it will read only variants with PASS filter, otherwise all</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>count</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of read chromosomes</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def read_all_snp(self, callback, sample=&#39;&#39;, ad_tag=&#39;AD&#39;, gt_tag=&#39;GT&#39;, filter=True):
    &#34;&#34;&#34;
    Read SNP/indel data for given sample name.

    Parameters
    ----------
    callback : callable
        Function to call after read a chromosome:
        callback(chrom, pos, ref, alt, nref, nalt, gt, flag, qual)
    sample : str
        Name of the sample (first one if empty - default)
    ad_tag : str
        AD tag in VCF file (default: &#39;AD&#39;)
    gt_tag : str
        GT tag in VCF file (default: &#39;GT&#39;)
    filter : bool
        If True it will read only variants with PASS filter, otherwise all

    Returns
    -------
    count : int
        Number of read chromosomes

    &#34;&#34;&#34;
    if sample == &#39;&#39;:
        sample = self.samples[0]
    pos = []
    ref = []
    alt = []
    nref = []
    nalt = []
    gt = []
    flag = []
    qual = []
    filter_stat = {}
    last_chrom = None
    count = 0
    alphabet = [&#39;A&#39;, &#39;T&#39;, &#39;G&#39;, &#39;C&#39;, &#39;.&#39;]

    try:
        for rec in self.file.fetch():
            if last_chrom is None:
                last_chrom = rec.chrom
            if last_chrom != rec.chrom:
                _logger.info(&#34;Chromosome &#39;%s&#39; read. Number of variants to store: %d.&#34; % (last_chrom, len(pos)))
                _logger.debug(&#34;Variants filter field statistics:&#34;)
                for f in filter_stat:
                    _logger.debug(&#34;    * &#39;%s&#39; : %d&#34; % (f, filter_stat[f]))
                callback(last_chrom, pos, ref, alt, nref, nalt, gt, flag, qual)
                pos = []
                ref = []
                alt = []
                nref = []
                nalt = []
                gt = []
                flag = []
                qual = []
                filter_stat = {}
                count += 1
            if len(rec.filter.keys()) == 0:
                if &#34;No filter (.)&#34; in filter_stat:
                    filter_stat[&#34;No filter (.)&#34;] += 1
                else:
                    filter_stat[&#34;No filter (.)&#34;] = 1
            for f in rec.filter.keys():
                if f in filter_stat:
                    filter_stat[f] += 1
                else:
                    filter_stat[f] = 1

            if (&#34;PASS&#34; in rec.filter.keys() or not filter) and rec.alts and len(rec.alts) &gt;= 1 and (
                    gt_tag in rec.samples[sample].keys()) and (
                    ad_tag in rec.samples[sample].keys()) and len(rec.samples[sample][gt_tag]) &gt; 1 and len(
                rec.samples[sample][ad_tag]) &gt; 1:
                if (len(rec.samples[sample][ad_tag]) &gt; 1) and (rec.ref in alphabet) and (
                        rec.alts[0] in alphabet) and (rec.samples[sample][gt_tag][0] is not None) and (
                        rec.samples[sample][gt_tag][1] is not None):
                    pos.append(rec.pos)
                    ref.append(rec.ref)
                    alt.append(rec.alts[0])
                    flag.append(2)  # Assign P region as default (-mask_snps updates this flag)
                    if rec.qual == &#34;.&#34; or rec.qual is None:
                        qual.append(0)
                    else:
                        qual.append(int(rec.qual / 10))  # divide QUAL by factor 10 and truncate to one byte
                    if qual[-1] &gt; 255:
                        qual[-1] = 255
                    try:
                        nref.append(int(rec.samples[sample][ad_tag][0]))
                        nalt.append(int(rec.samples[sample][ad_tag][1]))
                    except ValueError:
                        nref.append(0)
                        nalt.append(0)
                    try:
                        UNICODE_EXISTS = bool(type(unicode))
                    except NameError:
                        unicode = str
                    if isinstance(rec.samples[sample][gt_tag], str) or isinstance(rec.samples[sample][gt_tag],
                                                                                  unicode):
                        gt.append(gt_from_str(rec.samples[sample][gt_tag]))
                    else:
                        gt.append(gt_from_list(rec.samples[sample][gt_tag], rec.samples[sample].phased))

            last_chrom = rec.chrom
        _logger.debug(&#34;Chromosome &#39;%s&#39; read. Number of variants to store: %d.&#34; % (last_chrom, len(pos)))
        _logger.debug(&#34;Variants filter field statistics:&#34;)
        for f in filter_stat:
            _logger.debug(&#34; * &#39;%s&#39; : %d&#34; % (f, filter_stat[f]))
        callback(last_chrom, pos, ref, alt, nref, nalt, gt, flag, qual)
        count += 1
        return count
    except ValueError:
        _logger.error(&#34;Variant file reading problem.&#34;)
        exit(0)</code></pre>
</details>
</dd>
<dt id="cnvpytor.vcf.Vcf.read_all_snp_no_counts"><code class="name flex">
<span>def <span class="ident">read_all_snp_no_counts</span></span>(<span>self, callback, sample='', gt_tag='GT', filter=True)</span>
</code></dt>
<dd>
<section class="desc"><p>Read SNP/indel data without counts (AD tag) for given sample name.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>callback</code></strong> :&ensp;<code>callable</code></dt>
<dd>Function to call after read a chromosome:
callback(chrom, pos, ref, alt, gt, flag, qual)</dd>
<dt><strong><code>sample</code></strong> :&ensp;<code>str</code></dt>
<dd>Name of the sample (first one if empty - default)</dd>
<dt><strong><code>gt_tag</code></strong> :&ensp;<code>str</code></dt>
<dd>GT tag in VCF file (default: 'GT')</dd>
<dt><strong><code>filter</code></strong> :&ensp;<code>bool</code></dt>
<dd>If True it will read only variants with PASS filter, otherwise all</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>count</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of read chromosomes</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def read_all_snp_no_counts(self, callback, sample=&#39;&#39;, gt_tag=&#39;GT&#39;, filter=True):
    &#34;&#34;&#34;
    Read SNP/indel data without counts (AD tag) for given sample name.

    Parameters
    ----------
    callback : callable
        Function to call after read a chromosome:
        callback(chrom, pos, ref, alt, gt, flag, qual)
    sample : str
        Name of the sample (first one if empty - default)
    gt_tag : str
        GT tag in VCF file (default: &#39;GT&#39;)
    filter : bool
        If True it will read only variants with PASS filter, otherwise all

    Returns
    -------
    count : int
        Number of read chromosomes

    &#34;&#34;&#34;
    if sample == &#39;&#39;:
        sample = self.samples[0]
    pos = []
    ref = []
    alt = []
    gt = []
    flag = []
    qual = []
    last_chrom = None
    count = 0
    alphabet = [&#39;A&#39;, &#39;T&#39;, &#39;G&#39;, &#39;C&#39;, &#39;.&#39;]

    try:
        for rec in self.file.fetch():
            if last_chrom is None:
                last_chrom = rec.chrom
            if last_chrom != rec.chrom:
                callback(last_chrom, pos, ref, alt, gt, flag, qual)
                pos = []
                ref = []
                alt = []
                gt = []
                flag = []
                qual = []
                count += 1

            if (&#34;PASS&#34; in rec.filter.keys() or not filter) and rec.alts and len(rec.alts) &gt;= 1 and (
                    gt_tag in rec.samples[sample].keys()) and len(rec.samples[sample][gt_tag]) &gt; 1:
                if (rec.ref in alphabet) and (rec.alts[0] in alphabet) and (
                        rec.samples[sample][gt_tag][0] is not None) and (
                        rec.samples[sample][gt_tag][1] is not None):
                    pos.append(rec.pos)
                    ref.append(rec.ref)
                    alt.append(rec.alts[0])
                    flag.append(2)  # Assign P region as default (-mask_snps updates this flag)
                    if rec.qual == &#34;.&#34; or rec.qual is None:
                        qual.append(0)
                    else:
                        qual.append(int(rec.qual / 10))  # divide QUAL by factor 10 and truncate to one byte
                    if qual[-1] &gt; 255:
                        qual[-1] = 255
                    try:
                        UNICODE_EXISTS = bool(type(unicode))
                    except NameError:
                        unicode = str
                    if isinstance(rec.samples[sample][gt_tag], str) or isinstance(rec.samples[sample][gt_tag],
                                                                                  unicode):
                        gt.append(gt_from_str(rec.samples[sample][gt_tag]))
                    else:
                        gt.append(gt_from_list(rec.samples[sample][gt_tag], rec.samples[sample].phased))

            last_chrom = rec.chrom
        if len(pos) &gt; 0:
            callback(last_chrom, pos, ref, alt, gt, flag, qual)
            count += 1
        return count
    except ValueError:
        _logger.error(&#34;Variant file reading problem.&#34;)
        exit(0)</code></pre>
</details>
</dd>
<dt id="cnvpytor.vcf.Vcf.read_all_snp_positions"><code class="name flex">
<span>def <span class="ident">read_all_snp_positions</span></span>(<span>self, callback, filter=True)</span>
</code></dt>
<dd>
<section class="desc"><p>Read SNP/indel data positions.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>callback</code></strong> :&ensp;<code>callable</code></dt>
<dd>Function to call after read a chromosome:
callback(chrom, pos, ref, alt)</dd>
<dt><strong><code>sample</code></strong> :&ensp;<code>str</code></dt>
<dd>Name of the sample (first one if empty - default)</dd>
<dt><strong><code>filter</code></strong> :&ensp;<code>bool</code></dt>
<dd>If True it will read only variants with PASS filter, otherwise all</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>count</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of read chromosomes</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def read_all_snp_positions(self, callback, filter=True):
    &#34;&#34;&#34;
    Read SNP/indel data positions.

    Parameters
    ----------
    callback : callable
        Function to call after read a chromosome:
        callback(chrom, pos, ref, alt)
    sample : str
        Name of the sample (first one if empty - default)
    filter : bool
        If True it will read only variants with PASS filter, otherwise all

    Returns
    -------
    count : int
        Number of read chromosomes

    &#34;&#34;&#34;
    pos = []
    ref = []
    alt = []
    last_chrom = None
    count = 0
    alphabet = [&#39;A&#39;, &#39;T&#39;, &#39;G&#39;, &#39;C&#39;, &#39;.&#39;]

    try:
        for rec in self.file.fetch():
            if last_chrom is None:
                last_chrom = rec.chrom
            if last_chrom != rec.chrom:
                callback(last_chrom, pos, ref, alt)
                pos = []
                ref = []
                alt = []
                count += 1
            if (&#34;PASS&#34; in rec.filter.keys() or not filter) and rec.alts and len(rec.alts) &gt;= 1:
                if (rec.ref in alphabet) and (rec.alts[0] in alphabet):
                    pos.append(rec.pos)
                    ref.append(rec.ref)
                    alt.append(rec.alts[0])
            last_chrom = rec.chrom
        if len(pos) &gt; 0:
            callback(last_chrom, pos, ref, alt)
            count += 1
        return count
    except ValueError:
        _logger.error(&#34;Variant file reading problem.&#34;)
        exit(0)</code></pre>
</details>
</dd>
<dt id="cnvpytor.vcf.Vcf.read_chromosome_rd"><code class="name flex">
<span>def <span class="ident">read_chromosome_rd</span></span>(<span>self, chr_name, sample='', ad_tag='AD', dp_tag='DP', filter=False)</span>
</code></dt>
<dd>
<section class="desc"><p>Read RD data for given chromosome and sample.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>chr_name</code></strong> :&ensp;<code>str</code></dt>
<dd>Name of the chromosome.</dd>
<dt><strong><code>sample</code></strong> :&ensp;<code>str</code></dt>
<dd>Name of the sample (first one if empty - default)</dd>
<dt><strong><code>ad_tag</code></strong> :&ensp;<code>str</code></dt>
<dd>AD tag in VCF file (default: 'AD')</dd>
<dt><strong><code>dp_tag</code></strong> :&ensp;<code>str</code></dt>
<dd>DP tag in VCF file (default: 'DP')</dd>
<dt><strong><code>filter</code></strong> :&ensp;<code>bool</code></dt>
<dd>If True it will read only variants with PASS filter, otherwise all</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>rd_p</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>binned RD signal from DP tag (100 bins)</dd>
<dt><strong><code>rd_u</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>binned RD signal from AD tag (100 bins)</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def read_chromosome_rd(self, chr_name, sample=&#39;&#39;, ad_tag=&#39;AD&#39;, dp_tag=&#39;DP&#39;, filter=False):
    &#34;&#34;&#34;
    Read RD data for given chromosome and sample.

    Parameters
    ----------
    chr_name : str
        Name of the chromosome.
    sample : str
        Name of the sample (first one if empty - default)
    ad_tag : str
        AD tag in VCF file (default: &#39;AD&#39;)
    dp_tag : str
        DP tag in VCF file (default: &#39;DP&#39;)
    filter : bool
        If True it will read only variants with PASS filter, otherwise all

    Returns
    -------
    rd_p : list of int
        binned RD signal from DP tag (100 bins)
    rd_u : list of int
        binned RD signal from AD tag (100 bins)

    &#34;&#34;&#34;
    if sample == &#39;&#39;:
        sample = self.samples[0]
    pos = []
    rd_ad = []
    rd_dp = []
    try:
        for rec in self.file.fetch(chr_name):
            if (&#34;PASS&#34; in rec.filter.keys() or not filter) and (ad_tag in rec.samples[sample].keys()) and len(
                    rec.samples[sample][ad_tag]) &gt; 1 and (dp_tag in rec.samples[sample].keys()) and (
                    rec.samples[sample][dp_tag] is not None):
                try:
                    rd1 = sum(map(int, rec.samples[sample][ad_tag]))
                    rd2 = int(rec.samples[sample][dp_tag])
                    pos.append(rec.pos)
                    rd_ad.append(rd1)
                    rd_dp.append(rd2)
                except ValueError:
                    pass
    except ValueError:
        _logger.error(&#34;Variant file reading problem. Probably index file is missing or corrupted.&#34;)
        exit(0)
    n = 0
    if len(pos) == 0:
        return None, None
    n = pos[-1] + 1
    if self.lengths[chr_name] is not None:
        n = self.lengths[chr_name]
    n = n // 100 + 1
    rd_p = np.zeros(n)
    rd_u = np.zeros(n)
    count = np.zeros(n)
    for p, rd1, rd2 in zip(pos, rd_dp, rd_ad):
        cgb = (p - 1) // 10000 * 100
        rd_p[cgb] += rd1
        rd_u[cgb] += rd2
        count[cgb] += 1
    for i in range(n // 100 + 1):
        if (100 * i) &lt; n:
            if count[100 * i] != 0:
                rd_p[i * 100:(i + 1) * 100] = rd_p[i * 100] / count[i * 100]
                rd_u[i * 100:(i + 1) * 100] = rd_u[i * 100] / count[i * 100]
            else:
                rd_p[i * 100:(i + 1) * 100] = np.nan
                rd_u[i * 100:(i + 1) * 100] = np.nan
    return rd_p, rd_u</code></pre>
</details>
</dd>
<dt id="cnvpytor.vcf.Vcf.read_chromosome_snp"><code class="name flex">
<span>def <span class="ident">read_chromosome_snp</span></span>(<span>self, chr_name, sample='', ad_tag='AD', gt_tag='GT', filter=True)</span>
</code></dt>
<dd>
<section class="desc"><p>Read SNP/indel data for given chromosome and sample.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>chr_name</code></strong> :&ensp;<code>str</code></dt>
<dd>Name of the chromosome.</dd>
<dt><strong><code>sample</code></strong> :&ensp;<code>str</code></dt>
<dd>Name of the sample (first one if empty - default)</dd>
<dt><strong><code>ad_tag</code></strong> :&ensp;<code>str</code></dt>
<dd>AD tag in VCF file (default: 'AD')</dd>
<dt><strong><code>gt_tag</code></strong> :&ensp;<code>str</code></dt>
<dd>GT tag in VCF file (default: 'GT')</dd>
<dt><strong><code>filter</code></strong> :&ensp;<code>bool</code></dt>
<dd>If True it will read only variants with PASS filter, otherwise all</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>pos</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>List of SNP positions.</dd>
<dt><strong><code>ref</code></strong> :&ensp;<code>list</code> of <code>str</code></dt>
<dd>List of SNP reference base (A, T, G, C or .).</dd>
<dt><strong><code>alt</code></strong> :&ensp;<code>list</code> of <code>str</code></dt>
<dd>List of SNP alternative base (A, T, G, C or .).</dd>
<dt><strong><code>nref</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>Count of reads contains reference SNP.</dd>
<dt><strong><code>nalt</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>Count of reads contains alternative SNP.</dd>
<dt><strong><code>gt</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>List of genotypes (0 - "0/0", 1 - "0/1", 3- "1/1", 4 - "0|0" , 5 - "0|1", 6 - "1|0", 7 - "1|1").</dd>
<dt><strong><code>flag</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>Binary flag: first bit 1 - SNP exists in database, second bit 1 - SNP in P region of strict mask.</dd>
<dt><strong><code>qual</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>SNP quality (scale 0 - 255).</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def read_chromosome_snp(self, chr_name, sample=&#39;&#39;, ad_tag=&#39;AD&#39;, gt_tag=&#39;GT&#39;, filter=True):
    &#34;&#34;&#34;
    Read SNP/indel data for given chromosome and sample.

    Parameters
    ----------
    chr_name : str
        Name of the chromosome.
    sample : str
        Name of the sample (first one if empty - default)
    ad_tag : str
        AD tag in VCF file (default: &#39;AD&#39;)
    gt_tag : str
        GT tag in VCF file (default: &#39;GT&#39;)
    filter : bool
        If True it will read only variants with PASS filter, otherwise all

    Returns
    -------
    pos : list of int
        List of SNP positions.
    ref : list of str
        List of SNP reference base (A, T, G, C or .).
    alt : list of str
        List of SNP alternative base (A, T, G, C or .).
    nref : list of int
        Count of reads contains reference SNP.
    nalt : list of int
        Count of reads contains alternative SNP.
    gt : list of int
        List of genotypes (0 - &#34;0/0&#34;, 1 - &#34;0/1&#34;, 3- &#34;1/1&#34;, 4 - &#34;0|0&#34; , 5 - &#34;0|1&#34;, 6 - &#34;1|0&#34;, 7 - &#34;1|1&#34;).
    flag : list of int
        Binary flag: first bit 1 - SNP exists in database, second bit 1 - SNP in P region of strict mask.
    qual : list of int
        SNP quality (scale 0 - 255).

    &#34;&#34;&#34;
    if sample == &#39;&#39;:
        sample = self.samples[0]
    pos = []
    ref = []
    alt = []
    nref = []
    nalt = []
    gt = []
    flag = []
    qual = []
    filter_stat = {}
    alphabet = [&#39;A&#39;, &#39;T&#39;, &#39;G&#39;, &#39;C&#39;, &#39;.&#39;]
    try:
        for rec in self.file.fetch(chr_name):
            if len(rec.filter.keys()) == 0:
                if &#34;No filter (.)&#34; in filter_stat:
                    filter_stat[&#34;No filter (.)&#34;] += 1
                else:
                    filter_stat[&#34;No filter (.)&#34;] = 1
            for f in rec.filter.keys():
                if f in filter_stat:
                    filter_stat[f] += 1
                else:
                    filter_stat[f] = 1
            if (&#34;PASS&#34; in rec.filter.keys() or not filter) and rec.alts and len(rec.alts) &gt;= 1 and (
                    gt_tag in rec.samples[sample].keys()) and (
                    ad_tag in rec.samples[sample].keys()) and len(rec.samples[sample][gt_tag]) &gt; 1 and len(
                rec.samples[sample][ad_tag]) &gt; 1:
                if (len(rec.samples[sample][ad_tag]) &gt; 1) and (rec.ref in alphabet) and (rec.alts[0] in alphabet):
                    pos.append(rec.pos)
                    ref.append(rec.ref)
                    alt.append(rec.alts[0])
                    flag.append(2)  # Assign P region as default (-mask_snps updates this flag)
                    if rec.qual == &#34;.&#34; or rec.qual is None:
                        qual.append(0)
                    else:
                        qual.append(int(rec.qual / 10))  # divide QUAL by factor 10 and truncate to one byte
                    if qual[-1] &gt; 255:
                        qual[-1] = 255
                    try:
                        nref.append(int(rec.samples[sample][ad_tag][0]))
                        nalt.append(int(rec.samples[sample][ad_tag][1]))
                    except ValueError:
                        nref.append(0)
                        nalt.append(0)
                    try:
                        UNICODE_EXISTS = bool(type(unicode))
                    except NameError:
                        unicode = str
                    if isinstance(rec.samples[sample][gt_tag], str) or isinstance(rec.samples[sample][gt_tag],
                                                                                  unicode):
                        gt.append(gt_from_str(rec.samples[sample][gt_tag]))
                    else:
                        gt.append(gt_from_list(rec.samples[sample][gt_tag], rec.samples[sample].phased))
    except ValueError:
        _logger.error(&#34;Variant file reading problem. Probably index file is missing or corrupted.&#34;)
        exit(0)
    _logger.debug(&#34;Chromosome &#39;%s&#39; read. Number of variants to store: %d.&#34; % (chr_name, len(pos)))
    _logger.debug(&#34;Variants filter field statistics:&#34;)
    for f in filter_stat:
        _logger.debug(&#34; * &#39;%s&#39; : %d&#34; % (f, filter_stat[f]))
    return pos, ref, alt, nref, nalt, gt, flag, qual</code></pre>
</details>
</dd>
<dt id="cnvpytor.vcf.Vcf.read_chromosome_snp_no_counts"><code class="name flex">
<span>def <span class="ident">read_chromosome_snp_no_counts</span></span>(<span>self, chr_name, sample='', gt_tag='GT', filter=True)</span>
</code></dt>
<dd>
<section class="desc"><p>Read SNP/indel data without counts (AD tag) for given chromosome and sample.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>chr_name</code></strong> :&ensp;<code>str</code></dt>
<dd>Name of the chromosome.</dd>
<dt><strong><code>sample</code></strong> :&ensp;<code>str</code></dt>
<dd>Name of the sample (first one if empty - default)</dd>
<dt><strong><code>gt_tag</code></strong> :&ensp;<code>str</code></dt>
<dd>GT tag in VCF file (default: 'GT')</dd>
<dt><strong><code>filter</code></strong> :&ensp;<code>bool</code></dt>
<dd>If True it will read only variants with PASS filter, otherwise all</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>pos</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>List of SNP positions.</dd>
<dt><strong><code>ref</code></strong> :&ensp;<code>list</code> of <code>str</code></dt>
<dd>List of SNP reference base (A, T, G, C or .).</dd>
<dt><strong><code>alt</code></strong> :&ensp;<code>list</code> of <code>str</code></dt>
<dd>List of SNP alternative base (A, T, G, C or .).</dd>
<dt><strong><code>gt</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>List of genotypes (0 - "0/0", 1 - "0/1", 3- "1/1", 4 - "0|0" , 5 - "0|1", 6 - "1|0", 7 - "1|1").</dd>
<dt><strong><code>flag</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>Binary flag: first bit 1 - SNP exists in database, second bit 1 - SNP in P region of strict mask.</dd>
<dt><strong><code>qual</code></strong> :&ensp;<code>list</code> of <code>int</code></dt>
<dd>SNP quality (scale 0 - 255).</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def read_chromosome_snp_no_counts(self, chr_name, sample=&#39;&#39;, gt_tag=&#39;GT&#39;, filter=True):
    &#34;&#34;&#34;
    Read SNP/indel data without counts (AD tag) for given chromosome and sample.

    Parameters
    ----------
    chr_name : str
        Name of the chromosome.
    sample : str
        Name of the sample (first one if empty - default)
    gt_tag : str
        GT tag in VCF file (default: &#39;GT&#39;)
    filter : bool
        If True it will read only variants with PASS filter, otherwise all

    Returns
    -------
    pos : list of int
        List of SNP positions.
    ref : list of str
        List of SNP reference base (A, T, G, C or .).
    alt : list of str
        List of SNP alternative base (A, T, G, C or .).
    gt : list of int
        List of genotypes (0 - &#34;0/0&#34;, 1 - &#34;0/1&#34;, 3- &#34;1/1&#34;, 4 - &#34;0|0&#34; , 5 - &#34;0|1&#34;, 6 - &#34;1|0&#34;, 7 - &#34;1|1&#34;).
    flag : list of int
        Binary flag: first bit 1 - SNP exists in database, second bit 1 - SNP in P region of strict mask.
    qual : list of int
        SNP quality (scale 0 - 255).

    &#34;&#34;&#34;
    if sample == &#39;&#39;:
        sample = self.samples[0]
    pos = []
    ref = []
    alt = []
    gt = []
    flag = []
    qual = []
    alphabet = [&#39;A&#39;, &#39;T&#39;, &#39;G&#39;, &#39;C&#39;, &#39;.&#39;]
    try:
        for rec in self.file.fetch(chr_name):
            if (&#34;PASS&#34; in rec.filter.keys() or not filter) and rec.alts and len(rec.alts) &gt;= 1 and (
                    gt_tag in rec.samples[sample].keys()) and len(rec.samples[sample][gt_tag]) &gt; 1:
                if (rec.ref in alphabet) and (rec.alts[0] in alphabet):
                    pos.append(rec.pos)
                    ref.append(rec.ref)
                    alt.append(rec.alts[0])
                    flag.append(2)  # Assign P region as default (-mask_snps updates this flag)
                    if rec.qual == &#34;.&#34; or rec.qual is None:
                        qual.append(0)
                    else:
                        qual.append(int(rec.qual / 10))  # divide QUAL by factor 10 and truncate to one byte
                    if qual[-1] &gt; 255:
                        qual[-1] = 255
                    try:
                        UNICODE_EXISTS = bool(type(unicode))
                    except NameError:
                        unicode = str
                    if isinstance(rec.samples[sample][gt_tag], str) or isinstance(rec.samples[sample][gt_tag],
                                                                                  unicode):
                        gt.append(gt_from_str(rec.samples[sample][gt_tag]))
                    else:
                        gt.append(gt_from_list(rec.samples[sample][gt_tag], rec.samples[sample].phased))

    except ValueError:
        _logger.error(&#34;Variant file reading problem. Probably index file is missing or corrupted.&#34;)
        exit(0)
    return pos, ref, alt, gt, flag, qual</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="cnvpytor" href="index.html">cnvpytor</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="cnvpytor.vcf.Vcf" href="#cnvpytor.vcf.Vcf">Vcf</a></code></h4>
<ul class="">
<li><code><a title="cnvpytor.vcf.Vcf.get_chromosomes" href="#cnvpytor.vcf.Vcf.get_chromosomes">get_chromosomes</a></code></li>
<li><code><a title="cnvpytor.vcf.Vcf.get_samples" href="#cnvpytor.vcf.Vcf.get_samples">get_samples</a></code></li>
<li><code><a title="cnvpytor.vcf.Vcf.read_all_rd" href="#cnvpytor.vcf.Vcf.read_all_rd">read_all_rd</a></code></li>
<li><code><a title="cnvpytor.vcf.Vcf.read_all_snp" href="#cnvpytor.vcf.Vcf.read_all_snp">read_all_snp</a></code></li>
<li><code><a title="cnvpytor.vcf.Vcf.read_all_snp_no_counts" href="#cnvpytor.vcf.Vcf.read_all_snp_no_counts">read_all_snp_no_counts</a></code></li>
<li><code><a title="cnvpytor.vcf.Vcf.read_all_snp_positions" href="#cnvpytor.vcf.Vcf.read_all_snp_positions">read_all_snp_positions</a></code></li>
<li><code><a title="cnvpytor.vcf.Vcf.read_chromosome_rd" href="#cnvpytor.vcf.Vcf.read_chromosome_rd">read_chromosome_rd</a></code></li>
<li><code><a title="cnvpytor.vcf.Vcf.read_chromosome_snp" href="#cnvpytor.vcf.Vcf.read_chromosome_snp">read_chromosome_snp</a></code></li>
<li><code><a title="cnvpytor.vcf.Vcf.read_chromosome_snp_no_counts" href="#cnvpytor.vcf.Vcf.read_chromosome_snp_no_counts">read_chromosome_snp_no_counts</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.6.3</a>.</p>
</footer>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script>hljs.initHighlightingOnLoad()</script>
</body>
</html>