<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.6.3" />
<title>cnvpytor.viewer API documentation</title>
<meta name="description" content="cnvpytor.viewer â€¦" />
<link href='https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.0/normalize.min.css' rel='stylesheet'>
<link href='https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/8.0.0/sanitize.min.css' rel='stylesheet'>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet">
<style>.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{font-weight:bold}#index h4 + ul{margin-bottom:.6em}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase;cursor:pointer}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>cnvpytor.viewer</code></h1>
</header>
<section id="section-intro">
<p>cnvpytor.viewer</p>
<p>Class Viewer: ploting CNVpytor data</p>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">&#34;&#34;&#34; cnvpytor.viewer

Class Viewer: ploting CNVpytor data
&#34;&#34;&#34;
from __future__ import absolute_import, print_function, division

from .io import *
from .utils import *
from .genome import *
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
import numpy as np
import logging
import readline
import traceback

_logger = logging.getLogger(&#34;cnvpytor.viewer&#34;)


class Viewer:
    def __init__(self, files, output_filename):
        _logger.debug(&#34;Viewer class init: files [%s], png_prefix &#39;%s&#39;.&#34; % (&#34;, &#34;.join(files), output_filename))
        self.io = [IO(f, ro=True) for f in files]
        self.output_filename = output_filename
        self.set_style(&#39;classic&#39;)
        self.io_gc = self.io[0]
        self.io_mask = self.io[0]
        self.reference_genome = None
        self.interactive = False
        self.plot_files = [True for i in files]
        self.fig = None
        self.xkcd = False
        self.grid = &#34;auto&#34;
        self.plot_file = 0
        self.use_mask = True
        self.use_id = False
        self.baf_colors = {(0, 0): &#34;yellow&#34;, (0, 1): &#34;orange&#34;, (1, 0): &#34;cyan&#34;, (1, 1): &#34;blue&#34;, (2, 0): &#34;lime&#34;,
                           (2, 1): &#34;green&#34;, (3, 0): &#34;yellow&#34;, (3, 1): &#34;orange&#34;}
        if self.io[0].signal_exists(None, None, &#34;reference genome&#34;):
            rg_name = np.array(self.io[0].get_signal(None, None, &#34;reference genome&#34;)).astype(&#34;str&#34;)[0]
            self.reference_genome = Genome.reference_genomes[rg_name]
            if &#34;mask_file&#34; in Genome.reference_genomes[rg_name]:
                self.io_mask = IO(Genome.reference_genomes[rg_name][&#34;mask_file&#34;], ro=True, buffer=True)
            if &#34;gc_file&#34; in Genome.reference_genomes[rg_name]:
                self.io_gc = IO(Genome.reference_genomes[rg_name][&#34;gc_file&#34;], ro=True, buffer=True)

    def parse(self, command, args):
        current = &#34;regions&#34;
        regions = []

        for p in command:
            if p.isdigit() and (int(p) % 100) == 0:
                if current == &#34;rd&#34;:
                    self.rd(int(p), args.use_mask_with_rd)
                if current == &#34;likelihood&#34;:
                    self.likelihood(int(p))
                elif current == &#34;manhattan&#34;:
                    self.manhattan(int(p), args.use_mask_with_rd)
                elif current == &#34;stat&#34;:
                    self.stat(int(p))
                elif current == &#34;regions&#34;:
                    self.regions(int(p), regions, panels=args.panels)
                    regions = []
            elif p == &#34;rdstat&#34;:
                self.stat()
            elif p == &#34;baf&#34;:
                self.baf()
            elif p in [&#34;rd&#34;, &#34;manhattan&#34;, &#34;stat&#34;, &#34;regions&#34;, &#34;likelihood&#34;]:
                current = p
            elif current == &#34;regions&#34;:
                regions.append(p)
            else:
                current = p

    def plot(self, args):
        self.interactive = False
        self.use_mask = not args.no_mask
        self.use_id = args.use_id
        self.parse(args.plot, args)

    def prompt(self, bin_size, args):
        self.interactive = True
        self.use_mask = not args.no_mask
        self.use_id = args.use_id
        command_tree = {&#34;set&#34;: {&#34;bin_size&#34;: None,
                                &#34;panels&#34;: None,
                                &#34;use_mask_rd&#34;: None,
                                &#34;plot_files&#34;: None,
                                &#34;plot_file&#34;: None,
                                &#34;grid&#34;: None,
                                &#34;use_mask&#34;:None,
                                &#34;use_id&#34;:None,
                                &#34;xkcd&#34;: None,
                                &#34;style&#34;: {}
                                },
                        &#34;unset&#34;: {&#34;use_mask_rd&#34;: None,
                                  &#34;xkcd&#34;: None
                                  },
                        &#34;save&#34;: None, &#34;show&#34;: None, &#34;quit&#34;: None, &#34;rd&#34;: None, &#34;likelihood&#34;: None, &#34;baf&#34;: None,
                        &#34;stat&#34;: None, &#34;rdstat&#34;: None,
                        }
        chromosomes = set({})
        for f in self.io:
            chromosomes = chromosomes.union(set(f.rd_chromosomes()))
            chromosomes = chromosomes.union(set(f.snp_chromosomes()))
        for c in chromosomes:
            command_tree[c] = None
        command_tree[&#34;set&#34;][&#34;style&#34;] = dict(zip(plt.style.available, [None] * len(plt.style.available)))
        readline.parse_and_bind(&#34;tab: complete&#34;)
        completer = PromptCompleter(command_tree)
        readline.set_completer(completer.complete)
        quit = False
        try:
            while not quit:
                try:
                    line = raw_input(&#34;cnvpytor&gt; &#34;)
                except NameError:
                    line = input(&#34;cnvpytor&gt; &#34;)

                pre = line.split(&#34;&gt;&#34;)
                f = pre[0].strip().split(&#34; &#34;)
                n = len(f)
                if len(line) == 0:
                    continue
                elif f[0] == &#34;quit&#34;:
                    quit = True
                elif line[0] == &#34;|&#34;:
                    try:
                        eval(compile(line[1:], &#39;&lt;string&gt;&#39;, &#39;single&#39;))
                    except Exception as e:
                        print(traceback.format_exc())
                elif f[0] == &#34;save&#34; and n &gt; 1:
                    plt.savefig(f[1])
                elif f[0] in [&#34;draw&#34;, &#34;repaint&#34;, &#34;update&#34;] and n == 1:
                    self.fig.canvas.draw()
                elif f[0] == &#34;show&#34;:
                    if n == 1:
                        print(&#34;    Parameters&#34;)
                        print(&#34;        * bin_size:&#34;, bin_size)
                        print(&#34;        * panels:&#34;, args.panels)
                        print(&#34;        * use_mask_rd:&#34;, args.use_mask_with_rd)
                        print(&#34;        * use_mask:&#34;, self.use_mask)
                        print(&#34;        * use_id:&#34;, self.use_id)
                        print(&#34;        * plot_files:&#34;, list(zip(map(lambda x: x.filename, self.io), self.plot_files)))
                        print(&#34;        * plot_file:&#34;, self.plot_file)
                        print(&#34;        * grid:&#34;, self.grid)
                        print(&#34;    Available plot styles:&#34;, &#34;, &#34;.join(plt.style.available))
                elif f[0] == &#34;set&#34;:
                    if n &gt; 2 and f[1] == &#34;bin_size&#34;:
                        try:
                            bin_size = binsize_type(f[2])
                        except (ArgumentTypeError, ValueError):
                            _logger.warning(&#34;bin_size should be intiger divisible by 100&#34;)
                    elif n &gt; 3 and f[1] == &#34;grid&#34; and f[2].isdigit() and f[3].isdigit():
                        self.grid = (int(f[2]), int(f[3]))
                    elif n &gt; 2 and f[1] == &#34;plot_file&#34; and f[2].isdigit() and int(f[2]) &lt; len(self.io):
                        self.plot_file = int(f[2])
                    elif n &gt; 2 and f[1] == &#34;plot_files&#34;:
                        select = [int(i) for i in f[2:] if i.isdigit()]
                        self.plot_files = list([ix in select for ix in range(len(self.io))])
                    elif n &gt; 2 and f[1] == &#34;panels&#34;:
                        args.panels = f[2:]
                    elif n &gt; 2 and f[1] == &#34;style&#34;:
                        self.set_style(f[2])
                        if self.fig:
                            self.fig.canvas.draw()
                    elif n &gt; 1 and f[1] == &#34;use_mask_rd&#34;:
                        args.use_mask_with_rd = True
                    elif n &gt; 1 and f[1] == &#34;use_mask&#34;:
                        self.use_mask = True
                    elif n &gt; 1 and f[1] == &#34;use_id&#34;:
                        self.use_id = True
                    elif n &gt; 1 and f[1] == &#34;xkcd&#34;:
                        self.set_style(&#34;classic&#34;)
                        from matplotlib import patheffects
                        plt.xkcd()
                        plt.rcParams[&#34;path.effects&#34;] = [patheffects.withStroke(linewidth=0.5, foreground=&#34;w&#34;)]
                        self.xkcd = True
                    else:
                        print(&#34;Unrecognized set argument!&#34;)
                elif f[0] == &#34;unset&#34;:
                    if n &gt; 1 and f[1] == &#34;use_mask_rd&#34;:
                        args.use_mask_with_rd = False
                    elif n &gt; 1 and f[1] == &#34;use_mask&#34;:
                        self.use_mask = False
                    elif n &gt; 1 and f[1] == &#34;use_id&#34;:
                        self.use_id = False
                    elif n &gt; 1 and f[1] == &#34;xkcd&#34; and self.xkcd:
                        plt.rcdefaults()
                        self.set_style(&#39;classic&#39;)
                        self.xkcd = False

                    else:
                        print(&#34;Unrecognized unset argument!&#34;)
                else:

                    try:
                        if f[0] not in [&#34;rdstat&#34;, &#34;baf&#34;]:
                            self.parse(f + [str(bin_size)], args)
                        else:
                            self.parse(f, args)
                        if len(pre) &gt; 1:
                            fns = pre[1].strip().split(&#34; &#34;)
                            if fns[0] != &#34;&#34;:
                                plt.savefig(fns[0], dpi=200)
                    except Exception as e:
                        print(traceback.format_exc())
        except (EOFError, KeyboardInterrupt):
            print()
            return

    @staticmethod
    def set_style(style):
        if style in plt.style.available:
            plt.style.use(&#34;default&#34;)
            plt.style.use(style)

    def image_filename(self, sufix):
        parts = self.output_filename.split(&#34;.&#34;)
        if parts[-1] != &#34;png&#34; and parts[-1] != &#34;pdf&#34; and parts[-1] != &#34;jpg&#34; and parts[-1] != &#34;eps&#34; and parts[
            -1] != &#34;svg&#34;:
            _logger.warning(&#34;File extension should be: .jpg, .png, .svg, .eps or .pdf&#34;)
            exit(0)
        parts[-1] = sufix + &#34;.&#34; + parts[-1]
        return &#34;.&#34;.join(parts)

    def stat(self, his_bin_size=100):
        plt.clf()
        auto = self.io[0].signal_exists(None, his_bin_size, &#34;RD stat&#34;, FLAG_AUTO)
        sex = self.io[0].signal_exists(None, his_bin_size, &#34;RD stat&#34;, FLAG_SEX)
        mt = self.io[0].signal_exists(None, his_bin_size, &#34;RD stat&#34;, FLAG_MT) and (his_bin_size &lt; 1001)
        if not (auto or sex or mt):
            return
        cond = [auto, sex, mt]
        stat_list = []
        n_cols = sum(map(int, cond))
        ix = 1
        plt.rcParams[&#34;font.size&#34;] = 8
        self.fig = plt.figure(1, figsize=(4 * n_cols, 8), dpi=90, facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
        for t, c, flag in zip([&#34;Autosomes&#34;, &#34;X/Y&#34;, &#34;Mitochondria&#34;], cond, [FLAG_AUTO, FLAG_SEX, FLAG_MT]):
            if c:
                stat = self.io[0].get_signal(None, his_bin_size, &#34;RD stat&#34;, flag)
                stat_list.append(stat)
                max_rd = int(stat[0])
                bin_size = int(stat[1])
                n_bins = int(stat[2])
                lim_rd = int(max(2 * stat[4], stat[4] + 3 * stat[5]))
                _logger.info(&#34;RD stat for %s: %.2f +- %.2f&#34; % (t, stat[4], stat[5]))
                if t == &#34;Mitochondria&#34; and auto:
                    _logger.info(&#34;RD stat for %s - number of miochondria per cell: %.2f +- %.2f&#34; % (
                        t, 2 * stat[4] / stat_list[0][4],
                        2 * stat[5] / stat_list[0][4] + stat_list[0][5] * stat[4] / (
                                stat_list[0][4] * stat_list[0][4])))
                his_p = self.io[0].get_signal(None, his_bin_size, &#34;RD p dist&#34;, flag)
                his_u = self.io[0].get_signal(None, his_bin_size, &#34;RD u dist&#34;, flag)
                his_rd_gc = self.io[0].get_signal(None, his_bin_size, &#34;RD GC dist&#34;, flag)
                gc_corr = self.io[0].get_signal(None, his_bin_size, &#34;GC corr&#34;, flag)
                ax = plt.subplot(2, n_cols, ix)
                ax.set_xlabel(&#34;RD&#34;)
                ax.set_ylabel(&#34;GC [%]&#34;)
                ax.xaxis.set_ticklabels([])
                ax.set_title(t)
                his_rd_gc[0][0] = 0
                ax.imshow(his_rd_gc[:lim_rd // bin_size, :].T, aspect=&#34;auto&#34;, interpolation=&#39;nearest&#39;, origin=&#39;lower&#39;)
                ax.plot(gc_corr * stat[4] / bin_size, range(101), &#34;w-&#34;)

                ax = plt.subplot(2, n_cols, ix + n_cols)
                ax.set_ylabel(&#34;Normalised distribution&#34;)
                ax.set_xlabel(&#34;RD&#34;)
                ax.set_xlim([0, lim_rd])
                ax.set_ylim([0, 1.1])
                bins = range(0, max_rd, bin_size)
                x = np.arange(0, max_rd // bin_size * bin_size, 0.1 * bin_size)
                plt.plot(x, normal(x, 1, stat[4], stat[5]), &#34;g-&#34;)
                x = np.array(bins)
                plt.plot(x[:len(his_u)], his_u / stat[3], &#34;y*&#34;)
                plt.plot(x[:len(his_p)], his_p / stat[3], &#34;b*&#34;)
                ix += 1
        plt.subplots_adjust(bottom=0.08, top=0.95, wspace=0.25, hspace=0, left=0.05 * 3 / n_cols, right=0.95)
        if self.output_filename != &#34;&#34;:
            plt.savefig(self.image_filename(&#34;stat&#34;), dpi=150)
            plt.close(self.fig)
        elif self.interactive:
            plt.show(block=False)
            plt.draw()
        else:
            plt.show()

    def rd(self, bin_size, use_mask):
        plt.clf()
        if self.reference_genome is None:
            _logger.warning(&#34;Missing reference genome required for gview.&#34;)
            return
        plt.rcParams[&#34;font.size&#34;] = 8
        chroms = []
        for c, (l, t) in self.reference_genome[&#34;chromosomes&#34;].items():
            rd_chr = self.io[self.plot_file].rd_chromosome_name(c)
            if self.io[self.plot_file].signal_exists(rd_chr, bin_size, &#34;RD&#34;, 0) and \
                    self.io[self.plot_file].signal_exists(rd_chr, bin_size, &#34;RD&#34;, FLAG_GC_CORR) and \
                    (Genome.is_autosome(c) or Genome.is_sex_chrom(c)):
                chroms.append((rd_chr, l))
        sx, sy = self.panels_shape(len(chroms))
        self.fig = plt.figure(1, figsize=(sx, sy), dpi=200, facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
        ix = 1
        for c, l in chroms:
            flag = FLAG_MT if Genome.is_mt_chrom(c) else FLAG_SEX if Genome.is_sex_chrom(c) else FLAG_AUTO
            stat = self.io[self.plot_file].get_signal(None, bin_size, &#34;RD stat&#34;, flag)
            if stat is None:
                _logger.error(&#34;Data for bin size %d is missing in file &#39;%s&#39;!&#34; % (bin_size, self.io[self.plot_file].filename))
                exit(0)
            flag_rd = 0
            if use_mask:
                flag_rd = FLAG_USEMASK
            his_p = self.io[self.plot_file].get_signal(c, bin_size, &#34;RD&#34;, flag_rd)
            his_p_corr = self.io[self.plot_file].get_signal(c, bin_size, &#34;RD&#34;, flag_rd | FLAG_GC_CORR)
            ax = plt.subplot(sx, sy, ix)
            ax.set_title(c, position=(0.01, 0.9), fontdict={&#39;verticalalignment&#39;: &#39;top&#39;, &#39;horizontalalignment&#39;: &#39;left&#39;},
                         color=&#39;C0&#39;)
            ax.xaxis.set_ticklabels([])
            ax.yaxis.set_ticklabels([])
            ax.yaxis.set_ticks(np.arange(0, 3, 0.5) * stat[4], [])
            ax.xaxis.set_ticks(np.arange(0, (l + 10e6) // bin_size, 10e6 // bin_size), [])
            ax.set_ylim([0, max(3. * stat[4], stat[4] + 5. * stat[5])])
            n_bins = l // bin_size
            ax.set_xlim([-n_bins * 0.05, n_bins * 1.05])
            ax.grid()

            plt.step(his_p, &#34;grey&#34;)
            plt.step(his_p_corr, &#34;k&#34;)
            ix += 1
        plt.subplots_adjust(bottom=0., top=1., wspace=0, hspace=0, left=0., right=1.)
        if self.output_filename != &#34;&#34;:
            plt.savefig(self.image_filename(&#34;gview&#34;), dpi=150)
            plt.close(self.fig)
        elif self.interactive:
            plt.show(block=False)
            plt.draw()
        else:
            plt.show()

    def likelihood(self, bin_size):
        plt.clf()
        snp_flag = (FLAG_USEMASK if self.use_mask else 0) | (FLAG_USEID if self.use_id else 0)
        if self.reference_genome is None:
            _logger.warning(&#34;Missing reference genome required for gview.&#34;)
            return
        plt.rcParams[&#34;font.size&#34;] = 8
        chroms = []
        if self.reference_genome is None:
            chroms = self.io[self.plot_file].snp_chromosomes()
        else:
            for c, (l, t) in self.reference_genome[&#34;chromosomes&#34;].items():
                snp_chr = self.io[self.plot_file].snp_chromosome_name(c)
                if self.io[self.plot_file].signal_exists(snp_chr, bin_size, &#34;SNP likelihood&#34;, snp_flag) and (
                        Genome.is_autosome(c) or Genome.is_sex_chrom(c)):
                    chroms.append(snp_chr)
        sx, sy = self.panels_shape(len(chroms))
        self.fig = plt.figure(1, figsize=(sx, sy), dpi=200, facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
        ix = 1
        for c in chroms:
            likelihood = self.io[self.plot_file].get_signal(c, bin_size, &#34;SNP likelihood&#34;, snp_flag)
            img = np.array(likelihood).transpose()
            ax = plt.subplot(sx, sy, ix)
            ax.set_title(c, position=(0.01, 0.9), fontdict={&#39;verticalalignment&#39;: &#39;top&#39;, &#39;horizontalalignment&#39;: &#39;left&#39;},
                         color=&#39;C0&#39;)
            ax.imshow(img, aspect=&#39;auto&#39;)
            ax.xaxis.set_ticklabels([])
            ax.yaxis.set_ticklabels([])
            ax.xaxis.set_ticks(np.arange(0, likelihood.shape[0], 50), [])
            ix += 1
        plt.subplots_adjust(bottom=0., top=1., wspace=0, hspace=0, left=0., right=1.)
        if self.output_filename != &#34;&#34;:
            plt.savefig(self.image_filename(&#34;gview&#34;), dpi=150)
            plt.close(self.fig)
        elif self.interactive:
            plt.show(block=False)
            plt.draw()
        else:
            plt.show()

    def baf(self):
        plt.clf()
        plt.rcParams[&#34;font.size&#34;] = 8
        self.fig = plt.figure(1, figsize=(12, 8), dpi=90, facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
        chroms = []
        if self.reference_genome is None:
            chroms = self.io[self.plot_file].snp_chromosomes()
        else:
            for c, (l, t) in self.reference_genome[&#34;chromosomes&#34;].items():
                snp_chr = self.io[self.plot_file].snp_chromosome_name(c)
                if self.io[self.plot_file].signal_exists(snp_chr, None, &#34;SNP pos&#34;, 0) and \
                        self.io[self.plot_file].signal_exists(snp_chr, None, &#34;SNP desc&#34;, 0) and \
                        self.io[self.plot_file].signal_exists(snp_chr, None, &#34;SNP counts&#34;, 0) and \
                        self.io[self.plot_file].signal_exists(snp_chr, None, &#34;SNP qual&#34;, 0) and \
                        (Genome.is_autosome(c) or Genome.is_sex_chrom(c)):
                    chroms.append(snp_chr)
        sx, sy = self.panels_shape(len(chroms))
        ix = 1
        for c in chroms:
            pos, ref, alt, nref, nalt, gt, flag, qual = self.io[self.plot_file].read_snp(c)
            hpos = []
            baf = []
            color = []
            for i in range(len(pos)):
                if (nref[i] + nalt[i]) != 0:
                    hpos.append(pos[i])
                    if gt[i] % 4 != 2:
                        baf.append(1.0 * nalt[i] / (nref[i] + nalt[i]))
                    else:
                        baf.append(1.0 * nref[i] / (nref[i] + nalt[i]))
                    color.append(self.baf_colors[(gt[i] % 4, flag[i] &gt;&gt; 1)])

            ax = plt.subplot(sx, sy, ix)

            ax.set_title(c, position=(0.01, 0.9), fontdict={&#39;verticalalignment&#39;: &#39;top&#39;, &#39;horizontalalignment&#39;: &#39;left&#39;},
                         color=&#39;C0&#39;)
            ax.xaxis.set_ticklabels([])
            ax.yaxis.set_ticklabels([])
            ax.yaxis.set_ticks([0, 0.25, 0.5, 0.75, 1.0], [])
            l = max(pos)
            ax.xaxis.set_ticks(np.arange(0, (l + 10e6), 10e6), [])
            ax.set_ylim([0., 1.])
            ax.set_xlim([-0.05 * l, 1.05 * l])
            ax.grid()
            plt.scatter(hpos, baf, marker=&#39;.&#39;, edgecolor=color, c=color, s=10, alpha=0.7)
            ix += 1
        plt.subplots_adjust(bottom=0., top=1., wspace=0, hspace=0, left=0., right=1.)
        if self.output_filename != &#34;&#34;:
            plt.savefig(self.image_filename(&#34;gview&#34;), dpi=150)
            plt.close(self.fig)
        elif self.interactive:
            plt.show(block=False)
            plt.draw()
        else:
            plt.show()

    def panels_shape(self, n):
        sx, sy = 1, 1
        if n == 2:
            sx = 2
        elif n in [3, 4]:
            sx, sy = 2, 2
        elif n in [5, 6]:
            sx, sy = 3, 2
        elif n in [7, 8, 9]:
            sx, sy = 3, 3
        elif n in [10, 11, 12]:
            sx, sy = 4, 3
        elif n in [13, 14, 15, 16]:
            sx, sy = 4, 4
        elif n in [17, 18, 19, 20]:
            sx, sy = 5, 4
        elif n in [21, 22, 23, 24]:
            sx, sy = 6, 4
        else:
            while sx * sy &lt; n:
                sx += 1
                sy = int(2. * sx / 3 + 1.)
        return sx, sy

    def manhattan(self, bin_size, use_mask=False):
        plt.clf()
        if self.reference_genome is None:
            _logger.warning(&#34;Missing reference genome required for gview.&#34;)
            return
        chroms = []
        for c, (l, t) in self.reference_genome[&#34;chromosomes&#34;].items():
            rd_chr = self.io[self.plot_file].rd_chromosome_name(c)
            if self.io[self.plot_file].signal_exists(rd_chr, bin_size, &#34;RD&#34;, 0) and \
                    self.io[self.plot_file].signal_exists(rd_chr, bin_size, &#34;RD&#34;, FLAG_GC_CORR) and \
                    (Genome.is_autosome(c) or Genome.is_sex_chrom(c)):
                chroms.append((rd_chr, l))

        ix = 1
        apos = 0
        xticks = [0]

        plt.rcParams[&#34;font.size&#34;] = 8
        self.fig = plt.figure(1, figsize=(12, 4), dpi=90, facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
        ax = plt.gca()
        max_m = 0
        for c, l in chroms:
            flag = FLAG_MT if Genome.is_mt_chrom(c) else FLAG_SEX if Genome.is_sex_chrom(c) else FLAG_AUTO
            stat = self.io[self.plot_file].get_signal(None, bin_size, &#34;RD stat&#34;, flag)
            if stat[4] &gt; max_m:
                max_m = stat[4]
            flag_rd = 0
            if use_mask:
                flag_rd = FLAG_USEMASK
            his_p = self.io[self.plot_file].get_signal(c, bin_size, &#34;RD&#34;, flag_rd)
            his_p_corr = self.io[self.plot_file].get_signal(c, bin_size, &#34;RD&#34;, flag_rd | FLAG_GC_CORR)
            pos = range(apos, apos + len(his_p))
            ax.text(apos + len(his_p) // 2, stat[4] // 10, Genome.canonical_chrom_name(c),
                    fontsize=12, verticalalignment=&#39;bottom&#39;, horizontalalignment=&#39;center&#39;, )
            plt.plot(pos, his_p_corr, ls=&#39;&#39;, marker=&#39;.&#39;)
            apos += len(his_p)
            xticks.append(apos)

        ax.xaxis.set_ticklabels([])
        ax.yaxis.set_ticklabels([])
        ax.yaxis.set_ticks(np.arange(0, 3, 0.5) * max_m, [])
        ax.xaxis.set_ticks(xticks, [])
        ax.set_ylim([0, 2 * max_m])
        n_bins = apos
        ax.set_xlim([0, n_bins])
        ax.grid()
        plt.subplots_adjust(bottom=0.05, top=0.95, wspace=0, hspace=0, left=0.05, right=0.95)

        if self.output_filename != &#34;&#34;:
            plt.savefig(self.image_filename(&#34;manhattan&#34;), dpi=150)
            plt.close(self.fig)
        elif self.interactive:
            plt.show(block=False)
            plt.draw()
        else:
            plt.show()

    def regions(self, bin_size, regions, panels=[&#34;rd&#34;], use_mask_rd=False, sep_color=&#34;g&#34;):
        plt.clf()
        plt.rcParams[&#34;font.size&#34;] = 8
        self.fig = plt.figure(1, figsize=(12, 8), facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
        grid = gridspec.GridSpec(len(self.io), len(regions), wspace=0.2, hspace=0.2)
        ix = 0
        for i in self.io:
            for r in regions:
                self.region(i, grid[ix], bin_size, r, panels=panels, use_mask_rd=use_mask_rd, sep_color=sep_color)
                ix += 1
        plt.subplots_adjust(bottom=0.05, top=0.95, wspace=0, hspace=0, left=0.05, right=0.95)

        if self.output_filename != &#34;&#34;:
            plt.savefig(self.image_filename(&#34;regions&#34;), dpi=150)
            plt.close(self.fig)
        elif self.interactive:
            plt.show(block=False)
            plt.draw()
        else:
            plt.show()

    def region(self, io, element, bin_size, region, panels=[&#34;rd&#34;], use_mask_rd=False, sep_color=&#34;g&#34;):
        snp_flag = (FLAG_USEMASK if self.use_mask else 0) | (FLAG_USEID if self.use_id else 0)
        grid = gridspec.GridSpecFromSubplotSpec(len(panels), 1, subplot_spec=element, wspace=0, hspace=0.1)
        r = decode_region(region)
        for i in range(len(panels)):
            ax = self.fig.add_subplot(grid[i])
            if i == 0:
                ax.set_title(io.filename + &#34;: &#34; + region, position=(0.01, 0.9),
                             fontdict={&#39;verticalalignment&#39;: &#39;top&#39;, &#39;horizontalalignment&#39;: &#39;left&#39;},
                             color=&#39;C0&#39;)
            g_p = []
            g_p_corr = []
            if panels[i] == &#34;rd&#34;:
                mean, stdev = 0, 0
                borders = []
                for c, (pos1, pos2) in r:
                    flag = FLAG_MT if Genome.is_mt_chrom(c) else FLAG_SEX if Genome.is_sex_chrom(c) else FLAG_AUTO
                    flag_rd = 0
                    if use_mask_rd:
                        flag_rd = FLAG_USEMASK
                    stat = io.get_signal(None, bin_size, &#34;RD stat&#34;, flag)
                    mean = stat[4]
                    stdev = stat[5]
                    his_p = io.get_signal(c, bin_size, &#34;RD&#34;, flag_rd)
                    his_p_corr = io.get_signal(c, bin_size, &#34;RD&#34;, flag_rd | FLAG_GC_CORR)
                    start_bin = (pos1 - 1) // bin_size
                    end_bin = pos2 // bin_size
                    g_p.extend(list(his_p[start_bin:end_bin]))
                    g_p_corr.extend(list(his_p_corr[start_bin:end_bin]))
                    borders.append(len(g_p))

                # ax.xaxis.set_ticklabels([])
                ax.yaxis.set_ticklabels([])
                l = len(g_p)
                # ax.xaxis.set_ticks(np.arange(0, l, 10), [])
                ax.yaxis.set_ticks(np.arange(0, 3, 0.5) * mean, [])
                ax.set_ylim([0, max(3. * mean, mean + 5. * stdev)])
                ax.set_xlim([-l * 0.0, l * 1.0])

                ax.yaxis.grid()
                ax.step(g_p, &#34;grey&#34;)
                ax.step(g_p_corr, &#34;k&#34;)
                for i in borders[:-1]:
                    ax.axvline(i, color=sep_color, lw=1)
                self.fig.add_subplot(ax)
            elif panels[i] == &#34;baf&#34;:
                borders = []
                hpos = []
                baf = []
                color = []
                start_pos = 0
                for c, (pos1, pos2) in r:
                    pos, ref, alt, nref, nalt, gt, flag, qual = io.read_snp(c)
                    ix = 0
                    while ix &lt; len(pos) and pos[ix] &lt;= pos2:
                        if pos[ix] &gt;= pos1 and (nref[ix] + nalt[ix]) != 0:
                            hpos.append(start_pos + pos[ix] - pos1)
                            if gt[ix] % 4 != 2:
                                baf.append(1.0 * nalt[ix] / (nref[ix] + nalt[ix]))
                            else:
                                baf.append(1.0 * nref[ix] / (nref[ix] + nalt[ix]))
                            color.append(self.baf_colors[(gt[ix] % 4, flag[ix] &gt;&gt; 1)])
                        ix += 1
                    start_pos += pos2 - pos1
                    borders.append(start_pos)

                ax.xaxis.set_ticklabels([])
                ax.yaxis.set_ticklabels([])
                ax.yaxis.set_ticks([0, 0.25, 0.5, 0.75, 1.0], [])
                l = max(hpos)
                # ax.xaxis.set_ticks(np.arange(0, (l + 10e6), 10e6), [])
                ax.set_ylim([0., 1.])
                ax.set_xlim([0, borders[-1]])
                ax.yaxis.grid()
                ax.scatter(hpos, baf, marker=&#39;.&#39;, edgecolor=color, c=color, s=10, alpha=0.7)

                for i in borders[:-1]:
                    ax.axvline(i, color=sep_color, lw=1)
                self.fig.add_subplot(ax)

            elif panels[i] == &#34;likelihood&#34;:
                borders = []
                gl = []
                for c, (pos1, pos2) in r:
                    likelihood = io.get_signal(c, bin_size, &#34;SNP likelihood&#34;, snp_flag)
                    start_bin = (pos1 - 1) // bin_size
                    end_bin = pos2 // bin_size
                    gl.extend(list(likelihood[start_bin:end_bin]))
                    borders.append(len(gl))
                img = np.array(gl).transpose()
                ax.imshow(img, aspect=&#39;auto&#39;)
                ax.xaxis.set_ticklabels([])
                ax.yaxis.set_ticklabels([])
                ax.xaxis.set_ticks(np.arange(0, len(gl), 50), [])

                for i in borders[:-1]:
                    ax.axvline(i, color=sep_color, lw=1)
                self.fig.add_subplot(ax)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="cnvpytor.viewer.Viewer"><code class="flex name class">
<span>class <span class="ident">Viewer</span></span>
<span>(</span><span>files, output_filename)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">class Viewer:
    def __init__(self, files, output_filename):
        _logger.debug(&#34;Viewer class init: files [%s], png_prefix &#39;%s&#39;.&#34; % (&#34;, &#34;.join(files), output_filename))
        self.io = [IO(f, ro=True) for f in files]
        self.output_filename = output_filename
        self.set_style(&#39;classic&#39;)
        self.io_gc = self.io[0]
        self.io_mask = self.io[0]
        self.reference_genome = None
        self.interactive = False
        self.plot_files = [True for i in files]
        self.fig = None
        self.xkcd = False
        self.grid = &#34;auto&#34;
        self.plot_file = 0
        self.use_mask = True
        self.use_id = False
        self.baf_colors = {(0, 0): &#34;yellow&#34;, (0, 1): &#34;orange&#34;, (1, 0): &#34;cyan&#34;, (1, 1): &#34;blue&#34;, (2, 0): &#34;lime&#34;,
                           (2, 1): &#34;green&#34;, (3, 0): &#34;yellow&#34;, (3, 1): &#34;orange&#34;}
        if self.io[0].signal_exists(None, None, &#34;reference genome&#34;):
            rg_name = np.array(self.io[0].get_signal(None, None, &#34;reference genome&#34;)).astype(&#34;str&#34;)[0]
            self.reference_genome = Genome.reference_genomes[rg_name]
            if &#34;mask_file&#34; in Genome.reference_genomes[rg_name]:
                self.io_mask = IO(Genome.reference_genomes[rg_name][&#34;mask_file&#34;], ro=True, buffer=True)
            if &#34;gc_file&#34; in Genome.reference_genomes[rg_name]:
                self.io_gc = IO(Genome.reference_genomes[rg_name][&#34;gc_file&#34;], ro=True, buffer=True)

    def parse(self, command, args):
        current = &#34;regions&#34;
        regions = []

        for p in command:
            if p.isdigit() and (int(p) % 100) == 0:
                if current == &#34;rd&#34;:
                    self.rd(int(p), args.use_mask_with_rd)
                if current == &#34;likelihood&#34;:
                    self.likelihood(int(p))
                elif current == &#34;manhattan&#34;:
                    self.manhattan(int(p), args.use_mask_with_rd)
                elif current == &#34;stat&#34;:
                    self.stat(int(p))
                elif current == &#34;regions&#34;:
                    self.regions(int(p), regions, panels=args.panels)
                    regions = []
            elif p == &#34;rdstat&#34;:
                self.stat()
            elif p == &#34;baf&#34;:
                self.baf()
            elif p in [&#34;rd&#34;, &#34;manhattan&#34;, &#34;stat&#34;, &#34;regions&#34;, &#34;likelihood&#34;]:
                current = p
            elif current == &#34;regions&#34;:
                regions.append(p)
            else:
                current = p

    def plot(self, args):
        self.interactive = False
        self.use_mask = not args.no_mask
        self.use_id = args.use_id
        self.parse(args.plot, args)

    def prompt(self, bin_size, args):
        self.interactive = True
        self.use_mask = not args.no_mask
        self.use_id = args.use_id
        command_tree = {&#34;set&#34;: {&#34;bin_size&#34;: None,
                                &#34;panels&#34;: None,
                                &#34;use_mask_rd&#34;: None,
                                &#34;plot_files&#34;: None,
                                &#34;plot_file&#34;: None,
                                &#34;grid&#34;: None,
                                &#34;use_mask&#34;:None,
                                &#34;use_id&#34;:None,
                                &#34;xkcd&#34;: None,
                                &#34;style&#34;: {}
                                },
                        &#34;unset&#34;: {&#34;use_mask_rd&#34;: None,
                                  &#34;xkcd&#34;: None
                                  },
                        &#34;save&#34;: None, &#34;show&#34;: None, &#34;quit&#34;: None, &#34;rd&#34;: None, &#34;likelihood&#34;: None, &#34;baf&#34;: None,
                        &#34;stat&#34;: None, &#34;rdstat&#34;: None,
                        }
        chromosomes = set({})
        for f in self.io:
            chromosomes = chromosomes.union(set(f.rd_chromosomes()))
            chromosomes = chromosomes.union(set(f.snp_chromosomes()))
        for c in chromosomes:
            command_tree[c] = None
        command_tree[&#34;set&#34;][&#34;style&#34;] = dict(zip(plt.style.available, [None] * len(plt.style.available)))
        readline.parse_and_bind(&#34;tab: complete&#34;)
        completer = PromptCompleter(command_tree)
        readline.set_completer(completer.complete)
        quit = False
        try:
            while not quit:
                try:
                    line = raw_input(&#34;cnvpytor&gt; &#34;)
                except NameError:
                    line = input(&#34;cnvpytor&gt; &#34;)

                pre = line.split(&#34;&gt;&#34;)
                f = pre[0].strip().split(&#34; &#34;)
                n = len(f)
                if len(line) == 0:
                    continue
                elif f[0] == &#34;quit&#34;:
                    quit = True
                elif line[0] == &#34;|&#34;:
                    try:
                        eval(compile(line[1:], &#39;&lt;string&gt;&#39;, &#39;single&#39;))
                    except Exception as e:
                        print(traceback.format_exc())
                elif f[0] == &#34;save&#34; and n &gt; 1:
                    plt.savefig(f[1])
                elif f[0] in [&#34;draw&#34;, &#34;repaint&#34;, &#34;update&#34;] and n == 1:
                    self.fig.canvas.draw()
                elif f[0] == &#34;show&#34;:
                    if n == 1:
                        print(&#34;    Parameters&#34;)
                        print(&#34;        * bin_size:&#34;, bin_size)
                        print(&#34;        * panels:&#34;, args.panels)
                        print(&#34;        * use_mask_rd:&#34;, args.use_mask_with_rd)
                        print(&#34;        * use_mask:&#34;, self.use_mask)
                        print(&#34;        * use_id:&#34;, self.use_id)
                        print(&#34;        * plot_files:&#34;, list(zip(map(lambda x: x.filename, self.io), self.plot_files)))
                        print(&#34;        * plot_file:&#34;, self.plot_file)
                        print(&#34;        * grid:&#34;, self.grid)
                        print(&#34;    Available plot styles:&#34;, &#34;, &#34;.join(plt.style.available))
                elif f[0] == &#34;set&#34;:
                    if n &gt; 2 and f[1] == &#34;bin_size&#34;:
                        try:
                            bin_size = binsize_type(f[2])
                        except (ArgumentTypeError, ValueError):
                            _logger.warning(&#34;bin_size should be intiger divisible by 100&#34;)
                    elif n &gt; 3 and f[1] == &#34;grid&#34; and f[2].isdigit() and f[3].isdigit():
                        self.grid = (int(f[2]), int(f[3]))
                    elif n &gt; 2 and f[1] == &#34;plot_file&#34; and f[2].isdigit() and int(f[2]) &lt; len(self.io):
                        self.plot_file = int(f[2])
                    elif n &gt; 2 and f[1] == &#34;plot_files&#34;:
                        select = [int(i) for i in f[2:] if i.isdigit()]
                        self.plot_files = list([ix in select for ix in range(len(self.io))])
                    elif n &gt; 2 and f[1] == &#34;panels&#34;:
                        args.panels = f[2:]
                    elif n &gt; 2 and f[1] == &#34;style&#34;:
                        self.set_style(f[2])
                        if self.fig:
                            self.fig.canvas.draw()
                    elif n &gt; 1 and f[1] == &#34;use_mask_rd&#34;:
                        args.use_mask_with_rd = True
                    elif n &gt; 1 and f[1] == &#34;use_mask&#34;:
                        self.use_mask = True
                    elif n &gt; 1 and f[1] == &#34;use_id&#34;:
                        self.use_id = True
                    elif n &gt; 1 and f[1] == &#34;xkcd&#34;:
                        self.set_style(&#34;classic&#34;)
                        from matplotlib import patheffects
                        plt.xkcd()
                        plt.rcParams[&#34;path.effects&#34;] = [patheffects.withStroke(linewidth=0.5, foreground=&#34;w&#34;)]
                        self.xkcd = True
                    else:
                        print(&#34;Unrecognized set argument!&#34;)
                elif f[0] == &#34;unset&#34;:
                    if n &gt; 1 and f[1] == &#34;use_mask_rd&#34;:
                        args.use_mask_with_rd = False
                    elif n &gt; 1 and f[1] == &#34;use_mask&#34;:
                        self.use_mask = False
                    elif n &gt; 1 and f[1] == &#34;use_id&#34;:
                        self.use_id = False
                    elif n &gt; 1 and f[1] == &#34;xkcd&#34; and self.xkcd:
                        plt.rcdefaults()
                        self.set_style(&#39;classic&#39;)
                        self.xkcd = False

                    else:
                        print(&#34;Unrecognized unset argument!&#34;)
                else:

                    try:
                        if f[0] not in [&#34;rdstat&#34;, &#34;baf&#34;]:
                            self.parse(f + [str(bin_size)], args)
                        else:
                            self.parse(f, args)
                        if len(pre) &gt; 1:
                            fns = pre[1].strip().split(&#34; &#34;)
                            if fns[0] != &#34;&#34;:
                                plt.savefig(fns[0], dpi=200)
                    except Exception as e:
                        print(traceback.format_exc())
        except (EOFError, KeyboardInterrupt):
            print()
            return

    @staticmethod
    def set_style(style):
        if style in plt.style.available:
            plt.style.use(&#34;default&#34;)
            plt.style.use(style)

    def image_filename(self, sufix):
        parts = self.output_filename.split(&#34;.&#34;)
        if parts[-1] != &#34;png&#34; and parts[-1] != &#34;pdf&#34; and parts[-1] != &#34;jpg&#34; and parts[-1] != &#34;eps&#34; and parts[
            -1] != &#34;svg&#34;:
            _logger.warning(&#34;File extension should be: .jpg, .png, .svg, .eps or .pdf&#34;)
            exit(0)
        parts[-1] = sufix + &#34;.&#34; + parts[-1]
        return &#34;.&#34;.join(parts)

    def stat(self, his_bin_size=100):
        plt.clf()
        auto = self.io[0].signal_exists(None, his_bin_size, &#34;RD stat&#34;, FLAG_AUTO)
        sex = self.io[0].signal_exists(None, his_bin_size, &#34;RD stat&#34;, FLAG_SEX)
        mt = self.io[0].signal_exists(None, his_bin_size, &#34;RD stat&#34;, FLAG_MT) and (his_bin_size &lt; 1001)
        if not (auto or sex or mt):
            return
        cond = [auto, sex, mt]
        stat_list = []
        n_cols = sum(map(int, cond))
        ix = 1
        plt.rcParams[&#34;font.size&#34;] = 8
        self.fig = plt.figure(1, figsize=(4 * n_cols, 8), dpi=90, facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
        for t, c, flag in zip([&#34;Autosomes&#34;, &#34;X/Y&#34;, &#34;Mitochondria&#34;], cond, [FLAG_AUTO, FLAG_SEX, FLAG_MT]):
            if c:
                stat = self.io[0].get_signal(None, his_bin_size, &#34;RD stat&#34;, flag)
                stat_list.append(stat)
                max_rd = int(stat[0])
                bin_size = int(stat[1])
                n_bins = int(stat[2])
                lim_rd = int(max(2 * stat[4], stat[4] + 3 * stat[5]))
                _logger.info(&#34;RD stat for %s: %.2f +- %.2f&#34; % (t, stat[4], stat[5]))
                if t == &#34;Mitochondria&#34; and auto:
                    _logger.info(&#34;RD stat for %s - number of miochondria per cell: %.2f +- %.2f&#34; % (
                        t, 2 * stat[4] / stat_list[0][4],
                        2 * stat[5] / stat_list[0][4] + stat_list[0][5] * stat[4] / (
                                stat_list[0][4] * stat_list[0][4])))
                his_p = self.io[0].get_signal(None, his_bin_size, &#34;RD p dist&#34;, flag)
                his_u = self.io[0].get_signal(None, his_bin_size, &#34;RD u dist&#34;, flag)
                his_rd_gc = self.io[0].get_signal(None, his_bin_size, &#34;RD GC dist&#34;, flag)
                gc_corr = self.io[0].get_signal(None, his_bin_size, &#34;GC corr&#34;, flag)
                ax = plt.subplot(2, n_cols, ix)
                ax.set_xlabel(&#34;RD&#34;)
                ax.set_ylabel(&#34;GC [%]&#34;)
                ax.xaxis.set_ticklabels([])
                ax.set_title(t)
                his_rd_gc[0][0] = 0
                ax.imshow(his_rd_gc[:lim_rd // bin_size, :].T, aspect=&#34;auto&#34;, interpolation=&#39;nearest&#39;, origin=&#39;lower&#39;)
                ax.plot(gc_corr * stat[4] / bin_size, range(101), &#34;w-&#34;)

                ax = plt.subplot(2, n_cols, ix + n_cols)
                ax.set_ylabel(&#34;Normalised distribution&#34;)
                ax.set_xlabel(&#34;RD&#34;)
                ax.set_xlim([0, lim_rd])
                ax.set_ylim([0, 1.1])
                bins = range(0, max_rd, bin_size)
                x = np.arange(0, max_rd // bin_size * bin_size, 0.1 * bin_size)
                plt.plot(x, normal(x, 1, stat[4], stat[5]), &#34;g-&#34;)
                x = np.array(bins)
                plt.plot(x[:len(his_u)], his_u / stat[3], &#34;y*&#34;)
                plt.plot(x[:len(his_p)], his_p / stat[3], &#34;b*&#34;)
                ix += 1
        plt.subplots_adjust(bottom=0.08, top=0.95, wspace=0.25, hspace=0, left=0.05 * 3 / n_cols, right=0.95)
        if self.output_filename != &#34;&#34;:
            plt.savefig(self.image_filename(&#34;stat&#34;), dpi=150)
            plt.close(self.fig)
        elif self.interactive:
            plt.show(block=False)
            plt.draw()
        else:
            plt.show()

    def rd(self, bin_size, use_mask):
        plt.clf()
        if self.reference_genome is None:
            _logger.warning(&#34;Missing reference genome required for gview.&#34;)
            return
        plt.rcParams[&#34;font.size&#34;] = 8
        chroms = []
        for c, (l, t) in self.reference_genome[&#34;chromosomes&#34;].items():
            rd_chr = self.io[self.plot_file].rd_chromosome_name(c)
            if self.io[self.plot_file].signal_exists(rd_chr, bin_size, &#34;RD&#34;, 0) and \
                    self.io[self.plot_file].signal_exists(rd_chr, bin_size, &#34;RD&#34;, FLAG_GC_CORR) and \
                    (Genome.is_autosome(c) or Genome.is_sex_chrom(c)):
                chroms.append((rd_chr, l))
        sx, sy = self.panels_shape(len(chroms))
        self.fig = plt.figure(1, figsize=(sx, sy), dpi=200, facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
        ix = 1
        for c, l in chroms:
            flag = FLAG_MT if Genome.is_mt_chrom(c) else FLAG_SEX if Genome.is_sex_chrom(c) else FLAG_AUTO
            stat = self.io[self.plot_file].get_signal(None, bin_size, &#34;RD stat&#34;, flag)
            if stat is None:
                _logger.error(&#34;Data for bin size %d is missing in file &#39;%s&#39;!&#34; % (bin_size, self.io[self.plot_file].filename))
                exit(0)
            flag_rd = 0
            if use_mask:
                flag_rd = FLAG_USEMASK
            his_p = self.io[self.plot_file].get_signal(c, bin_size, &#34;RD&#34;, flag_rd)
            his_p_corr = self.io[self.plot_file].get_signal(c, bin_size, &#34;RD&#34;, flag_rd | FLAG_GC_CORR)
            ax = plt.subplot(sx, sy, ix)
            ax.set_title(c, position=(0.01, 0.9), fontdict={&#39;verticalalignment&#39;: &#39;top&#39;, &#39;horizontalalignment&#39;: &#39;left&#39;},
                         color=&#39;C0&#39;)
            ax.xaxis.set_ticklabels([])
            ax.yaxis.set_ticklabels([])
            ax.yaxis.set_ticks(np.arange(0, 3, 0.5) * stat[4], [])
            ax.xaxis.set_ticks(np.arange(0, (l + 10e6) // bin_size, 10e6 // bin_size), [])
            ax.set_ylim([0, max(3. * stat[4], stat[4] + 5. * stat[5])])
            n_bins = l // bin_size
            ax.set_xlim([-n_bins * 0.05, n_bins * 1.05])
            ax.grid()

            plt.step(his_p, &#34;grey&#34;)
            plt.step(his_p_corr, &#34;k&#34;)
            ix += 1
        plt.subplots_adjust(bottom=0., top=1., wspace=0, hspace=0, left=0., right=1.)
        if self.output_filename != &#34;&#34;:
            plt.savefig(self.image_filename(&#34;gview&#34;), dpi=150)
            plt.close(self.fig)
        elif self.interactive:
            plt.show(block=False)
            plt.draw()
        else:
            plt.show()

    def likelihood(self, bin_size):
        plt.clf()
        snp_flag = (FLAG_USEMASK if self.use_mask else 0) | (FLAG_USEID if self.use_id else 0)
        if self.reference_genome is None:
            _logger.warning(&#34;Missing reference genome required for gview.&#34;)
            return
        plt.rcParams[&#34;font.size&#34;] = 8
        chroms = []
        if self.reference_genome is None:
            chroms = self.io[self.plot_file].snp_chromosomes()
        else:
            for c, (l, t) in self.reference_genome[&#34;chromosomes&#34;].items():
                snp_chr = self.io[self.plot_file].snp_chromosome_name(c)
                if self.io[self.plot_file].signal_exists(snp_chr, bin_size, &#34;SNP likelihood&#34;, snp_flag) and (
                        Genome.is_autosome(c) or Genome.is_sex_chrom(c)):
                    chroms.append(snp_chr)
        sx, sy = self.panels_shape(len(chroms))
        self.fig = plt.figure(1, figsize=(sx, sy), dpi=200, facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
        ix = 1
        for c in chroms:
            likelihood = self.io[self.plot_file].get_signal(c, bin_size, &#34;SNP likelihood&#34;, snp_flag)
            img = np.array(likelihood).transpose()
            ax = plt.subplot(sx, sy, ix)
            ax.set_title(c, position=(0.01, 0.9), fontdict={&#39;verticalalignment&#39;: &#39;top&#39;, &#39;horizontalalignment&#39;: &#39;left&#39;},
                         color=&#39;C0&#39;)
            ax.imshow(img, aspect=&#39;auto&#39;)
            ax.xaxis.set_ticklabels([])
            ax.yaxis.set_ticklabels([])
            ax.xaxis.set_ticks(np.arange(0, likelihood.shape[0], 50), [])
            ix += 1
        plt.subplots_adjust(bottom=0., top=1., wspace=0, hspace=0, left=0., right=1.)
        if self.output_filename != &#34;&#34;:
            plt.savefig(self.image_filename(&#34;gview&#34;), dpi=150)
            plt.close(self.fig)
        elif self.interactive:
            plt.show(block=False)
            plt.draw()
        else:
            plt.show()

    def baf(self):
        plt.clf()
        plt.rcParams[&#34;font.size&#34;] = 8
        self.fig = plt.figure(1, figsize=(12, 8), dpi=90, facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
        chroms = []
        if self.reference_genome is None:
            chroms = self.io[self.plot_file].snp_chromosomes()
        else:
            for c, (l, t) in self.reference_genome[&#34;chromosomes&#34;].items():
                snp_chr = self.io[self.plot_file].snp_chromosome_name(c)
                if self.io[self.plot_file].signal_exists(snp_chr, None, &#34;SNP pos&#34;, 0) and \
                        self.io[self.plot_file].signal_exists(snp_chr, None, &#34;SNP desc&#34;, 0) and \
                        self.io[self.plot_file].signal_exists(snp_chr, None, &#34;SNP counts&#34;, 0) and \
                        self.io[self.plot_file].signal_exists(snp_chr, None, &#34;SNP qual&#34;, 0) and \
                        (Genome.is_autosome(c) or Genome.is_sex_chrom(c)):
                    chroms.append(snp_chr)
        sx, sy = self.panels_shape(len(chroms))
        ix = 1
        for c in chroms:
            pos, ref, alt, nref, nalt, gt, flag, qual = self.io[self.plot_file].read_snp(c)
            hpos = []
            baf = []
            color = []
            for i in range(len(pos)):
                if (nref[i] + nalt[i]) != 0:
                    hpos.append(pos[i])
                    if gt[i] % 4 != 2:
                        baf.append(1.0 * nalt[i] / (nref[i] + nalt[i]))
                    else:
                        baf.append(1.0 * nref[i] / (nref[i] + nalt[i]))
                    color.append(self.baf_colors[(gt[i] % 4, flag[i] &gt;&gt; 1)])

            ax = plt.subplot(sx, sy, ix)

            ax.set_title(c, position=(0.01, 0.9), fontdict={&#39;verticalalignment&#39;: &#39;top&#39;, &#39;horizontalalignment&#39;: &#39;left&#39;},
                         color=&#39;C0&#39;)
            ax.xaxis.set_ticklabels([])
            ax.yaxis.set_ticklabels([])
            ax.yaxis.set_ticks([0, 0.25, 0.5, 0.75, 1.0], [])
            l = max(pos)
            ax.xaxis.set_ticks(np.arange(0, (l + 10e6), 10e6), [])
            ax.set_ylim([0., 1.])
            ax.set_xlim([-0.05 * l, 1.05 * l])
            ax.grid()
            plt.scatter(hpos, baf, marker=&#39;.&#39;, edgecolor=color, c=color, s=10, alpha=0.7)
            ix += 1
        plt.subplots_adjust(bottom=0., top=1., wspace=0, hspace=0, left=0., right=1.)
        if self.output_filename != &#34;&#34;:
            plt.savefig(self.image_filename(&#34;gview&#34;), dpi=150)
            plt.close(self.fig)
        elif self.interactive:
            plt.show(block=False)
            plt.draw()
        else:
            plt.show()

    def panels_shape(self, n):
        sx, sy = 1, 1
        if n == 2:
            sx = 2
        elif n in [3, 4]:
            sx, sy = 2, 2
        elif n in [5, 6]:
            sx, sy = 3, 2
        elif n in [7, 8, 9]:
            sx, sy = 3, 3
        elif n in [10, 11, 12]:
            sx, sy = 4, 3
        elif n in [13, 14, 15, 16]:
            sx, sy = 4, 4
        elif n in [17, 18, 19, 20]:
            sx, sy = 5, 4
        elif n in [21, 22, 23, 24]:
            sx, sy = 6, 4
        else:
            while sx * sy &lt; n:
                sx += 1
                sy = int(2. * sx / 3 + 1.)
        return sx, sy

    def manhattan(self, bin_size, use_mask=False):
        plt.clf()
        if self.reference_genome is None:
            _logger.warning(&#34;Missing reference genome required for gview.&#34;)
            return
        chroms = []
        for c, (l, t) in self.reference_genome[&#34;chromosomes&#34;].items():
            rd_chr = self.io[self.plot_file].rd_chromosome_name(c)
            if self.io[self.plot_file].signal_exists(rd_chr, bin_size, &#34;RD&#34;, 0) and \
                    self.io[self.plot_file].signal_exists(rd_chr, bin_size, &#34;RD&#34;, FLAG_GC_CORR) and \
                    (Genome.is_autosome(c) or Genome.is_sex_chrom(c)):
                chroms.append((rd_chr, l))

        ix = 1
        apos = 0
        xticks = [0]

        plt.rcParams[&#34;font.size&#34;] = 8
        self.fig = plt.figure(1, figsize=(12, 4), dpi=90, facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
        ax = plt.gca()
        max_m = 0
        for c, l in chroms:
            flag = FLAG_MT if Genome.is_mt_chrom(c) else FLAG_SEX if Genome.is_sex_chrom(c) else FLAG_AUTO
            stat = self.io[self.plot_file].get_signal(None, bin_size, &#34;RD stat&#34;, flag)
            if stat[4] &gt; max_m:
                max_m = stat[4]
            flag_rd = 0
            if use_mask:
                flag_rd = FLAG_USEMASK
            his_p = self.io[self.plot_file].get_signal(c, bin_size, &#34;RD&#34;, flag_rd)
            his_p_corr = self.io[self.plot_file].get_signal(c, bin_size, &#34;RD&#34;, flag_rd | FLAG_GC_CORR)
            pos = range(apos, apos + len(his_p))
            ax.text(apos + len(his_p) // 2, stat[4] // 10, Genome.canonical_chrom_name(c),
                    fontsize=12, verticalalignment=&#39;bottom&#39;, horizontalalignment=&#39;center&#39;, )
            plt.plot(pos, his_p_corr, ls=&#39;&#39;, marker=&#39;.&#39;)
            apos += len(his_p)
            xticks.append(apos)

        ax.xaxis.set_ticklabels([])
        ax.yaxis.set_ticklabels([])
        ax.yaxis.set_ticks(np.arange(0, 3, 0.5) * max_m, [])
        ax.xaxis.set_ticks(xticks, [])
        ax.set_ylim([0, 2 * max_m])
        n_bins = apos
        ax.set_xlim([0, n_bins])
        ax.grid()
        plt.subplots_adjust(bottom=0.05, top=0.95, wspace=0, hspace=0, left=0.05, right=0.95)

        if self.output_filename != &#34;&#34;:
            plt.savefig(self.image_filename(&#34;manhattan&#34;), dpi=150)
            plt.close(self.fig)
        elif self.interactive:
            plt.show(block=False)
            plt.draw()
        else:
            plt.show()

    def regions(self, bin_size, regions, panels=[&#34;rd&#34;], use_mask_rd=False, sep_color=&#34;g&#34;):
        plt.clf()
        plt.rcParams[&#34;font.size&#34;] = 8
        self.fig = plt.figure(1, figsize=(12, 8), facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
        grid = gridspec.GridSpec(len(self.io), len(regions), wspace=0.2, hspace=0.2)
        ix = 0
        for i in self.io:
            for r in regions:
                self.region(i, grid[ix], bin_size, r, panels=panels, use_mask_rd=use_mask_rd, sep_color=sep_color)
                ix += 1
        plt.subplots_adjust(bottom=0.05, top=0.95, wspace=0, hspace=0, left=0.05, right=0.95)

        if self.output_filename != &#34;&#34;:
            plt.savefig(self.image_filename(&#34;regions&#34;), dpi=150)
            plt.close(self.fig)
        elif self.interactive:
            plt.show(block=False)
            plt.draw()
        else:
            plt.show()

    def region(self, io, element, bin_size, region, panels=[&#34;rd&#34;], use_mask_rd=False, sep_color=&#34;g&#34;):
        snp_flag = (FLAG_USEMASK if self.use_mask else 0) | (FLAG_USEID if self.use_id else 0)
        grid = gridspec.GridSpecFromSubplotSpec(len(panels), 1, subplot_spec=element, wspace=0, hspace=0.1)
        r = decode_region(region)
        for i in range(len(panels)):
            ax = self.fig.add_subplot(grid[i])
            if i == 0:
                ax.set_title(io.filename + &#34;: &#34; + region, position=(0.01, 0.9),
                             fontdict={&#39;verticalalignment&#39;: &#39;top&#39;, &#39;horizontalalignment&#39;: &#39;left&#39;},
                             color=&#39;C0&#39;)
            g_p = []
            g_p_corr = []
            if panels[i] == &#34;rd&#34;:
                mean, stdev = 0, 0
                borders = []
                for c, (pos1, pos2) in r:
                    flag = FLAG_MT if Genome.is_mt_chrom(c) else FLAG_SEX if Genome.is_sex_chrom(c) else FLAG_AUTO
                    flag_rd = 0
                    if use_mask_rd:
                        flag_rd = FLAG_USEMASK
                    stat = io.get_signal(None, bin_size, &#34;RD stat&#34;, flag)
                    mean = stat[4]
                    stdev = stat[5]
                    his_p = io.get_signal(c, bin_size, &#34;RD&#34;, flag_rd)
                    his_p_corr = io.get_signal(c, bin_size, &#34;RD&#34;, flag_rd | FLAG_GC_CORR)
                    start_bin = (pos1 - 1) // bin_size
                    end_bin = pos2 // bin_size
                    g_p.extend(list(his_p[start_bin:end_bin]))
                    g_p_corr.extend(list(his_p_corr[start_bin:end_bin]))
                    borders.append(len(g_p))

                # ax.xaxis.set_ticklabels([])
                ax.yaxis.set_ticklabels([])
                l = len(g_p)
                # ax.xaxis.set_ticks(np.arange(0, l, 10), [])
                ax.yaxis.set_ticks(np.arange(0, 3, 0.5) * mean, [])
                ax.set_ylim([0, max(3. * mean, mean + 5. * stdev)])
                ax.set_xlim([-l * 0.0, l * 1.0])

                ax.yaxis.grid()
                ax.step(g_p, &#34;grey&#34;)
                ax.step(g_p_corr, &#34;k&#34;)
                for i in borders[:-1]:
                    ax.axvline(i, color=sep_color, lw=1)
                self.fig.add_subplot(ax)
            elif panels[i] == &#34;baf&#34;:
                borders = []
                hpos = []
                baf = []
                color = []
                start_pos = 0
                for c, (pos1, pos2) in r:
                    pos, ref, alt, nref, nalt, gt, flag, qual = io.read_snp(c)
                    ix = 0
                    while ix &lt; len(pos) and pos[ix] &lt;= pos2:
                        if pos[ix] &gt;= pos1 and (nref[ix] + nalt[ix]) != 0:
                            hpos.append(start_pos + pos[ix] - pos1)
                            if gt[ix] % 4 != 2:
                                baf.append(1.0 * nalt[ix] / (nref[ix] + nalt[ix]))
                            else:
                                baf.append(1.0 * nref[ix] / (nref[ix] + nalt[ix]))
                            color.append(self.baf_colors[(gt[ix] % 4, flag[ix] &gt;&gt; 1)])
                        ix += 1
                    start_pos += pos2 - pos1
                    borders.append(start_pos)

                ax.xaxis.set_ticklabels([])
                ax.yaxis.set_ticklabels([])
                ax.yaxis.set_ticks([0, 0.25, 0.5, 0.75, 1.0], [])
                l = max(hpos)
                # ax.xaxis.set_ticks(np.arange(0, (l + 10e6), 10e6), [])
                ax.set_ylim([0., 1.])
                ax.set_xlim([0, borders[-1]])
                ax.yaxis.grid()
                ax.scatter(hpos, baf, marker=&#39;.&#39;, edgecolor=color, c=color, s=10, alpha=0.7)

                for i in borders[:-1]:
                    ax.axvline(i, color=sep_color, lw=1)
                self.fig.add_subplot(ax)

            elif panels[i] == &#34;likelihood&#34;:
                borders = []
                gl = []
                for c, (pos1, pos2) in r:
                    likelihood = io.get_signal(c, bin_size, &#34;SNP likelihood&#34;, snp_flag)
                    start_bin = (pos1 - 1) // bin_size
                    end_bin = pos2 // bin_size
                    gl.extend(list(likelihood[start_bin:end_bin]))
                    borders.append(len(gl))
                img = np.array(gl).transpose()
                ax.imshow(img, aspect=&#39;auto&#39;)
                ax.xaxis.set_ticklabels([])
                ax.yaxis.set_ticklabels([])
                ax.xaxis.set_ticks(np.arange(0, len(gl), 50), [])

                for i in borders[:-1]:
                    ax.axvline(i, color=sep_color, lw=1)
                self.fig.add_subplot(ax)</code></pre>
</details>
<h3>Static methods</h3>
<dl>
<dt id="cnvpytor.viewer.Viewer.set_style"><code class="name flex">
<span>def <span class="ident">set_style</span></span>(<span>style)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">@staticmethod
def set_style(style):
    if style in plt.style.available:
        plt.style.use(&#34;default&#34;)
        plt.style.use(style)</code></pre>
</details>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="cnvpytor.viewer.Viewer.baf"><code class="name flex">
<span>def <span class="ident">baf</span></span>(<span>self)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def baf(self):
    plt.clf()
    plt.rcParams[&#34;font.size&#34;] = 8
    self.fig = plt.figure(1, figsize=(12, 8), dpi=90, facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
    chroms = []
    if self.reference_genome is None:
        chroms = self.io[self.plot_file].snp_chromosomes()
    else:
        for c, (l, t) in self.reference_genome[&#34;chromosomes&#34;].items():
            snp_chr = self.io[self.plot_file].snp_chromosome_name(c)
            if self.io[self.plot_file].signal_exists(snp_chr, None, &#34;SNP pos&#34;, 0) and \
                    self.io[self.plot_file].signal_exists(snp_chr, None, &#34;SNP desc&#34;, 0) and \
                    self.io[self.plot_file].signal_exists(snp_chr, None, &#34;SNP counts&#34;, 0) and \
                    self.io[self.plot_file].signal_exists(snp_chr, None, &#34;SNP qual&#34;, 0) and \
                    (Genome.is_autosome(c) or Genome.is_sex_chrom(c)):
                chroms.append(snp_chr)
    sx, sy = self.panels_shape(len(chroms))
    ix = 1
    for c in chroms:
        pos, ref, alt, nref, nalt, gt, flag, qual = self.io[self.plot_file].read_snp(c)
        hpos = []
        baf = []
        color = []
        for i in range(len(pos)):
            if (nref[i] + nalt[i]) != 0:
                hpos.append(pos[i])
                if gt[i] % 4 != 2:
                    baf.append(1.0 * nalt[i] / (nref[i] + nalt[i]))
                else:
                    baf.append(1.0 * nref[i] / (nref[i] + nalt[i]))
                color.append(self.baf_colors[(gt[i] % 4, flag[i] &gt;&gt; 1)])

        ax = plt.subplot(sx, sy, ix)

        ax.set_title(c, position=(0.01, 0.9), fontdict={&#39;verticalalignment&#39;: &#39;top&#39;, &#39;horizontalalignment&#39;: &#39;left&#39;},
                     color=&#39;C0&#39;)
        ax.xaxis.set_ticklabels([])
        ax.yaxis.set_ticklabels([])
        ax.yaxis.set_ticks([0, 0.25, 0.5, 0.75, 1.0], [])
        l = max(pos)
        ax.xaxis.set_ticks(np.arange(0, (l + 10e6), 10e6), [])
        ax.set_ylim([0., 1.])
        ax.set_xlim([-0.05 * l, 1.05 * l])
        ax.grid()
        plt.scatter(hpos, baf, marker=&#39;.&#39;, edgecolor=color, c=color, s=10, alpha=0.7)
        ix += 1
    plt.subplots_adjust(bottom=0., top=1., wspace=0, hspace=0, left=0., right=1.)
    if self.output_filename != &#34;&#34;:
        plt.savefig(self.image_filename(&#34;gview&#34;), dpi=150)
        plt.close(self.fig)
    elif self.interactive:
        plt.show(block=False)
        plt.draw()
    else:
        plt.show()</code></pre>
</details>
</dd>
<dt id="cnvpytor.viewer.Viewer.image_filename"><code class="name flex">
<span>def <span class="ident">image_filename</span></span>(<span>self, sufix)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def image_filename(self, sufix):
    parts = self.output_filename.split(&#34;.&#34;)
    if parts[-1] != &#34;png&#34; and parts[-1] != &#34;pdf&#34; and parts[-1] != &#34;jpg&#34; and parts[-1] != &#34;eps&#34; and parts[
        -1] != &#34;svg&#34;:
        _logger.warning(&#34;File extension should be: .jpg, .png, .svg, .eps or .pdf&#34;)
        exit(0)
    parts[-1] = sufix + &#34;.&#34; + parts[-1]
    return &#34;.&#34;.join(parts)</code></pre>
</details>
</dd>
<dt id="cnvpytor.viewer.Viewer.likelihood"><code class="name flex">
<span>def <span class="ident">likelihood</span></span>(<span>self, bin_size)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def likelihood(self, bin_size):
    plt.clf()
    snp_flag = (FLAG_USEMASK if self.use_mask else 0) | (FLAG_USEID if self.use_id else 0)
    if self.reference_genome is None:
        _logger.warning(&#34;Missing reference genome required for gview.&#34;)
        return
    plt.rcParams[&#34;font.size&#34;] = 8
    chroms = []
    if self.reference_genome is None:
        chroms = self.io[self.plot_file].snp_chromosomes()
    else:
        for c, (l, t) in self.reference_genome[&#34;chromosomes&#34;].items():
            snp_chr = self.io[self.plot_file].snp_chromosome_name(c)
            if self.io[self.plot_file].signal_exists(snp_chr, bin_size, &#34;SNP likelihood&#34;, snp_flag) and (
                    Genome.is_autosome(c) or Genome.is_sex_chrom(c)):
                chroms.append(snp_chr)
    sx, sy = self.panels_shape(len(chroms))
    self.fig = plt.figure(1, figsize=(sx, sy), dpi=200, facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
    ix = 1
    for c in chroms:
        likelihood = self.io[self.plot_file].get_signal(c, bin_size, &#34;SNP likelihood&#34;, snp_flag)
        img = np.array(likelihood).transpose()
        ax = plt.subplot(sx, sy, ix)
        ax.set_title(c, position=(0.01, 0.9), fontdict={&#39;verticalalignment&#39;: &#39;top&#39;, &#39;horizontalalignment&#39;: &#39;left&#39;},
                     color=&#39;C0&#39;)
        ax.imshow(img, aspect=&#39;auto&#39;)
        ax.xaxis.set_ticklabels([])
        ax.yaxis.set_ticklabels([])
        ax.xaxis.set_ticks(np.arange(0, likelihood.shape[0], 50), [])
        ix += 1
    plt.subplots_adjust(bottom=0., top=1., wspace=0, hspace=0, left=0., right=1.)
    if self.output_filename != &#34;&#34;:
        plt.savefig(self.image_filename(&#34;gview&#34;), dpi=150)
        plt.close(self.fig)
    elif self.interactive:
        plt.show(block=False)
        plt.draw()
    else:
        plt.show()</code></pre>
</details>
</dd>
<dt id="cnvpytor.viewer.Viewer.manhattan"><code class="name flex">
<span>def <span class="ident">manhattan</span></span>(<span>self, bin_size, use_mask=False)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def manhattan(self, bin_size, use_mask=False):
    plt.clf()
    if self.reference_genome is None:
        _logger.warning(&#34;Missing reference genome required for gview.&#34;)
        return
    chroms = []
    for c, (l, t) in self.reference_genome[&#34;chromosomes&#34;].items():
        rd_chr = self.io[self.plot_file].rd_chromosome_name(c)
        if self.io[self.plot_file].signal_exists(rd_chr, bin_size, &#34;RD&#34;, 0) and \
                self.io[self.plot_file].signal_exists(rd_chr, bin_size, &#34;RD&#34;, FLAG_GC_CORR) and \
                (Genome.is_autosome(c) or Genome.is_sex_chrom(c)):
            chroms.append((rd_chr, l))

    ix = 1
    apos = 0
    xticks = [0]

    plt.rcParams[&#34;font.size&#34;] = 8
    self.fig = plt.figure(1, figsize=(12, 4), dpi=90, facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
    ax = plt.gca()
    max_m = 0
    for c, l in chroms:
        flag = FLAG_MT if Genome.is_mt_chrom(c) else FLAG_SEX if Genome.is_sex_chrom(c) else FLAG_AUTO
        stat = self.io[self.plot_file].get_signal(None, bin_size, &#34;RD stat&#34;, flag)
        if stat[4] &gt; max_m:
            max_m = stat[4]
        flag_rd = 0
        if use_mask:
            flag_rd = FLAG_USEMASK
        his_p = self.io[self.plot_file].get_signal(c, bin_size, &#34;RD&#34;, flag_rd)
        his_p_corr = self.io[self.plot_file].get_signal(c, bin_size, &#34;RD&#34;, flag_rd | FLAG_GC_CORR)
        pos = range(apos, apos + len(his_p))
        ax.text(apos + len(his_p) // 2, stat[4] // 10, Genome.canonical_chrom_name(c),
                fontsize=12, verticalalignment=&#39;bottom&#39;, horizontalalignment=&#39;center&#39;, )
        plt.plot(pos, his_p_corr, ls=&#39;&#39;, marker=&#39;.&#39;)
        apos += len(his_p)
        xticks.append(apos)

    ax.xaxis.set_ticklabels([])
    ax.yaxis.set_ticklabels([])
    ax.yaxis.set_ticks(np.arange(0, 3, 0.5) * max_m, [])
    ax.xaxis.set_ticks(xticks, [])
    ax.set_ylim([0, 2 * max_m])
    n_bins = apos
    ax.set_xlim([0, n_bins])
    ax.grid()
    plt.subplots_adjust(bottom=0.05, top=0.95, wspace=0, hspace=0, left=0.05, right=0.95)

    if self.output_filename != &#34;&#34;:
        plt.savefig(self.image_filename(&#34;manhattan&#34;), dpi=150)
        plt.close(self.fig)
    elif self.interactive:
        plt.show(block=False)
        plt.draw()
    else:
        plt.show()</code></pre>
</details>
</dd>
<dt id="cnvpytor.viewer.Viewer.panels_shape"><code class="name flex">
<span>def <span class="ident">panels_shape</span></span>(<span>self, n)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def panels_shape(self, n):
    sx, sy = 1, 1
    if n == 2:
        sx = 2
    elif n in [3, 4]:
        sx, sy = 2, 2
    elif n in [5, 6]:
        sx, sy = 3, 2
    elif n in [7, 8, 9]:
        sx, sy = 3, 3
    elif n in [10, 11, 12]:
        sx, sy = 4, 3
    elif n in [13, 14, 15, 16]:
        sx, sy = 4, 4
    elif n in [17, 18, 19, 20]:
        sx, sy = 5, 4
    elif n in [21, 22, 23, 24]:
        sx, sy = 6, 4
    else:
        while sx * sy &lt; n:
            sx += 1
            sy = int(2. * sx / 3 + 1.)
    return sx, sy</code></pre>
</details>
</dd>
<dt id="cnvpytor.viewer.Viewer.parse"><code class="name flex">
<span>def <span class="ident">parse</span></span>(<span>self, command, args)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def parse(self, command, args):
    current = &#34;regions&#34;
    regions = []

    for p in command:
        if p.isdigit() and (int(p) % 100) == 0:
            if current == &#34;rd&#34;:
                self.rd(int(p), args.use_mask_with_rd)
            if current == &#34;likelihood&#34;:
                self.likelihood(int(p))
            elif current == &#34;manhattan&#34;:
                self.manhattan(int(p), args.use_mask_with_rd)
            elif current == &#34;stat&#34;:
                self.stat(int(p))
            elif current == &#34;regions&#34;:
                self.regions(int(p), regions, panels=args.panels)
                regions = []
        elif p == &#34;rdstat&#34;:
            self.stat()
        elif p == &#34;baf&#34;:
            self.baf()
        elif p in [&#34;rd&#34;, &#34;manhattan&#34;, &#34;stat&#34;, &#34;regions&#34;, &#34;likelihood&#34;]:
            current = p
        elif current == &#34;regions&#34;:
            regions.append(p)
        else:
            current = p</code></pre>
</details>
</dd>
<dt id="cnvpytor.viewer.Viewer.plot"><code class="name flex">
<span>def <span class="ident">plot</span></span>(<span>self, args)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def plot(self, args):
    self.interactive = False
    self.use_mask = not args.no_mask
    self.use_id = args.use_id
    self.parse(args.plot, args)</code></pre>
</details>
</dd>
<dt id="cnvpytor.viewer.Viewer.prompt"><code class="name flex">
<span>def <span class="ident">prompt</span></span>(<span>self, bin_size, args)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def prompt(self, bin_size, args):
    self.interactive = True
    self.use_mask = not args.no_mask
    self.use_id = args.use_id
    command_tree = {&#34;set&#34;: {&#34;bin_size&#34;: None,
                            &#34;panels&#34;: None,
                            &#34;use_mask_rd&#34;: None,
                            &#34;plot_files&#34;: None,
                            &#34;plot_file&#34;: None,
                            &#34;grid&#34;: None,
                            &#34;use_mask&#34;:None,
                            &#34;use_id&#34;:None,
                            &#34;xkcd&#34;: None,
                            &#34;style&#34;: {}
                            },
                    &#34;unset&#34;: {&#34;use_mask_rd&#34;: None,
                              &#34;xkcd&#34;: None
                              },
                    &#34;save&#34;: None, &#34;show&#34;: None, &#34;quit&#34;: None, &#34;rd&#34;: None, &#34;likelihood&#34;: None, &#34;baf&#34;: None,
                    &#34;stat&#34;: None, &#34;rdstat&#34;: None,
                    }
    chromosomes = set({})
    for f in self.io:
        chromosomes = chromosomes.union(set(f.rd_chromosomes()))
        chromosomes = chromosomes.union(set(f.snp_chromosomes()))
    for c in chromosomes:
        command_tree[c] = None
    command_tree[&#34;set&#34;][&#34;style&#34;] = dict(zip(plt.style.available, [None] * len(plt.style.available)))
    readline.parse_and_bind(&#34;tab: complete&#34;)
    completer = PromptCompleter(command_tree)
    readline.set_completer(completer.complete)
    quit = False
    try:
        while not quit:
            try:
                line = raw_input(&#34;cnvpytor&gt; &#34;)
            except NameError:
                line = input(&#34;cnvpytor&gt; &#34;)

            pre = line.split(&#34;&gt;&#34;)
            f = pre[0].strip().split(&#34; &#34;)
            n = len(f)
            if len(line) == 0:
                continue
            elif f[0] == &#34;quit&#34;:
                quit = True
            elif line[0] == &#34;|&#34;:
                try:
                    eval(compile(line[1:], &#39;&lt;string&gt;&#39;, &#39;single&#39;))
                except Exception as e:
                    print(traceback.format_exc())
            elif f[0] == &#34;save&#34; and n &gt; 1:
                plt.savefig(f[1])
            elif f[0] in [&#34;draw&#34;, &#34;repaint&#34;, &#34;update&#34;] and n == 1:
                self.fig.canvas.draw()
            elif f[0] == &#34;show&#34;:
                if n == 1:
                    print(&#34;    Parameters&#34;)
                    print(&#34;        * bin_size:&#34;, bin_size)
                    print(&#34;        * panels:&#34;, args.panels)
                    print(&#34;        * use_mask_rd:&#34;, args.use_mask_with_rd)
                    print(&#34;        * use_mask:&#34;, self.use_mask)
                    print(&#34;        * use_id:&#34;, self.use_id)
                    print(&#34;        * plot_files:&#34;, list(zip(map(lambda x: x.filename, self.io), self.plot_files)))
                    print(&#34;        * plot_file:&#34;, self.plot_file)
                    print(&#34;        * grid:&#34;, self.grid)
                    print(&#34;    Available plot styles:&#34;, &#34;, &#34;.join(plt.style.available))
            elif f[0] == &#34;set&#34;:
                if n &gt; 2 and f[1] == &#34;bin_size&#34;:
                    try:
                        bin_size = binsize_type(f[2])
                    except (ArgumentTypeError, ValueError):
                        _logger.warning(&#34;bin_size should be intiger divisible by 100&#34;)
                elif n &gt; 3 and f[1] == &#34;grid&#34; and f[2].isdigit() and f[3].isdigit():
                    self.grid = (int(f[2]), int(f[3]))
                elif n &gt; 2 and f[1] == &#34;plot_file&#34; and f[2].isdigit() and int(f[2]) &lt; len(self.io):
                    self.plot_file = int(f[2])
                elif n &gt; 2 and f[1] == &#34;plot_files&#34;:
                    select = [int(i) for i in f[2:] if i.isdigit()]
                    self.plot_files = list([ix in select for ix in range(len(self.io))])
                elif n &gt; 2 and f[1] == &#34;panels&#34;:
                    args.panels = f[2:]
                elif n &gt; 2 and f[1] == &#34;style&#34;:
                    self.set_style(f[2])
                    if self.fig:
                        self.fig.canvas.draw()
                elif n &gt; 1 and f[1] == &#34;use_mask_rd&#34;:
                    args.use_mask_with_rd = True
                elif n &gt; 1 and f[1] == &#34;use_mask&#34;:
                    self.use_mask = True
                elif n &gt; 1 and f[1] == &#34;use_id&#34;:
                    self.use_id = True
                elif n &gt; 1 and f[1] == &#34;xkcd&#34;:
                    self.set_style(&#34;classic&#34;)
                    from matplotlib import patheffects
                    plt.xkcd()
                    plt.rcParams[&#34;path.effects&#34;] = [patheffects.withStroke(linewidth=0.5, foreground=&#34;w&#34;)]
                    self.xkcd = True
                else:
                    print(&#34;Unrecognized set argument!&#34;)
            elif f[0] == &#34;unset&#34;:
                if n &gt; 1 and f[1] == &#34;use_mask_rd&#34;:
                    args.use_mask_with_rd = False
                elif n &gt; 1 and f[1] == &#34;use_mask&#34;:
                    self.use_mask = False
                elif n &gt; 1 and f[1] == &#34;use_id&#34;:
                    self.use_id = False
                elif n &gt; 1 and f[1] == &#34;xkcd&#34; and self.xkcd:
                    plt.rcdefaults()
                    self.set_style(&#39;classic&#39;)
                    self.xkcd = False

                else:
                    print(&#34;Unrecognized unset argument!&#34;)
            else:

                try:
                    if f[0] not in [&#34;rdstat&#34;, &#34;baf&#34;]:
                        self.parse(f + [str(bin_size)], args)
                    else:
                        self.parse(f, args)
                    if len(pre) &gt; 1:
                        fns = pre[1].strip().split(&#34; &#34;)
                        if fns[0] != &#34;&#34;:
                            plt.savefig(fns[0], dpi=200)
                except Exception as e:
                    print(traceback.format_exc())
    except (EOFError, KeyboardInterrupt):
        print()
        return</code></pre>
</details>
</dd>
<dt id="cnvpytor.viewer.Viewer.rd"><code class="name flex">
<span>def <span class="ident">rd</span></span>(<span>self, bin_size, use_mask)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def rd(self, bin_size, use_mask):
    plt.clf()
    if self.reference_genome is None:
        _logger.warning(&#34;Missing reference genome required for gview.&#34;)
        return
    plt.rcParams[&#34;font.size&#34;] = 8
    chroms = []
    for c, (l, t) in self.reference_genome[&#34;chromosomes&#34;].items():
        rd_chr = self.io[self.plot_file].rd_chromosome_name(c)
        if self.io[self.plot_file].signal_exists(rd_chr, bin_size, &#34;RD&#34;, 0) and \
                self.io[self.plot_file].signal_exists(rd_chr, bin_size, &#34;RD&#34;, FLAG_GC_CORR) and \
                (Genome.is_autosome(c) or Genome.is_sex_chrom(c)):
            chroms.append((rd_chr, l))
    sx, sy = self.panels_shape(len(chroms))
    self.fig = plt.figure(1, figsize=(sx, sy), dpi=200, facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
    ix = 1
    for c, l in chroms:
        flag = FLAG_MT if Genome.is_mt_chrom(c) else FLAG_SEX if Genome.is_sex_chrom(c) else FLAG_AUTO
        stat = self.io[self.plot_file].get_signal(None, bin_size, &#34;RD stat&#34;, flag)
        if stat is None:
            _logger.error(&#34;Data for bin size %d is missing in file &#39;%s&#39;!&#34; % (bin_size, self.io[self.plot_file].filename))
            exit(0)
        flag_rd = 0
        if use_mask:
            flag_rd = FLAG_USEMASK
        his_p = self.io[self.plot_file].get_signal(c, bin_size, &#34;RD&#34;, flag_rd)
        his_p_corr = self.io[self.plot_file].get_signal(c, bin_size, &#34;RD&#34;, flag_rd | FLAG_GC_CORR)
        ax = plt.subplot(sx, sy, ix)
        ax.set_title(c, position=(0.01, 0.9), fontdict={&#39;verticalalignment&#39;: &#39;top&#39;, &#39;horizontalalignment&#39;: &#39;left&#39;},
                     color=&#39;C0&#39;)
        ax.xaxis.set_ticklabels([])
        ax.yaxis.set_ticklabels([])
        ax.yaxis.set_ticks(np.arange(0, 3, 0.5) * stat[4], [])
        ax.xaxis.set_ticks(np.arange(0, (l + 10e6) // bin_size, 10e6 // bin_size), [])
        ax.set_ylim([0, max(3. * stat[4], stat[4] + 5. * stat[5])])
        n_bins = l // bin_size
        ax.set_xlim([-n_bins * 0.05, n_bins * 1.05])
        ax.grid()

        plt.step(his_p, &#34;grey&#34;)
        plt.step(his_p_corr, &#34;k&#34;)
        ix += 1
    plt.subplots_adjust(bottom=0., top=1., wspace=0, hspace=0, left=0., right=1.)
    if self.output_filename != &#34;&#34;:
        plt.savefig(self.image_filename(&#34;gview&#34;), dpi=150)
        plt.close(self.fig)
    elif self.interactive:
        plt.show(block=False)
        plt.draw()
    else:
        plt.show()</code></pre>
</details>
</dd>
<dt id="cnvpytor.viewer.Viewer.region"><code class="name flex">
<span>def <span class="ident">region</span></span>(<span>self, io, element, bin_size, region, panels=['rd'], use_mask_rd=False, sep_color='g')</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def region(self, io, element, bin_size, region, panels=[&#34;rd&#34;], use_mask_rd=False, sep_color=&#34;g&#34;):
    snp_flag = (FLAG_USEMASK if self.use_mask else 0) | (FLAG_USEID if self.use_id else 0)
    grid = gridspec.GridSpecFromSubplotSpec(len(panels), 1, subplot_spec=element, wspace=0, hspace=0.1)
    r = decode_region(region)
    for i in range(len(panels)):
        ax = self.fig.add_subplot(grid[i])
        if i == 0:
            ax.set_title(io.filename + &#34;: &#34; + region, position=(0.01, 0.9),
                         fontdict={&#39;verticalalignment&#39;: &#39;top&#39;, &#39;horizontalalignment&#39;: &#39;left&#39;},
                         color=&#39;C0&#39;)
        g_p = []
        g_p_corr = []
        if panels[i] == &#34;rd&#34;:
            mean, stdev = 0, 0
            borders = []
            for c, (pos1, pos2) in r:
                flag = FLAG_MT if Genome.is_mt_chrom(c) else FLAG_SEX if Genome.is_sex_chrom(c) else FLAG_AUTO
                flag_rd = 0
                if use_mask_rd:
                    flag_rd = FLAG_USEMASK
                stat = io.get_signal(None, bin_size, &#34;RD stat&#34;, flag)
                mean = stat[4]
                stdev = stat[5]
                his_p = io.get_signal(c, bin_size, &#34;RD&#34;, flag_rd)
                his_p_corr = io.get_signal(c, bin_size, &#34;RD&#34;, flag_rd | FLAG_GC_CORR)
                start_bin = (pos1 - 1) // bin_size
                end_bin = pos2 // bin_size
                g_p.extend(list(his_p[start_bin:end_bin]))
                g_p_corr.extend(list(his_p_corr[start_bin:end_bin]))
                borders.append(len(g_p))

            # ax.xaxis.set_ticklabels([])
            ax.yaxis.set_ticklabels([])
            l = len(g_p)
            # ax.xaxis.set_ticks(np.arange(0, l, 10), [])
            ax.yaxis.set_ticks(np.arange(0, 3, 0.5) * mean, [])
            ax.set_ylim([0, max(3. * mean, mean + 5. * stdev)])
            ax.set_xlim([-l * 0.0, l * 1.0])

            ax.yaxis.grid()
            ax.step(g_p, &#34;grey&#34;)
            ax.step(g_p_corr, &#34;k&#34;)
            for i in borders[:-1]:
                ax.axvline(i, color=sep_color, lw=1)
            self.fig.add_subplot(ax)
        elif panels[i] == &#34;baf&#34;:
            borders = []
            hpos = []
            baf = []
            color = []
            start_pos = 0
            for c, (pos1, pos2) in r:
                pos, ref, alt, nref, nalt, gt, flag, qual = io.read_snp(c)
                ix = 0
                while ix &lt; len(pos) and pos[ix] &lt;= pos2:
                    if pos[ix] &gt;= pos1 and (nref[ix] + nalt[ix]) != 0:
                        hpos.append(start_pos + pos[ix] - pos1)
                        if gt[ix] % 4 != 2:
                            baf.append(1.0 * nalt[ix] / (nref[ix] + nalt[ix]))
                        else:
                            baf.append(1.0 * nref[ix] / (nref[ix] + nalt[ix]))
                        color.append(self.baf_colors[(gt[ix] % 4, flag[ix] &gt;&gt; 1)])
                    ix += 1
                start_pos += pos2 - pos1
                borders.append(start_pos)

            ax.xaxis.set_ticklabels([])
            ax.yaxis.set_ticklabels([])
            ax.yaxis.set_ticks([0, 0.25, 0.5, 0.75, 1.0], [])
            l = max(hpos)
            # ax.xaxis.set_ticks(np.arange(0, (l + 10e6), 10e6), [])
            ax.set_ylim([0., 1.])
            ax.set_xlim([0, borders[-1]])
            ax.yaxis.grid()
            ax.scatter(hpos, baf, marker=&#39;.&#39;, edgecolor=color, c=color, s=10, alpha=0.7)

            for i in borders[:-1]:
                ax.axvline(i, color=sep_color, lw=1)
            self.fig.add_subplot(ax)

        elif panels[i] == &#34;likelihood&#34;:
            borders = []
            gl = []
            for c, (pos1, pos2) in r:
                likelihood = io.get_signal(c, bin_size, &#34;SNP likelihood&#34;, snp_flag)
                start_bin = (pos1 - 1) // bin_size
                end_bin = pos2 // bin_size
                gl.extend(list(likelihood[start_bin:end_bin]))
                borders.append(len(gl))
            img = np.array(gl).transpose()
            ax.imshow(img, aspect=&#39;auto&#39;)
            ax.xaxis.set_ticklabels([])
            ax.yaxis.set_ticklabels([])
            ax.xaxis.set_ticks(np.arange(0, len(gl), 50), [])

            for i in borders[:-1]:
                ax.axvline(i, color=sep_color, lw=1)
            self.fig.add_subplot(ax)</code></pre>
</details>
</dd>
<dt id="cnvpytor.viewer.Viewer.regions"><code class="name flex">
<span>def <span class="ident">regions</span></span>(<span>self, bin_size, regions, panels=['rd'], use_mask_rd=False, sep_color='g')</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def regions(self, bin_size, regions, panels=[&#34;rd&#34;], use_mask_rd=False, sep_color=&#34;g&#34;):
    plt.clf()
    plt.rcParams[&#34;font.size&#34;] = 8
    self.fig = plt.figure(1, figsize=(12, 8), facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
    grid = gridspec.GridSpec(len(self.io), len(regions), wspace=0.2, hspace=0.2)
    ix = 0
    for i in self.io:
        for r in regions:
            self.region(i, grid[ix], bin_size, r, panels=panels, use_mask_rd=use_mask_rd, sep_color=sep_color)
            ix += 1
    plt.subplots_adjust(bottom=0.05, top=0.95, wspace=0, hspace=0, left=0.05, right=0.95)

    if self.output_filename != &#34;&#34;:
        plt.savefig(self.image_filename(&#34;regions&#34;), dpi=150)
        plt.close(self.fig)
    elif self.interactive:
        plt.show(block=False)
        plt.draw()
    else:
        plt.show()</code></pre>
</details>
</dd>
<dt id="cnvpytor.viewer.Viewer.stat"><code class="name flex">
<span>def <span class="ident">stat</span></span>(<span>self, his_bin_size=100)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def stat(self, his_bin_size=100):
    plt.clf()
    auto = self.io[0].signal_exists(None, his_bin_size, &#34;RD stat&#34;, FLAG_AUTO)
    sex = self.io[0].signal_exists(None, his_bin_size, &#34;RD stat&#34;, FLAG_SEX)
    mt = self.io[0].signal_exists(None, his_bin_size, &#34;RD stat&#34;, FLAG_MT) and (his_bin_size &lt; 1001)
    if not (auto or sex or mt):
        return
    cond = [auto, sex, mt]
    stat_list = []
    n_cols = sum(map(int, cond))
    ix = 1
    plt.rcParams[&#34;font.size&#34;] = 8
    self.fig = plt.figure(1, figsize=(4 * n_cols, 8), dpi=90, facecolor=&#39;w&#39;, edgecolor=&#39;k&#39;)
    for t, c, flag in zip([&#34;Autosomes&#34;, &#34;X/Y&#34;, &#34;Mitochondria&#34;], cond, [FLAG_AUTO, FLAG_SEX, FLAG_MT]):
        if c:
            stat = self.io[0].get_signal(None, his_bin_size, &#34;RD stat&#34;, flag)
            stat_list.append(stat)
            max_rd = int(stat[0])
            bin_size = int(stat[1])
            n_bins = int(stat[2])
            lim_rd = int(max(2 * stat[4], stat[4] + 3 * stat[5]))
            _logger.info(&#34;RD stat for %s: %.2f +- %.2f&#34; % (t, stat[4], stat[5]))
            if t == &#34;Mitochondria&#34; and auto:
                _logger.info(&#34;RD stat for %s - number of miochondria per cell: %.2f +- %.2f&#34; % (
                    t, 2 * stat[4] / stat_list[0][4],
                    2 * stat[5] / stat_list[0][4] + stat_list[0][5] * stat[4] / (
                            stat_list[0][4] * stat_list[0][4])))
            his_p = self.io[0].get_signal(None, his_bin_size, &#34;RD p dist&#34;, flag)
            his_u = self.io[0].get_signal(None, his_bin_size, &#34;RD u dist&#34;, flag)
            his_rd_gc = self.io[0].get_signal(None, his_bin_size, &#34;RD GC dist&#34;, flag)
            gc_corr = self.io[0].get_signal(None, his_bin_size, &#34;GC corr&#34;, flag)
            ax = plt.subplot(2, n_cols, ix)
            ax.set_xlabel(&#34;RD&#34;)
            ax.set_ylabel(&#34;GC [%]&#34;)
            ax.xaxis.set_ticklabels([])
            ax.set_title(t)
            his_rd_gc[0][0] = 0
            ax.imshow(his_rd_gc[:lim_rd // bin_size, :].T, aspect=&#34;auto&#34;, interpolation=&#39;nearest&#39;, origin=&#39;lower&#39;)
            ax.plot(gc_corr * stat[4] / bin_size, range(101), &#34;w-&#34;)

            ax = plt.subplot(2, n_cols, ix + n_cols)
            ax.set_ylabel(&#34;Normalised distribution&#34;)
            ax.set_xlabel(&#34;RD&#34;)
            ax.set_xlim([0, lim_rd])
            ax.set_ylim([0, 1.1])
            bins = range(0, max_rd, bin_size)
            x = np.arange(0, max_rd // bin_size * bin_size, 0.1 * bin_size)
            plt.plot(x, normal(x, 1, stat[4], stat[5]), &#34;g-&#34;)
            x = np.array(bins)
            plt.plot(x[:len(his_u)], his_u / stat[3], &#34;y*&#34;)
            plt.plot(x[:len(his_p)], his_p / stat[3], &#34;b*&#34;)
            ix += 1
    plt.subplots_adjust(bottom=0.08, top=0.95, wspace=0.25, hspace=0, left=0.05 * 3 / n_cols, right=0.95)
    if self.output_filename != &#34;&#34;:
        plt.savefig(self.image_filename(&#34;stat&#34;), dpi=150)
        plt.close(self.fig)
    elif self.interactive:
        plt.show(block=False)
        plt.draw()
    else:
        plt.show()</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="cnvpytor" href="index.html">cnvpytor</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="cnvpytor.viewer.Viewer" href="#cnvpytor.viewer.Viewer">Viewer</a></code></h4>
<ul class="two-column">
<li><code><a title="cnvpytor.viewer.Viewer.baf" href="#cnvpytor.viewer.Viewer.baf">baf</a></code></li>
<li><code><a title="cnvpytor.viewer.Viewer.image_filename" href="#cnvpytor.viewer.Viewer.image_filename">image_filename</a></code></li>
<li><code><a title="cnvpytor.viewer.Viewer.likelihood" href="#cnvpytor.viewer.Viewer.likelihood">likelihood</a></code></li>
<li><code><a title="cnvpytor.viewer.Viewer.manhattan" href="#cnvpytor.viewer.Viewer.manhattan">manhattan</a></code></li>
<li><code><a title="cnvpytor.viewer.Viewer.panels_shape" href="#cnvpytor.viewer.Viewer.panels_shape">panels_shape</a></code></li>
<li><code><a title="cnvpytor.viewer.Viewer.parse" href="#cnvpytor.viewer.Viewer.parse">parse</a></code></li>
<li><code><a title="cnvpytor.viewer.Viewer.plot" href="#cnvpytor.viewer.Viewer.plot">plot</a></code></li>
<li><code><a title="cnvpytor.viewer.Viewer.prompt" href="#cnvpytor.viewer.Viewer.prompt">prompt</a></code></li>
<li><code><a title="cnvpytor.viewer.Viewer.rd" href="#cnvpytor.viewer.Viewer.rd">rd</a></code></li>
<li><code><a title="cnvpytor.viewer.Viewer.region" href="#cnvpytor.viewer.Viewer.region">region</a></code></li>
<li><code><a title="cnvpytor.viewer.Viewer.regions" href="#cnvpytor.viewer.Viewer.regions">regions</a></code></li>
<li><code><a title="cnvpytor.viewer.Viewer.set_style" href="#cnvpytor.viewer.Viewer.set_style">set_style</a></code></li>
<li><code><a title="cnvpytor.viewer.Viewer.stat" href="#cnvpytor.viewer.Viewer.stat">stat</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.6.3</a>.</p>
</footer>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script>hljs.initHighlightingOnLoad()</script>
</body>
</html>